
The Willisornis vidua nigrigula genome was assembled using Supernova. Since the assembly contiguity can vary with the number of input reads, it was assembled with varying values for the max numbers of reads: 450, 500, 550, 600, 650, 700, and 750. The assemblies were done on a supercomputer run by Compute Canada, and the jobs were submitted as bash scripts, archived below in Appendix A.

The assembly reports which give assembly statistics are given in this Github repository under `QC_Reports/Assembly_reports`. Note that the values in these reports differ slightly from those presented in the manuscript because Supernova automatically discounts scaffolds smaller than 10 kb in its calculations, but I incuded small scaffolds in the statistics that I reported. 

After genome assembly, the assembly was turned into a pseudohaploid genome sequence in fasta format:

### Required software:
* supernova v2.1.1

## Make a Fasta
After assembly, we generate a fasta file from the supernova run with the following command:

```bash
/opt/NGS_Mapping/supernova-2.1.1/supernova mkoutput --style=pseudohap --asmdir=/path/to/folder/0_GENOME_ASSEMBLIES/Willisornis_JTW11144__700milReads__2019Nov12/Willisornis_JTW11144__700milReads__2019Nov12/outs/assembly --outprefix=/path/to/folder/0_GENOME_ASSEMBLIES/Willisornis_JTW11144__700milReads__2019Nov12/Willisornis_JTW11144__700milReads__2019Nov12/FASTAOUTPUT/Willisornis_poecilinotus_JTW1144_700mill --minsize=1000 --headers=short
```
* `minsize`: the minimum size of scaffold that will be placed into the output folder. `--minsize=1000` excludes very small scaffolds from the assembly as these are not useful for most analyses.

This took 5 minutes and produced `Willisornis_poecilinotus_JTW1144_700mill.fasta.gz`

**Output**:  
```
supernova mkoutput (2.1.1)
Copyright (c) 2018 10x Genomics, Inc.  All rights reserved.
-------------------------------------------------------------------------------

Thu Nov 21 23:48:23 2019: using 1 threads

--------------------------------------------------------------------------------
Thu Nov 21 23:48:23 2019 on weirlab999, pid=30503 [Aug 28 2018 15:15:30 6bb16452a]
MakeFasta OUT_HEAD=Willisornis_poecilinotus_JTW1144_700mill                    \
          DIR=/path/to/folder/0_GENOME_ASSEMBLIES \
          /Willisornis_JTW11144__700milReads__2019Nov12/Willisornis_JTW11144__ \
          700milReads__2019Nov12/outs/assembly/a.base/gang/micro/final         \
          FLAVOR=pseudohap MINSIZE=1000 ABBREV=True
--------------------------------------------------------------------------------

version=1.10 style=3
Thu Nov 21 23:48:23 2019: loading assembly files
Thu Nov 21 23:49:33 2019: cell reinsertion
Thu Nov 21 23:51:08 2019: building lines-of-lines
Thu Nov 21 23:51:08 2019: forming lines of lines
Thu Nov 21 23:51:09 2019: found 230797 lines, and 226629 lines of lines
Thu Nov 21 23:51:10 2019: writing the FASTA portion of output to Willisornis_poecilinotus_JTW1144_700mill.fasta
writing 100 dots in 10 groups of 10:
.......... .......... .......... .......... ..........
.......... .......... .......... .......... ..........
writing gzipped object
Thu Nov 21 23:53:11 2019: graph info
final graph has 1645078 edges
reinserted cells start at edge 1192718
14 cell(s) at dead-ends
completed conversion in 4.8 minutes
```

# Appendix A: Bash scripts used to run supernova for genome assembly
Here are the contents of scripts used to run the supernova commands.
```bash
cat > 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_450mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_450mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__450milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=450000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```

```bash
cat > 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_500mill.sh
```
```bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_500mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__500milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=500000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS

```

```bash
cat > 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_550mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_550mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__550milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=550000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```

```bash
cat > 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_600mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_600mill
#SBATCH --output=OUTPUT_TEST_600mill_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__550milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=600000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```

```bash
cat > 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_650mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_650mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__650milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=650000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```

```bash
cat 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_700mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_700mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__700milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=700000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```

```bash
cat 0_GRAHAM_submitFile__ChromiumXSupernova_Willisornis_750mill.sh
```
```bash
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --cpus-per-task=32
#SBATCH --mem=240G
#SBATCH --time=53:59:00
#SBATCH --job-name weir_ChromiumX_Xiphorhynchus_750mill
#SBATCH --output=OUTPUT_TEST1_%.txt

module load supernova/2.1.1

# EXECUTION COMMAND; 
export OMP_NUM_THREADS=32

#User sets these options
   WD=/path/to/folder/
   SAMPLE=Willisornis_JTW11144
   VERSION=__750milReads__2019Nov12
   OUTPUT=/path/to/folder/0_ASSEMBLIES/$SAMPLE$VERSION
   INPUT1=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2017Sept
   INPUT2=/path/to/folder/0_RAW_GENOMIC_DATA/Willisornis_poecilinotus_JTW1144/2018Jan

   NCORES=32
   LOCALMEM=220
   MAXREADS=750000000


  #automated code to run (don't change)
   ASSEMBLY=$SAMPLE$VERSION
   mkdir $OUTPUT
   cd $OUTPUT
   supernova run --id=$ASSEMBLY \
              --fastqs=$INPUT1,$INPUT2 --sample=4717-JW-0008,4717-JW-0008 \
              --localcores=$NCORES   --localmem=$LOCALMEM --maxreads=$MAXREADS
```