# BUSCO Analysis.

Now we have a genome sequence. How good is it? One way of assessing is using BUSCO, which uses databases of conserved single-copy genes and checks how many are present in your genome sequence.

```bash
#set up folders
mkdir -p /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis
mkdir -p /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco

#obtain copy of the genome sequence and uncompress it
cp /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis_poecilinotus_JTW1144_700mill.fasta.gz ./Willisornis.fasta.gz
gunzip Willisornis.fasta.gz

#get config file. Edit if needed (make sure paths to prerequisite programs are correct. My default also has tblastn not forced to single threaded, and augustus species is chicken).
cp /home/0_PROGRAMS/busco/config/config.ini ./Willisornis_busco_config.ini

#set up environment with the prerequisites for busco
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/Willisornis_busco_config.ini"
export PATH=$PATH:/home/0_PROGRAMS/hmmer-3.3/bin/

#Run BUSCO!
python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i Willisornis.fasta -o Willisornis_BUSCO -l /home/0_PROGRAMS/busco/aves_odb9 -m geno -c 24 -sp chicken
```
* `-m geno`: assessment mode (genome, rather than transcriptome or proteome)
* `-l /home/0_PROGRAMS/busco/aves_odb9`: path to the lineage file of gene models
* `-i Willisornis.fasta`: input genome sequence
* `-o Willisornis_BUSCO`: output prefix
* `-c 24`: number of cores
* `-sp chicken`: Augustus gene model to use. chicken is the only bird currently available.

**Notes**: 
* I had to specify `-sp chicken` on the command line as it otherwise seemed to override what I wrote in the config.ini file and defaulted to human, which is not desired.
* The location of Augustus config must have read/write permission for you. I did not so I added permissions to all my conda environments with `chmod -R 755 envs`.

**Timing**: Took only 4h24m to run on 24 threads.
**Output**:  
```
WARNING	An augustus species is mentioned in the config file, dataset default species (human) will be ignored
INFO	****************** Start a BUSCO 3.1.0 analysis, current time: 11/22/2019 07:54:32 ******************
INFO	Configuration loaded from /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/Willisornis_busco_config.ini
INFO	Init tools...
INFO	Check dependencies...
INFO	Check input file...
INFO	To reproduce this run: python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i Willisornis.fasta -o Willisornis_BUSCO -l /home/0_PROGRAMS/busco/aves_odb9/ -m genome -c 24 -sp chicken
INFO	Mode is: genome
INFO	The lineage dataset is: aves_odb9 (eukaryota)
INFO	Temp directory is ./tmp/
INFO	****** Phase 1 of 2, initial predictions ******
INFO	****** Step 1/3, current time: 11/22/2019 07:54:43 ******
INFO	Create blast database...
INFO	[makeblastdb]	Building a new DB, current time: 11/22/2019 07:54:43
INFO	[makeblastdb]	New DB name:   /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/tmp/Willisornis_BUSCO_3459228854
INFO	[makeblastdb]	New DB title:  Willisornis.fasta
INFO	[makeblastdb]	Sequence type: Nucleotide
INFO	[makeblastdb]	Keep MBits: T
INFO	[makeblastdb]	Maximum file size: 1000000000B
INFO	[makeblastdb]	Adding sequences from FASTA; added 23036 sequences in 15.4716 seconds.
INFO	[makeblastdb]	1 of 1 task(s) completed at 11/22/2019 07:54:58
INFO	Running tblastn, writing output to /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/blast_output/tblastn_Willisornis_BUSCO.tsv...
INFO	[tblastn]	1 of 1 task(s) completed at 11/22/2019 08:45:00
INFO	****** Step 2/3, current time: 11/22/2019 08:45:00 ******
INFO	Maximum number of candidate contig per BUSCO limited to: 3
INFO	Getting coordinates for candidate regions...
INFO	Pre-Augustus scaffold extraction...
INFO	Running Augustus prediction using chicken as species:
INFO	[augustus]	Please find all logs related to Augustus errors here: /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/augustus_output/augustus.log
INFO	[augustus]	696 of 6952 task(s) completed at 11/22/2019 08:59:17
INFO	[augustus]	1391 of 6952 task(s) completed at 11/22/2019 09:11:36
INFO	[augustus]	2086 of 6952 task(s) completed at 11/22/2019 09:24:00
INFO	[augustus]	2781 of 6952 task(s) completed at 11/22/2019 09:35:48
INFO	[augustus]	3477 of 6952 task(s) completed at 11/22/2019 09:46:51
INFO	[augustus]	4172 of 6952 task(s) completed at 11/22/2019 09:58:17
INFO	[augustus]	4867 of 6952 task(s) completed at 11/22/2019 10:11:01
INFO	[augustus]	5562 of 6952 task(s) completed at 11/22/2019 10:21:05
INFO	[augustus]	6257 of 6952 task(s) completed at 11/22/2019 10:32:19
INFO	[augustus]	6952 of 6952 task(s) completed at 11/22/2019 10:44:19
INFO	Extracting predicted proteins...
INFO	****** Step 3/3, current time: 11/22/2019 10:45:11 ******
INFO	Running HMMER to confirm orthology of predicted proteins:
INFO	[hmmsearch]	6228 of 6919 task(s) completed at 11/22/2019 10:45:37
INFO	[hmmsearch]	6919 of 6919 task(s) completed at 11/22/2019 10:45:39
INFO	Results:
INFO	C:89.7%[S:88.2%,D:1.5%],F:6.5%,M:3.8%,n:4915
INFO	4411 Complete BUSCOs (C)
INFO	4336 Complete and single-copy BUSCOs (S)
INFO	75 Complete and duplicated BUSCOs (D)
INFO	319 Fragmented BUSCOs (F)
INFO	185 Missing BUSCOs (M)
INFO	4915 Total BUSCO groups searched
INFO	****** Phase 2 of 2, predictions using species specific training ******
INFO	****** Step 1/3, current time: 11/22/2019 10:45:40 ******
INFO	Extracting missing and fragmented buscos from the ancestral_variants file...
INFO	Running tblastn, writing output to /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/blast_output/tblastn_Willisornis_BUSCO_missing_and_frag_rerun.tsv...
INFO	[tblastn]	1 of 1 task(s) completed at 11/22/2019 11:27:59
INFO	Maximum number of candidate contig per BUSCO limited to: 3
INFO	Getting coordinates for candidate regions...
INFO	****** Step 2/3, current time: 11/22/2019 11:28:02 ******
INFO	Training Augustus using Single-Copy Complete BUSCOs:
INFO	Converting predicted genes to short genbank files at 11/22/2019 11:28:02...
INFO	All files converted to short genbank files, now running the training scripts at 11/22/2019 12:03:08...
INFO	Pre-Augustus scaffold extraction...
INFO	Re-running Augustus with the new metaparameters, number of target BUSCOs: 504
INFO	[augustus]	79 of 788 task(s) completed at 11/22/2019 12:04:48
INFO	[augustus]	158 of 788 task(s) completed at 11/22/2019 12:05:59
INFO	[augustus]	237 of 788 task(s) completed at 11/22/2019 12:07:07
INFO	[augustus]	316 of 788 task(s) completed at 11/22/2019 12:08:22
INFO	[augustus]	394 of 788 task(s) completed at 11/22/2019 12:09:51
INFO	[augustus]	473 of 788 task(s) completed at 11/22/2019 12:11:14
INFO	[augustus]	552 of 788 task(s) completed at 11/22/2019 12:12:54
INFO	[augustus]	631 of 788 task(s) completed at 11/22/2019 12:14:24
INFO	[augustus]	710 of 788 task(s) completed at 11/22/2019 12:16:02
INFO	[augustus]	788 of 788 task(s) completed at 11/22/2019 12:18:44
INFO	Extracting predicted proteins...
INFO	****** Step 3/3, current time: 11/22/2019 12:18:50 ******
INFO	Running HMMER to confirm orthology of predicted proteins:
INFO	[hmmsearch]	Error: Failed to open sequence file /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/augustus_output/extracted_proteins/EOG090F0AQS.faa.2 for reading
INFO	[hmmsearch]	Error: Failed to open sequence file /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/augustus_output/extracted_proteins/EOG090F0A90.faa.2 for reading
INFO	[hmmsearch]	Error: Failed to open sequence file /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/augustus_output/extracted_proteins/EOG090F00W8.faa.2 for reading
INFO	[hmmsearch]	386 of 771 task(s) completed at 11/22/2019 12:18:51
INFO	[hmmsearch]	771 of 771 task(s) completed at 11/22/2019 12:18:52
INFO	Results:
INFO	C:91.6%[S:90.1%,D:1.5%],F:5.1%,M:3.3%,n:4915
INFO	4502 Complete BUSCOs (C)
INFO	4427 Complete and single-copy BUSCOs (S)
INFO	75 Complete and duplicated BUSCOs (D)
INFO	250 Fragmented BUSCOs (F)
INFO	163 Missing BUSCOs (M)
INFO	4915 Total BUSCO groups searched
INFO	BUSCO analysis done with WARNING(s). Total running time: 15862.3075092 seconds
INFO	Results written in /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/busco/run_Willisornis_BUSCO/
```

#### Interpretation
A quite complete genome! Only 3.3% of buscos are missing.

# Sequencing Quality
In this step, I am looking at the raw sequence reads.  
I am also assesssing "mappability". This is not an analysis for the publication. In this test, I want to see what proportion of the sequencing reads are able to be mapped back to the genome assembly. We can use a standard mapping tool, `bwa-mem` to map our reads. This is also nice because it uses soft-clipping, so is able to ignore any low-quality or adapter sequences at the ends of our reads, so we do not have to do trimming before hand. Of course, if we cared about doing downstream variant calling we probably *would* care about trimming adaptrs, but for this exercise it is not strictly necessary. Mappability should be high since we are testing the same data that was used for assembly, and would be concerning if it were low.

We have sequenced these reads:

in`/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/0_RAW_GENOMIC_DATA/Hudson_Alpha_ChromiumX_2017_2018_2019/Willisornis_poecilinotus_JTW11144/2017Sept`:
* `4717-JW-0008_S5_L006_R1_001.fastq.gz` + `4717-JW-0008_S5_L006_R2_001.fastq.gz` 
* `4717-JW-0008_S5_L007_R1_001.fastq.gz` + `4717-JW-0008_S5_L007_R2_001.fastq.gz`
in `2018Jan`:
* `4717-JW-0008_S4_L002_R1_001.fastq.gz` + `4717-JW-0008_S4_L002_R2_001.fastq.gz` 
* `4717-JW-0008_S4_L003_R1_001.fastq.gz` + `4717-JW-0008_S4_L003_R2_001.fastq.gz`

```bash
#set up environment with executables
export PATH=$PATH:/home/0_PROGRAMS/bwa/ #for bwa executable
export PATH=$PATH:/home/0_PROGRAMS/samtools-1.9/ #for samtools executable
export PATH=$PATH:/home/0_PROGRAMS/bamtools/bin/

mkdir -p /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/genome
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/genome

#get a copy of the genome sequence
cp Willisornis_poecilinotus_JTW1144_700mill.fasta.gz genome/Willisornis700.fasta.gz
gunzip genome/Willisornis700.fasta.gz

#Index the genome for bwa and samtools
bwa index /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/genome/Willisornis700.fasta
time samtools faidx /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/genome/Willisornis700.fasta

#get the sequencing reads
scp $USERNAME@$IP_ADDRESS:/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/0_RAW_GENOMIC_DATA/Hudson_Alpha_ChromiumX_2017_2018_2019/Willisornis_poecilinotus_JTW1144/2017Sept/4717-JW-0008_S5_L006_R*_001.fastq.gz .

scp $USERNAME@$IP_ADDRESS:/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/0_RAW_GENOMIC_DATA/Hudson_Alpha_ChromiumX_2017_2018_2019/Willisornis_poecilinotus_JTW1144/2017Sept/4717-JW-0008_S5_L007_R*_001.fastq.gz .

scp $USERNAME@$IP_ADDRESS:/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/0_RAW_GENOMIC_DATA/Hudson_Alpha_ChromiumX_2017_2018_2019/Willisornis_poecilinotus_JTW1144/2018Jan/4717-JW-0008_S4_L002_R*_001.fastq.gz .

scp $USERNAME@$IP_ADDRESS:/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/0_RAW_GENOMIC_DATA/Hudson_Alpha_ChromiumX_2017_2018_2019/Willisornis_poecilinotus_JTW1144/2018Jan/4717-JW-0008_S4_L003_R*_001.fastq.gz .

#glimpse the raw sequencing quality
conda activate fastqc
time fastqc ./*fastq.gz -t 8

#trim adapters and low quality trailing bases
time java -jar /hhome/$USERNAME/tools/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 6 4717-JW-0008_S4_L002_R1_001.fastq.gz 4717-JW-0008_S4_L002_R2_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L002_R1_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S4_L002_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L002_R2_001.fastq.gz  fltrd_adaptrlss_unpaired4717-JW-0008_S4_L002_R2_001.fastq.gz HEADCROP:16 TRAILING:3 LEADING:3 SLIDINGWINDOW:4:15 MINLEN:40 2> 4717-JW-0008_S4_L002_qualtrim_out.log #it is important that HEADCROP is done first!

#trim adapters and low quality trailing bases
time java -jar /hhome/$USERNAME/tools/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 6 4717-JW-0008_S4_L003_R1_001.fastq.gz 4717-JW-0008_S4_L003_R2_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L003_R1_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S4_L003_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L003_R2_001.fastq.gz  fltrd_adaptrlss_unpaired4717-JW-0008_S4_L003_R2_001.fastq.gz HEADCROP:16 TRAILING:3 LEADING:3 SLIDINGWINDOW:4:15 MINLEN:40 2> 4717-JW-0008_S4_L003_qualtrim_out.log #it is important that HEADCROP is done first!

time java -jar /hhome/$USERNAME/tools/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 6 4717-JW-0008_S5_L006_R1_001.fastq.gz 4717-JW-0008_S5_L006_R2_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L006_R1_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S5_L006_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L006_R2_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S5_L006_R2_001.fastq.gz HEADCROP:16 TRAILING:3 LEADING:3 SLIDINGWINDOW:4:15 MINLEN:40 2> 4717-JW-0008_S5_L006_qualtrim_out.log #it is important that HEADCROP is done first!

time java -jar /hhome/$USERNAME/tools/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 6 4717-JW-0008_S5_L007_R1_001.fastq.gz 4717-JW-0008_S5_L007_R2_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L007_R1_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S5_L007_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L007_R2_001.fastq.gz fltrd_adaptrlss_unpaired4717-JW-0008_S5_L007_R2_001.fastq.gz HEADCROP:16 TRAILING:3 LEADING:3 SLIDINGWINDOW:4:15 MINLEN:40 2> 4717-JW-0008_S5_L007_qualtrim_out.log #it is important that HEADCROP is done first!

#then, map your reads to the assembly. Compress the mapping files to bam format as you go to save space.
#raw umtrimmed reads, to observe effects of trimming
time bwa mem -t 23 Willisornis700.fasta 4717-JW-0008_S5_L006_R1_001.fastq.gz 4717-JW-0008_S5_L006_R2_001.fastq.gz 2> Willisornis1_mapping.log | time samtools view -S -b -@ 8 -h > Willisornis1.bam 


#map trimmed reads
time bwa mem -t 23 Willisornis700.fasta fltrd_adaptrlss4717-JW-0008_S5_L006_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L006_R2_001.fastq.gz 2> 4717-JW-0008_S5_L006_mapping.log | time samtools view -S -b -@ 8 -h > 4717-JW-0008_S5_L006.bam 

time bwa mem -t 23 Willisornis700.fasta fltrd_adaptrlss4717-JW-0008_S4_L002_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L002_R2_001.fastq.gz 2> 4717-JW-0008_S4_L002_mapping.log | time samtools view -S -b -@ 8 -h > 4717-JW-0008_S4_L002.bam 

time bwa mem -t 23 Willisornis700.fasta fltrd_adaptrlss4717-JW-0008_S4_L003_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S4_L003_R2_001.fastq.gz 2> 4717-JW-0008_S4_L003_mapping.log | time samtools view -S -b -@ 8 -h > 4717-JW-0008_S4_L003.bam 

time bwa mem -t 23 Willisornis700.fasta fltrd_adaptrlss4717-JW-0008_S5_L007_R1_001.fastq.gz fltrd_adaptrlss4717-JW-0008_S5_L007_R2_001.fastq.gz 2> 4717-JW-0008_S5_L007_mapping.log | time samtools view -S -b -@ 8 -h > 4717-JW-0008_S5_L007.bam 

### Statistics
#view some mapping statistics
run=4717-JW-0008_S5_L006
run=4717-JW-0008_S4_L002
run=4717-JW-0008_S4_L003
run=4717-JW-0008_S5_L007

time samtools flagstat "$run".bam > "$run".flagstats.txt
time bamtools stats -in "$run".bam > "$run".bamstats.txt
time samtools stats "$run".bam > "$run".samstats.txt

#Sort by position and index the bam file in order to get indexstats
time samtools sort --threads=14 -o "$run"_sorted.bam "$run".bam
time samtools index "$run"_sorted.bam
time samtools idxstats "$run"_sorted.bam > "$run".idxstats.txt

#take a more in depth look
time ~/tools/qualimap_v2.2.1/qualimap bamqc -bam "$run"_sorted.bam -c -ip -nt 23 --java-mem-size=4G -outdir "$run"_qualimap

#at this point, delete all the raw sequencing reads (assuming they are intact on the main server), and unsorted bam files to save space. Keep the sorted bam files and the trimmed sequencing reads for the next analyses.
```
* `-t`: number of threads
* `-S`: the input is a SAM file (no longer required as this can be automatically detected)
* `-b`: the output will be BAM format
* `-@`: number of threads to use for BAM compression
* `-h`: include the header in the output

**Timing**: indexing genome: 1185s real, 1172s CPU; bwa raw data 182m53 real, 4183m44 user; bwa trimmed data 51m7 real, 1161m38 user / 87m27 real, 2001m8 user (23 threads). bamstats: 7m7s. samstats: 16m45 (1 thread). flagstats: 2m46 (1 thread). bamstats: 6m53 (1 thread).
Trimmomatic: 21m36/45m3 (64m27/111m30 user) on 6 threads; sorting takes 49m13 (98m48 user) on 4 threads. Qualimap: 26m45 (211m52 user) on 23 threads.

**Results**
I have deposited the raw output files in this Github repository in the folder QC_reports. 
* FASTQC tells you # of reads, %GC, and a visualization of various sequence quality metrics
* samtools flagstat tells you % mapping, % properly paired, % singleton and a bit more.
* bamstats tells you the same thing as flagstat but I like to double check everything: # reads, %mapped, %proper pairs, % singletone
* samtools stats tells much much of the same thing but also average mapping quality, and insert sizes.

Sequence set|Reads, untrimmed (FASTQC)|%GC (FASTQC)|Reads, trimmed (FastQC)|Raw mapping % (flagstats/bamstats)|properly paired % (flagstats/bamstats)|singleton % (flagstats/bamstats)|average insert size, bp (stdev) (samstats)|avg quality (samstats)|PE survival trimming|
---|---|---|---|---|---|---|---|---|---|
4717-JW-0008_S4_L002| 64,732,037|50|49,236,912|95.66%/94.9195%|83.66%/75.6627%|1.39%/2.38742%|270.6 (140.7)|38.4|76.06%|
4717-JW-0008_S4_L003|63,905,978|51|49,396,233|95.83%/95.8261%|84.11%/82.2016%|1.30%/1.42399%|272.2 (140.5)|38.5|77.30%|
4717-JW-0008_S5_L006|110,595,610|48|91,752,571|96.53%/94.9195%|85.96%/75.6627%|1.04%/2.38742%|270.9 (129.2)|38.8|82.96%|
4717-JW-0008_S5_L006 (untrimmed)|110,595,610|48|NA|94.92|78.27/75.7|1.39%/2.07%|278.1 (139.3)|35.7|82.96|NA|
4717-JW-0008_S5_L007|110,780,612|47|93,279,371|96.81%/96.8131%|86.49%/84.6005%|0.95%/1.03162%|262.5 (121.8)|38.8|84.20%|

In the read2 reads, there is a sharp secondary peak at 50% GC content that appears to be caused by an overrepresented AACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCT repetitive sequence (and its reverse complement) in the reads. 

The trimmed reads have more mapping success than the unfiltered reads as expected (quality trimming removes poor quality reads that are less likely to map properly)

**Mappability (% reads that are mappable to the assembly)**:
* Will report the % of trimmed reads that map to the assembly by taking the weighted average of the sequencing runs. I will use the values reported by Qualimap/flagstats rather than bamstats because I believe bamstats is not counting the multimapped reads, but in my case I am just interested in what proportion of the reads can be mapped to the genome so I do want to include these possibly repetitive sequences if they have been assembled into the genome.
* `[(95.66*49,236,912)+(95.83*49,396,233)+(96.53*91,752,571)+(96.81*93,279,371)]/(91,752,571+93,279,371+49,236,912+49,396,233)` = **96.349**%

**Reportable values for manuscript**:
In the genome report manuscript I will need to report:  
1) The number of reads that were sequenced. I am reported the number of raw, untrimmed sequences (that is what was given as input to the assembler, the trimming was just for mapping/kmer analyses)
* `64,732,037+63,905,978+110,595,610+110,780,612=`**350,014,237** read pairs (700,028,474 individual reads)
* `700,028,474 reads * 151 bp/read =` **105,704,299,574 bp** of sequencing data (106 Gbp)

2) Average insert size
* I am using the average insert size given by the actual assembler (supernova, 308 bp median) since that is able to use the information about the linked reads to place the reads into an assembly, while that information is not used when simply mapping the reads to the genome.

