
## TE-free Protein Database
Ensure you are not masking real proteins! See a discussion [here](https://www.biostars.org/p/347385/)

The goal is to download a database of "real" bird protein sequences, which we would *not* like to mask, and then remove any putative "transposons" in our library that actually match one of these real proteins.  
You can use the uniprot-swiss prot database, which is a [high quality, curated database of protein sequences](https://www.uniprot.org/help/uniprotkb_sections).  
The high quality, intensively curated nature of of SwissProt means that it does not contain all possible proteins that have been discovered - there are not enough resources to curate them all.  
We can supplement (or replace) the SwissProt database with some more proteomes closer to our species of interest. I chose four favourite model taxa which have reference-quality proteomes; different species can be chosen if preferred for your taxon of interest.  
In the end I used only the four avian proteomes (46.5 Mb) without the entire SwissProt database (276.7 Mb) to speed it up.  

Many TEs actually do contain "real" proteins (such as proteins involved in TE transposition) - but we actually do want to mask these TE sequences! So, before accidentally removing all the TE proteins from our TE library, we need to remove TE proteins from our protein database.  

There is a curated TE protein database available through RepeatMasker, which comes with the program. I have it in `~/tools/RepeatMasker/Libraries/RepeatPeps.lib` but it can also be downloaded separately from [here](http://www.repeatmasker.org/libraries/).

```bash
#mkdir ~/Genetic_resources #(run this if this folder is not already made)
cd Genetic_resources 

#get the uniprot/swissprot database if desired
#wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
#gunzip uniprot_sprot.fasta.gz

#Ficedula albicollis proteome
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000016665_59894.fasta.gz
gunzip UP000016665_59894.fasta.gz
#rename the file
mv UP000016665_59894.fasta FicAlbProteome_UP000016665_59894.fasta

#Manacus vitellinus proteome
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000053258_328815.fasta.gz
gunzip UP000053258_328815.fasta.gz
#rename the file
mv UP000053258_328815.fasta ManVitProteome_UP000053258_328815.fasta 

#Gallus gallus proteome
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000000539_9031.fasta.gz
gunzip UP000000539_9031.fasta.gz
#rename the file
mv UP000000539_9031.fasta GalGalProteome_UP000053258_328815.fasta

#Taeniopygia guttata proteome
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000007754_59729.fasta.gz
gunzip UP000007754_59729.fasta.gz
#rename the file
mv UP000007754_59729.fasta TaeGutProteome_UP000007754_59729.fasta

#Some Zebra Finch and Chicken sequences are already in Swiss Prot, to avoid duplicates, just download the other proteins (use this if you are including SwissProt, otherwise ignore)
#Get non-SwissProt Zebra Finch sequences from here: https://www.uniprot.org/uniprot/?query=proteome:UP000007754%20reviewed:no#
#Download -> Go
#mv ~/Downloads/uniprot-proteome%3AUP000007754+reviewed%3Ano.fasta.gz ./
#gunzip uniprot-proteome%3AUP000007754+reviewed%3Ano.fasta.gz
#rename the file
#mv uniprot-proteome%3AUP000007754+reviewed%3Ano.fasta TaeGutProteome_UP000007754.fasta #Taeniopygia guttata

#Get non-SwissProt Chicken sequences from here: https://www.uniprot.org/uniprot/?query=proteome:UP000000539%20reviewed:no
#Download -> Go
#mv ~/Downloads/uniprot-proteome%3AUP000000539+reviewed%3Ano.fasta.gz ./
#gunzip uniprot-proteome%3AUP000000539+reviewed%3Ano.fasta.gz
#rename the file
#mv uniprot-proteome%3AUP000000539+reviewed%3Ano.fasta GalGalProteome_000000539.fasta #Gallus gallus

#combine the Ficedula, Taeniopygia, Manacus, and Gallus proteomes with or without the SwissProt database, and also create a smaller database with only the bird proteomes.
#cat FicAlbProteome_UP000016665_59894.fasta TaeGutProteome_UP000007754.fasta ManVitProteome_UP000053258_328815.fasta GalGalProteome_000000539.fasta uniprot_sprot.fasta > SwissProt_RefAves_proteins.fasta
cat FicAlbProteome_UP000016665_59894.fasta TaeGutProteome_UP000007754_59729.fasta ManVitProteome_UP000053258_328815.fasta GalGalProteome_UP000053258_328815.fasta > RefAves_proteins.fasta

#get the RepeatMasker RepeatPeps.lib library of TE proteins
cp ~/tools/RepeatMasker/Libraries/RepeatPeps.lib ./
```

Now, BLAST this `RefAves_proteins.fasta` protein database with the RepeatMasker TE proteins. Then, remove any proteins that match TE proteins. 

```bash
#create a blast database
~/tools/ncbi-blast-2.9.0+/bin/makeblastdb -dbtype prot -in RepeatPeps.lib -out RepeatPeps.lib
#get the taxID database
wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz
gunzip taxdb.tar.gz
tar -xvf taxdb.tar

#run blastp
~/tools/ncbi-blast-2.9.0+/bin/blastp -query RefAves_proteins.fasta -db RepeatPeps.lib -outfmt '6 qseqid staxids bitscore std sscinames sskingdoms stitle' -max_target_seqs 25 -culling_limit 2 -num_threads 23 -evalue 1e-5 -out RefAves_proteins.fasta.vs.RepeatPeps.25cul2.1e5.blastp.out

#remove proteins that match the TE proteins in BLAST
~/tools/assemblage/fastaqual_select.pl -f RefAves_proteins.fasta -e <(awk '{print $1}' RefAves_proteins.fasta.vs.RepeatPeps.25cul2.1e5.blastp.out | sort | uniq) > RefAves_proteins_no_tes.fa
 ```
 * -f: fasta file
 * -e: exclude file  
fastaqual_select comes from [this](https://blaxter-lab-documentation.readthedocs.io/en/latest/filter-repeatmodeler-library.html) pipeline.

The file now contains 60,780 bird proteins after removing TEs.
Now we have a database of bird proteins without TE proteins: `RefAves_proteins_no_tes.fa`.

# CDS Database
This will be a database used to filter proteins out of the TE library so that they do not get masked. It requires coding sequences (nucleotide) that do not have UTRs or introns, and should not contain TEs either. There are some tips on how to get CDS sequences on [Biostars](https://www.biostars.org/p/74870/). I will again choose my favourite model birds: *Ficedula, Taeniopygia, Manacus,* and *Gallus* to build the library. I chose these because I think they have been well annotated, and to represent oscines and suboscines.
```bash
cd /home/else/genetic_resources
esearch -db protein -query 'Ficedula albicollis [ORGN] NOT ("low quality" OR "Uncharacterized")' | efetch -format fasta_cds_na > FicAlb_CDS.lib
esearch -db protein -query 'Taeniopygia guttata [ORGN] NOT ("low quality" OR "Uncharacterized")' | efetch -format fasta_cds_na > TaeGut_CDS.lib
esearch -db protein -query 'Manacus vitellinus [ORGN] NOT ("low quality" OR "Uncharacterized")' | efetch -format fasta_cds_na > ManVit_CDS.lib
esearch -db protein -query 'Gallus gallus [ORGN] NOT ("low quality" OR "Uncharacterized")' | efetch -format fasta_cds_na > GalGal_CDS.lib

cat FicAlb_CDS.lib TaeGut_CDS.lib ManVit_CDS.lib GalGal_CDS.lib > ref_aves_CDS.lib

mv ./* /hhome/else/ #put onto main server, will move to /home/0_GENOMES1/0_Genetic_resources folder
```

