# Repeat curation.

Now we have a de novo library of repetitive elements. Next we need to curate these repeat sequences. This could be done mannually by hand and eye, but we received _**5000**_ repeat sequences from the EDTA pipeline. I do not have time to look at 5000 TE sequences. I need to reduce this number to a reasonable diversity without throwing too much away. I believe that redundancy is a major issue with this library so I will try to reduce it with CD-Hit. Then I will curate the remainder with an automated pipeline.

[CD-hit](http://weizhongli-lab.org/cd-hit/) is a fairly old program which can be used to cluster protein or nucleotide sequences by similarity, and then keep only a single (longest) example from each cluster of similar sequences. There is a guide [here](https://github.com/weizhongli/cdhit/wiki/3.-User's-Guide#CDHIT_tools).

```bash
#set up environment 
#cd-hit v4.8.1
export PATH=$PATH:/home/0_PROGRAMS/cdhit/

mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation

cp /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeatmodeler/Anteater_repeat_library.lib .

#First, count how many sequences we are dealing with 
grep -c "^>" Anteater_repeat_library.lib

time cd-hit-est -i Anteater_repeat_library.lib -o Anteater_repeat_library_95.lib -c 0.95 -n 10 -p 1 -d 0 -T 0 > Anteater_repeat_library_90.log

#count how many are left after clustering
grep -c "^>" Anteater_repeat_library_95.lib
#grep -c "^>" Anteater_repeat_library_90.lib

```
* `-i`: "input filename in fasta format, required"
* `-o`:	"output filename, required"
* `-c`:	"sequence identity threshold, default 0.9 this is the default cd-hit's "global sequence identity" calculated as: number of identical bases in alignment divided by the full length of the shorter sequence"
* `-T`:	"number of threads, default 1; with 0, all CPUs will be used"
* `-p`: "1 or 0, default 0; if set to 1, print alignment overlap in .clstr file"
* `-d`: "length of description in .clstr file, default 20; if set to 0, it takes the fasta defline and stops at first space"
* `-n`: word size

step|starting sequences|ending sequences|real time|user time|system time|threads|
---|---|---|---|---|---|---|
CD-hit 95%|5709|5646|6s|55s|0.3s|24|
80-80 rule BLAST|5646|3548|fast|-|1|
min 20 hits|3548|216!|fast|-|-|1|

Ok. That is still a lot of repeats.
Next we will do a BLASTn search to find the best matches for each repeat in the genome.
```bash
#set up environment
export PATH=$PATH:/home/0_PROGRAMS/ncbi-blast-2.9.0+/bin
export PATH=$PATH:/home/0_PROGRAMS/bioinfoTools
export PATH=$PATH:/home/0_PROGRAMS/ #for muscle executable

cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation

#get the genome sequence
cp ../Willisornis_poecilinotus_JTW1144_700mill.fasta.gz .
gunzip Willisornis_poecilinotus_JTW1144_700mill.fasta.gz
mv Willisornis_poecilinotus_JTW1144_700mill.fasta Anteater700.fasta

#BLAST your redundancy-reduced library against the genome sequence to get all hits
makeblastdb -in Anteater700.fasta -out Anteater700 -dbtype nucl
time blastn -db Anteater700 -query Anteater_repeat_library_95.lib -num_threads 22 -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen" -out Anteater_repeat_library_95_autocurate.tsv

#remove hits with less than 80% identity (column 3) to reduce false hits
awk '$3 > 80' Anteater_repeat_library_95_autocurate.tsv > temp
#remove hits with less than 80% length of consensus to reduce TEs matching a different TE species
awk '$4 > 0.8 * $13' temp > temp2
#count how many hits and unique TEs are retained
wc -l temp2 #108951
cat temp2 | cut -f 1 | sort | uniq | wc -l #3548

#filtering anything with less than 20 hits. This script was a gift from Uriel Garcilazo.
python filterreps.py #this requires that the script filterreps.py is present in the working directory and the file that you are filtering is named temp2. It produces the file Filtered_repeat_elements.txt and will overwrite it if it exists.

wc -l Filtered_repeat_elements.txt #104840 lines
cat Filtered_repeat_elements.txt | cut -f 1 | sort | uniq | wc -l #216!

#Now we have our candidate multi-copy TEs. We want to extract the flanking regions and align them.

#The following script will fail if there are "/" in the TE names! So we will remove them
grep "/" Filtered_repeat_elements.txt #just check that removing "/" wont be a problem
grep "/" Anteater_repeat_library_95.lib #just check that removing "/" wont be a problem
grep -c "@" Anteater_repeat_library_95.lib #check that there are no "@" symbols anywhere
grep -c "@" Filtered_repeat_elements.txt
#replace all "/" with "@"
sed 's/\//@/g' Anteater_repeat_library_95.lib > Anteater_repeat_library_95.lib.safe
sed 's/\//@/g' Filtered_repeat_elements.txt > Filtered_repeat_elements.txt.safe

#Load Perl modules
export PERLBREW_ROOT=/opt/perl5 #perlbrew will be installed in opt
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

perl /home/0_PROGRAMS/bioinfoTools/extractAlignTEs.pl --genome Anteater700.fasta --blast Filtered_repeat_elements.txt.safe --consTEs Anteater_repeat_library_95.lib.safe --seqBuffer 2000 --seqNum 104840

#Get rid of anything that has only 1 hit (the consensus sequence with no passing hits)
grep -c "^>" *.fas #lists how many sequences are in each alignment
grep -c "^>" *.fas | grep ".fas:1$" #lists all the sequences with no blast hits

grep -c "^>" *.fas | grep ".fas:1$" | sed 's/:1//g' | cat > seqs2filter.txt
mkdir garbage_can
cat seqs2filter.txt | while read line ; do mv "$line" garbage_can ; done
#now delete everything in the garbage (rm garbage_can/*)
```
* `--genome`: input genome file (in fasta format)           
* `--blast`: input blast file (tab format [6])              
* `--consTEs`: file of consensus elements                    
* `--seqBuffer`: additional 5'/3' bases in extracted sequence  
* `--seqNum`: number of sequences to extract. I am setting this very high in order to get all filtered hits.                 
* `--align` aligns the sequences with MUSCLE. I will do this outsid eof the script to do that step in parallel - much faster!

**Timing**: makeblastdb:15s; blastn: real	3m9.765s, user	5m41.139s, sys	0m4.777s

Ok. First, let's see how many sequences were found for each TE. There should be min 21 sequences per alignment (consensus query + min 20 blast hits)
```bash
grep -c "^>" ./*fas #lists how many sequences are in each alignment
```
Ok. Some have fewer than 50... but some have more than 5000! That is going to take a while to align. Better get started now.

```bash
find ./*.fas > TEs2align.txt #make a list of all TEs to align. Assuming there are no other .fas files in your working directory!
less TEs2align.txt #double check there is nothing unwanted in there
wc -l TEs2align.txt #yep still 216

cat TEs2align.txt | parallel time muscle -in {1} -out {1}.muscle
```
This will take quite a while to run...

Oh no, I had to log out of my account before it finished running. That's fine, I can start where I left off (65 out of 216 were done after about 16 hours).
```bash
find ./*.fas.muscle | sed 's/.muscle//g' > donealigning.txt #get a list of files that finished and produced the .muscle output
grep -xvf donealigning.txt TEs2align.txt > notdone.txt #remove the finished samples from the list of samples to get a list of unfinished samples
cat notdone.txt | parallel time muscle -in {1} -out {1}.muscle #run muscle on the unfinished samples.
```

Well, still some failed (the largest ones). 194 done and 22 left. I tried again with MUSCLE, however continued to get segmentation fault (core dumped), or out of memory warnings. I tried running only 8 at a time but still this problem persisted, so I am turning to MAFFT, which I did not use before because I had suspected that MUSCLE would be faster.
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation

export PATH=$PATH:/home/0_PROGRAMS/ #for muscle executable
export PATH=$PATH:/home/else/tools/mafft-7.453-with-extensions/bin #for mafft, on main server

find ./*.fas.muscle | sed 's/.muscle//g' > donealigning2.txt #get a list of files that finished and produced the .muscle output
grep -xvf donealigning2.txt TEs2align.txt > notdone2.txt #remove the finished samples from the list of samples to get a list of unfinished samples
cat notdone2.txt | parallel --jobs 8 time muscle -in {1} -out {1}.muscle #run muscle on the unfinished samples. #seg faults, failed

cat notdone2.txt | while read line ; do time mafft --auto --thread 24 "$line" > "$line".mafft ; done

```
* `--jobs`: limit the number of jobs that can be sent in parallel, for example to avoid running out of memory.

MAFFT was slow, taking real	211m41.921s, user	344m44.416s, sys	17m38.168s each to do one at a time, but at least they ran!

Here is an example of a few timings:  

#seqs to align|time done|
---|---|
35|17m16
96|23m43
105|24m48
80|38m44
68|45m48
82|26m37
99|52m42
71|54m43
69|20m14
90|44m20
125|64m31
62|77m30
76|56m59
161|107m42
160|124m22
55|44m50
173|129m51
80|101m5
141|179m48
133|182m6
39|18m9
140|199m44
42|27m11

etc! I won't write out all 216.

Anyways, it took a few days. The last to finish was a CR1 with 1547 sequences, which took 6437 minutes to align! In the end, 194 of 216 finished - 22 failed. I will deal with them later (actually I wrote that part above).

Next, we need to trim the alignments. This can be done by hand, but is tiresome and slow. Instead we can do it using [Trimal](http://trimal.cgenomics.org/downloads) and Gblocks.

```bash
#set up environment
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation
PATH=$PATH:/home/0_PROGRAMS/Gblocks_0.91b #for Gblocks
PATH=$PATH:/home/0_PROGRAMS/trimAl/source #for TrimAl

#make lists of all the alignments to work with
find ./*.fas.muscle > aligned_TEs.txt 
find ./*.fas.mafft > aligned_TEsmafft.txt 

#first, remove gap positions where less than 66% of sequences have data (33-100% gaps). It is pretty quick.
cat aligned_TEs.txt | while read line ; do trimal -in "$line" -out "$line".trim -htmlout "$line".html -fasta -gapthreshold 0.66 ; done
cat aligned_TEsmafft.txt | while read line ; do trimal -in "$line" -out "$line".trim -htmlout "$line".html -fasta -gapthreshold 0.66 ; done

#Then select the conserved blocks
cat aligned_TEs.txt | while read line ; do Gblocks "$line".trim -t=d -b1=$(($(grep -c '^>' "$line".trim)/2 +1)) -b2=$(echo "$(grep -c '^>' "$line".trim) * 0.8" | bc) -b3=6 -b4=60 -b5=h -s=y -e=gblok > "$line".trim.gbloklog ; done
cat aligned_TEsmafft.txt | while read line ; do Gblocks "$line".trim -t=d -b1=$(($(grep -c '^>' "$line".trim)/2 +1)) -b2=$(echo "$(grep -c '^>' "$line".trim) * 0.8" | bc) -b3=6 -b4=60 -b5=h -s=y -e=gblok > "$line".trim.gbloklog ; done

#Take a view at the output
cat *.gbloklog

#see how many are in a single block
cat *.gbloklog | grep -c "in 1 selected"

#see the alignments which are likely incomplete as they took up over 80% of the original alignment
cat *.gbloklog | grep -B 2 "8. %\|9. %" | grep -v "%\|--\|Original" > incomplete_rnd1

```
* `-t`: Type Of Sequence (Protein, DNA, Codons)
* `-b1=`: "Minimum Number Of Sequences For A Conserved Position (50% of the number of sequences + 1)" Leaving this as default rather than defining since I have different numbers in each alignment"
* `-b2=`: "Minimum Number Of Sequences For A Flank Position (85% of the number of sequences)" Leaving this as default rather than defining since I have different numbers in each alignment"
* `-b3=`	"Maximum Number Of Contiguous Nonconserved Positions (8)"
* `-b4=`	"Minimum Length Of A Block (10)"
* `-b5=`	"Allowed Gap Positions (None (n), With Half (h) , All (a))"
* `-s`: "Selected Blocks (Yes, No)"
* `-p`: "Results And Parameters File (Yes, Text, Short Text, No)"
* `-e`: "Generic File Extension (-gb)"
* `-fasta`: "print fasta format output"
* `-gapthreshold`: "1 - (fraction of sequences with a gap allowed)."
* `-simthreshold`: "Minimum average similarity allowed."
* `-cons`: "Minimum percentage of the positions in the original alignment to conserve."

I experimented with many settings. With -b3 6 -b4 30 I was still getting some spurious blocks appearing in the final alignment. When I increased it to -b4 60 all 191 were present in only a single block, and I did not seem to be losing the true alignment - I was satisfied. (Increasing `-b4` removes all the smaller, spurious blocks smaller than b4, but does not affect the larger blocks which should contain the true alignment, in this case I had no spurious blocks larger than 60 but some were larger than 30.

Now that we have a trimmed alignment, we will take the consensus of this alignment to create a consensus of the repeat.  
Note that I DO NOT like making consensuses of alignments because it DOES NOT reconstruct the "ancestral sequence" of the element, it just represents the most frequent bases.

I am using  Emboss [cons](http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/cons.html), but there is also Advanced Consensus maker from an HIV database, but it is online only.

```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation
PATH=$PATH:/home/0_PROGRAMS/EMBOSS-6.6.0/bin #for cons

#trim the file names first so that the fasta names are cleaner
cat aligned_TEs.txt | sed 's/.fas.muscle//g' | sed 's/.\///g' | parallel cons -sequence {1}.fas.muscle.trimgblok -outseq {1}.fas.muscle.trimgblok.consensus.fasta -name {1}
cat aligned_TEsmafft.txt | sed 's/.fas.mafft//g' | sed 's/.\///g' | parallel cons -sequence {1}.fas.mafft.trimgblok -outseq {1}.fas.mafft.trimgblok.consensus.fasta -name {1}

```
* `-sequence`: File containing a sequence alignment.
* `-outseq`: Sequence filename and optional format (output USA)
* `-name`: name of the consensus

It is rather quick for a single alignment but to do all of them took a few mins.

Now we have a curated TE library, with no manual editing.
```bash
cat aligned_TEs.txt aligned_TEsmafft.txt | while read line ; do cat "$line".trimgblok.consensus.fasta >> Anteater_curated_TE_library_consensus.fasta ; done

#quickly check if there is redundancy
export PATH=$PATH:/home/0_PROGRAMS/cdhit/

time cd-hit-est -i Anteater_curated_TE_library_consensus.fasta -o Anteater_curated_TE_library_consensus.95.fasta -c 0.95 -n 10 -p 1 -d 0 -T 4 > Anteater_curated_TE_library_consensus.95.log

#count how many are left after clustering
grep -c "^>" Anteater_curated_TE_library_consensus.fasta
grep -c "^>" Anteater_curated_TE_library_consensus.95.fasta

#get a list of what was removed
cat Anteater_curated_TE_library_consensus.fasta | grep ">" | grep -v "$(grep ">" Anteater_curated_TE_library_consensus.95.fasta)" | sed 's/>//g' > removed_redundant_curated_consensi.txt

grep -vf removed_redundant_curated_consensi.txt incomplete_rnd1 | sed 's/.fas.muscle.trim//g' | sed 's/.\///g' > incomplete_rnd1_tolengthen #was some overlap
```
In the end, 158 of 216 remain.

rescue incomplete stuff.
```bash
cat incomplete_rnd1_tolengthen | while read line ; do awk -v seq="$line" -v RS='>' '$1 == seq {print RS $0}' Anteater_curated_TE_library_consensus.fasta >> seqs2lengthen_rnd1.fasta ; done

time blastn -db Anteater700 -query seqs2lengthen_rnd1.fasta -num_threads 6 -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen" -out seqs2lengthen_rnd1.tsv #very fast, 8 seconds for 28 seqs
awk '$3 > 80' seqs2lengthen_rnd1.tsv | awk '$4 > 0.8 * $13' > temp2
wc -l temp2 #1528
cat temp2 | cut -f 1 | sort | uniq | wc -l #25
python filterreps.py #this requires that the script filterreps.py is present in the working directory and the file that you are filtering is named temp2. It produces the file Filtered_repeat_elements.txt and will overwrite it if it exists.
wc -l Filtered_repeat_elements.txt #1486 lines
cat Filtered_repeat_elements.txt | cut -f 1 | sort | uniq | wc -l #17

#extract seqs
export PERLBREW_ROOT=/opt/perl5 #perlbrew will be installed in opt
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH
perl /home/0_PROGRAMS/bioinfoTools/extractAlignTEs.pl --genome Anteater700.fasta --blast Filtered_repeat_elements.txt --consTEs seqs2lengthen_rnd1.fasta --seqBuffer 4000 --seqNum 104840 #warning overwrites original files

#Get rid of anything that had only 1 hit (the consensus sequence with no passing hits)
grep -c "^>" *.fas | grep ".fas:1$" | sed 's/:1//g' | cat > seqs2filter2.txt
mkdir garbage_can
cat seqs2filter2.txt | while read line ; do mv "$line" garbage_can ; done

#align
export PATH=$PATH:/home/else/tools/ #for muscle executable onmain server
cat incomplete_rnd1_tolengthen | parallel cp {1}.fas ~/Downloads
cat incomplete_rnd1_tolengthen | parallel cp ~/Downloads/{1}.fas .

cat incomplete_rnd1_tolengthen | parallel time muscle -in {1}.fas -out {1}.fas.round2.muscle #longest took 130 min
cat incomplete_rnd1_tolengthen | parallel cp {1}.fas.round2.muscle ~/Downloads
cat incomplete_rnd1_tolengthen | parallel cp ~/Downloads/{1}.fas.round2.muscle .

#Then select the conserved blocks
find ./*.fas.round2.muscle > aligned_TEs2.txt 
PATH=$PATH:/home/0_PROGRAMS/Gblocks_0.91b #for Gblocks
PATH=$PATH:/home/0_PROGRAMS/trimAl/source #for TrimAl
cat aligned_TEs2.txt | while read line ; do trimal -in "$line" -out "$line".trim -htmlout "$line".html -fasta -gapthreshold 0.66 ; done
cat aligned_TEs2.txt | while read line ; do Gblocks "$line".trim -t=d -b1=$(($(grep -c '^>' "$line".trim)/2 +1)) -b2=$(echo "$(grep -c '^>' "$line".trim) * 0.8" | bc) -b3=6 -b4=60 -b5=h -s=y -e=gblok > "$line".trim.gbloklog ; done

#Take a view at the output
cat *round2*.gbloklog

#see how many are in a single block
cat *.gbloklog | grep -c "in 1 selected"

#see the alignments which are likely incomplete as they took up over 80% of the original alignment
cat *.gbloklog | grep -B 2 "8. %\|9. %" | grep -v "%\|--\|Original" > incomplete_rnd1

PATH=$PATH:/home/0_PROGRAMS/EMBOSS-6.6.0/bin #for cons
#trim the file names first so that the fasta names are cleaner
cat aligned_TEs2.txt | sed 's/.fas.round2.muscle//g' | sed 's/.\///g' | parallel cons -sequence {1}.fas.round2.muscle.trimgblok -outseq {1}.fas.muscle.trimgblok.consensus.fasta -name {1}

#update the library
cat aligned_TEs.txt aligned_TEsmafft.txt | while read line ; do cat "$line".trimgblok.consensus.fasta >> Anteater_curated_TE_library_consensus.fasta ; done

#quickly check if there is redundancy
export PATH=$PATH:/home/0_PROGRAMS/cdhit/

time cd-hit-est -i Anteater_curated_TE_library_consensus.fasta -o Anteater_curated_TE_library_consensus.95.fasta -c 0.95 -n 10 -p 1 -d 0 -T 4 > Anteater_curated_TE_library_consensus.95.log

#count how many are left after clustering
grep -c "^>" Anteater_curated_TE_library_consensus.fasta
grep -c "^>" Anteater_curated_TE_library_consensus.95.fasta

#get a list of what was removed
cat Anteater_curated_TE_library_consensus.fasta | grep ">" | grep -v "$(grep ">" Anteater_curated_TE_library_consensus.95.fasta)" | sed 's/>//g' > removed_redundant_curated_consensi.txt

grep -vf removed_redundant_curated_consensi.txt incomplete_rnd1 | sed 's/.fas.muscle.trim//g' | sed 's/.\///g' > incomplete_rnd1_tolengthen #was some overlap
```
157 out of 216, final library.

# Classify curated repeats

Now, let's clean up the classification a bit with TE_sorter now that we have full(er) length repeats!
```bash
#set up environment
export PATH=$PATH:/home/0_PROGRAMS/hmmer-3.3/bin
export PATH=$PATH:/home/0_PROGRAMS/ncbi-blast-2.9.0+/bin

mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation/TEsorter
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation/TEsorter

conda activate python2 #make sure python 2.7, pp and biopython are installed!

#Run TEsorter
python /home/0_PROGRAMS/TEsorter/TEsorter.py ../Anteater_curated_TE_library_consensus.95.fasta -p 20

cat /home/0_GENOMES1/0_Genetic_resources/Repeatomes/68015-fAlb15_rm3.0.lib /home/0_GENOMES1/0_Genetic_resources/Repeatomes/uraCya_rm2.45.fasta /home/0_GENOMES1/0_Genetic_resources/Repeatomes/AviRTE_TEs.fa > FicalbUracya_replibs.fasta

makeblastdb -in FicalbUracya_replibs.fasta -out FicalbUracya_replibs -dbtype nucl
time blastn -db FicalbUracya_replibs -query ../Anteater_curated_TE_library_consensus.95.fasta -num_threads 22 -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen" -out Anteater_curated_TE_library_consensus.95.fasta.tsv

#remove hits with less than 80% identity (column 3) to reduce false hits
#remove hits with less than 80% length of consensus to reduce TEs matching a different TE species
| awk '$4 > 0.8 * $13'
awk '$3 > 80' Anteater_curated_TE_library_consensus.95.fasta.tsv > Anteater_curated_TE_library_consensus.95.fasta.8080.tsv 
#count how many hits and unique TEs are retained
wc -l Anteater_curated_TE_library_consensus.95.fasta.8080.tsv #26
cat Anteater_curated_TE_library_consensus.95.fasta.tsv | cut -f 1 | sort | uniq | wc -l #10

```
* `-p`: number of threads to use

**Results**:TEsorter results are in the Repeat_Library folder in this github repo.

# Further Identification
Next, upload your library to GIRI to use [Censor](https://www.girinst.org/censor/). I am not using the command line version because the database is too expensive for my purposes to download, but I am thankful to be able to search it on their website.

Then, submit it to [TEclass](http://www.bioinformatics.uni-muenster.de/tools/teclass/generate/index.pl?lang=en) (give a name for the title, but not ID. The queue goes pretty quick.)

**Results**:Censor and TE_class results are in the Repeat_Library folder in this github repo.

# Change names
Now I am going to rename the TEs to be more meaningful, based on the results of TEClass, Censor, and TEsorter.

Here is the list of TEs before renaming:
```bash
cat > curated_TE_list_names
sed -i 's/>//g' curated_TE_list_names
```
```
>rnd-1_family-100#Unknown
>rnd-1_family-101#Unknown
>rnd-1_family-102#Unknown
>rnd-1_family-103#LTR@ERVK
>rnd-1_family-104#Unknown
>rnd-1_family-105#Unknown
>rnd-1_family-107#LTR@ERVL
>rnd-1_family-113#LINE@CR1
>rnd-1_family-114#Unknown
>rnd-1_family-117#Unknown
>rnd-1_family-119#Unknown
>rnd-1_family-122#LTR@ERVL
>rnd-1_family-126#Unknown
>rnd-1_family-128#Unknown
>rnd-1_family-129#Unknown
>rnd-1_family-130#Unknown
>rnd-1_family-134#Unknown
>rnd-1_family-138#Unknown
>rnd-1_family-143#Unknown
>rnd-1_family-145#Unknown
>rnd-1_family-150#Unknown
>rnd-1_family-152#Unknown
>rnd-1_family-16#LINE@CR1
>rnd-1_family-18#LINE@CR1
>rnd-1_family-21#LINE@CR1
>rnd-1_family-26#LINE@CR1
>rnd-1_family-27#LINE@RTE-BovB
>rnd-1_family-29#Unknown
>rnd-1_family-30#Unknown
>rnd-1_family-31#LINE@CR1
>rnd-1_family-43#Unknown
>rnd-1_family-44#Unknown
>rnd-1_family-45#Unknown
>rnd-1_family-47#LTR@ERVL
>rnd-1_family-51#Unknown
>rnd-1_family-52#LTR@ERVL
>rnd-1_family-53#Unknown
>rnd-1_family-54#LTR@ERVL
>rnd-1_family-57#LTR@ERVL
>rnd-1_family-59#Unknown
>rnd-1_family-63#Unknown
>rnd-1_family-64#Unknown
>rnd-1_family-68#Unknown
>rnd-1_family-79#LTR@ERVL
>rnd-1_family-81#Unknown
>rnd-1_family-84#Unknown
>rnd-1_family-88#Unknown
>rnd-1_family-97#Unknown
>rnd-1_family-98#Unknown
>rnd-1_family-99#Unknown
>rnd-3_family-53#LINE@CR1
>rnd-4_family-1062#Unknown
>rnd-4_family-336#Unknown
>rnd-5_family-10335#Unknown
>rnd-5_family-10372#Satellite
>rnd-5_family-10982#Unknown
>rnd-5_family-1147#LTR@ERV1
>rnd-5_family-11568#Unknown
>rnd-5_family-12159#LINE@CR1
>rnd-5_family-1373#LTR@ERVK
>rnd-5_family-14611#Unknown
>rnd-5_family-1533#Unknown
>rnd-5_family-15382#Unknown
>rnd-5_family-1599#Unknown
>rnd-5_family-16196#Unknown
>rnd-5_family-16640#Unknown
>rnd-5_family-1684#Unknown
>rnd-5_family-1704#LTR@ERVL
>rnd-5_family-1706#Unknown
>rnd-5_family-17134#Unknown
>rnd-5_family-18364#LTR@ERVK
>rnd-5_family-18768#Unknown
>rnd-5_family-2249#Unknown
>rnd-5_family-2267#LTR@ERV1
>rnd-5_family-25903#LTR@ERVK
>rnd-5_family-26406#Unknown
>rnd-5_family-3138#Unknown
>rnd-5_family-3181#Unknown
>rnd-5_family-3861#Unknown
>rnd-5_family-4174#Unknown
>rnd-5_family-42824#Unknown
>rnd-5_family-439#LTR@ERVK
>rnd-5_family-4633#LINE@CR1
>rnd-5_family-5294#LTR@ERVK
>rnd-5_family-6012#Unknown
>rnd-5_family-6177#Unknown
>rnd-5_family-6256#Unknown
>rnd-5_family-6290#Unknown
>rnd-5_family-6381#Unknown
>rnd-5_family-6742#Unknown
>rnd-5_family-6995#Unknown
>rnd-5_family-7082#LTR@ERVK
>rnd-5_family-7126#Unknown
>rnd-5_family-764#Unknown
>rnd-5_family-796#Unknown
>rnd-5_family-9233#tRNA
>rnd-5_family-963#LTR@ERV1
>rnd-5_family-965#Unknown
>rnd-5_family-992#Unknown
>TE_00000000#MITE@unknown
>TE_00000001#MITE@unknown
>TE_00000002#MITE@unknown
>TE_00000010#MITE@unknown
>TE_00000012#MITE@unknown
>TE_00000015#MITE@unknown
>TE_00000018#MITE@unknown
>TE_00000021#MITE@unknown
>TE_00000022#MITE@unknown
>TE_00000023#MITE@unknown
>TE_00000024#MITE@unknown
>TE_00000028#MITE@unknown
>TE_00000030#MITE@unknown
>TE_00000032#MITE@unknown
>TE_00000033#MITE@unknown
>TE_00000034#MITE@unknown
>TE_00000036#MITE@unknown
>TE_00000040#MITE@unknown
>TE_00000043#MITE@unknown
>TE_00000046#MITE@unknown
>TE_00000049#MITE@unknown
>TE_00000052#MITE@unknown
>TE_00000068#MITE@unknown
>TE_00000071#MITE@unknown
>TE_00000073#MITE@unknown
>TE_00001107_LTR#LTR@unknown
>TE_00001177_LTR#LTR@unknown
>TE_00001214_LTR#LTR@Gypsy
>TE_00001324_LTR#LTR@unknown
>TE_00002249_INT#LTR@Gypsy
>TE_00002464_LTR#LTR@unknown
>TE_00002563_LTR#LTR@unknown
>TE_00002676_LTR#LTR@unknown
>TE_00002739_LTR#LTR@Copia
>TE_00002799_LTR#LTR@unknown
>TE_00002809_LTR#LTR@Gypsy
>TE_00002832_LTR#LTR@unknown
>TE_00002986_LTR#LTR@unknown
>TE_00003054_INT#LTR@unknown
>TE_00003386_LTR#LTR@Gypsy
>TE_00003421_LTR#LTR@Gypsy
>TE_00003595_INT#LTR@Gypsy
>TE_00003627_LTR#LTR@unknown
>TE_00003665_LTR#LTR@Gypsy
>TE_00003759_LTR#LTR@unknown
>TE_00003992_INT#LTR@unknown
>TE_00004015_LTR#LTR@Gypsy
>TE_00004237_LTR#LTR@unknown
>TE_00004268_LTR#LTR@unknown
>TE_00004276_LTR#LTR@unknown
>TE_00004455_LTR#LTR@unknown
>TE_00004571_LTR#LTR@unknown
>TE_00004612_LTR#LTR@unknown
>TE_00004824_LTR#LTR@unknown
>TE_00004863_INT#LTR@unknown
>TE_00004880_LTR#LTR@unknown
>TE_00004920_LTR#LTR@unknown
>TE_00004995_LTR#LTR@Copia
>TE_00005070_LTR#LTR@unknown
```

Here is what I will rename them:
```bash
cat > new_TE_classification
```
```
Wilpoe000#LTR/Unknown
Wilpoe001#LTR/Unknown
Wilpoe002#Unknown
Wilpoe003#LTR/ERVK
Wilpoe004#LTR/Unknown
Wilpoe005#LTR/Unknown
Wilpoe006#LTR/ERVL
Wilpoe007#LINE/CR1
Wilpoe008#LTR/Unknown
Wilpoe009#LTR/Unknown
Wilpoe010#LTR/Unknown
Wilpoe011#LTR/ERVL
Wilpoe012#LTR/ERVL
Wilpoe013#LTR/Unknown
Wilpoe014#LTR/Unknown
Wilpoe015#LTR/Unknown
Wilpoe016#LINE/Unknown
Wilpoe017#LTR/Unknown
Wilpoe018#LTR/Unknown
Wilpoe019#LTR/Unknown
Wilpoe020#LTR/Unknown
Wilpoe021#LTR/Unknown
Wilpoe022#LINE/CR1
Wilpoe023#LINE/CR1
Wilpoe024#LINE/CR1
Wilpoe025#LINE/CR1
Wilpoe026#LINE/RTE
Wilpoe027#LTR/Unknown
Wilpoe028#LTR/Unknown
Wilpoe029#LINE/CR1
Wilpoe030#LTR/Unknown
Wilpoe031#LTR/ERVL
Wilpoe032#LTR/ERVL
Wilpoe033#LTR/ERVL
Wilpoe034#LTR/ERVL
Wilpoe035#LTR/ERVL
Wilpoe036#LTR/Unknown
Wilpoe037#LTR/ERVL
Wilpoe038#LTR/ERVL
Wilpoe039#LTR/Unknown
Wilpoe040#LTR/Unknown
Wilpoe041#LTR/Unknown
Wilpoe042#LTR/Unknown
Wilpoe043#LTR/ERVL
Wilpoe044#LINE/Unknown
Wilpoe045#Unknown
Wilpoe046#Unknown
Wilpoe047#Unknown
Wilpoe048#LTR/Unknown
Wilpoe049#Unknown
Wilpoe050#LINE/CR1
Wilpoe051#LTR/Unknown
Wilpoe052#LTR/Unknown
Wilpoe053#LTR/Unknown
Wilpoe054#Satellite
Wilpoe055#Unknown
Wilpoe056#LTR/ERV1
Wilpoe057#Unknown
Wilpoe058#LINE/CR1
Wilpoe059#LTR/ERVK
Wilpoe060#LTR/ERVL
Wilpoe061#LTR/Unknown
Wilpoe062#LTR/Unknown
Wilpoe063#LTR/Unknown
Wilpoe064#LTR/Unknown
Wilpoe065#LTR/Unknown
Wilpoe066#LTR/Unknown
Wilpoe067#LTR/ERVL
Wilpoe068#LTR/ERVL
Wilpoe069#LTR/Unknown
Wilpoe070#LTR/Unknown
Wilpoe071#LTR/ERVL
Wilpoe072#LTR/Unknown
Wilpoe073#LTR/ERV1
Wilpoe074#LTR/ERVK
Wilpoe075#LTR/Unknown
Wilpoe076#Unknown
Wilpoe077#LTR/Unknown
Wilpoe078#LTR/Unknown
Wilpoe079#LINE/CR1
Wilpoe080#LTR/Unknown
Wilpoe081#LTR/ERVK
Wilpoe082#LINE/CR1
Wilpoe083#LTR/ERV2
Wilpoe084#Unknown
Wilpoe085#LTR/Unknown
Wilpoe086#Unknown
Wilpoe087#Unknown
Wilpoe088#LTR/ERVL
Wilpoe089#LTR/Unknown
Wilpoe090#Unknown
Wilpoe091#LTR/ERVK
Wilpoe092#Unknown
Wilpoe093#LTR/Unknown
Wilpoe094#LTR/Unknown
Wilpoe095#SINE/RTE
Wilpoe096#LTR/ERV1
Wilpoe097#LTR/Unknown
Wilpoe098#LTR/Unknown
Wilpoe099#Unknown
Wilpoe100#LTR/Unknown
Wilpoe101#SINE/SINE2
Wilpoe102#SINE/AviRTE
Wilpoe103#LTR/ERVL
Wilpoe104#LTR/Unknown
Wilpoe105#LTR/Unknown
Wilpoe106#LTR/Unknown
Wilpoe107#LTR/Unknown
Wilpoe108#LTR/Unknown
Wilpoe109#LTR/Unknown
Wilpoe110#LTR/ERVL
Wilpoe111#LTR/Unknown
Wilpoe112#LTR/Unknown
Wilpoe113#LTR/Unknown
Wilpoe114#LTR/Unknown
Wilpoe115#LTR/Unknown
Wilpoe116#LTR/Unknown
Wilpoe117#LTR/Unknown
Wilpoe118#LTR/Unknown
Wilpoe119#Unknown
Wilpoe120#LTR/Unknown
Wilpoe121#LINE/CR1
Wilpoe122#LINE/CR1
Wilpoe123#LTR/Unknown
Wilpoe124#LTR/Unknown
Wilpoe125#LTR/Unknown
Wilpoe126#LTR/Unknown
Wilpoe127#LTR/Unknown
Wilpoe128#LTR/ERVK
Wilpoe129#LINE/CR1
Wilpoe130#LTR/Unknown
Wilpoe131#LINE/CR1
Wilpoe132#LTR/Unknown
Wilpoe133#LTR/Unknown
Wilpoe134#LTR/Unknown
Wilpoe135#LINE/CR1
Wilpoe136#LINE/CR1
Wilpoe137#LTR/ERVL
Wilpoe138#LTR/Unknown
Wilpoe139#LTR/Unknown
Wilpoe140#LTR/ERVK
Wilpoe141#LTR/Unknown
Wilpoe142#LTR/Unknown
Wilpoe143#LTR/Unknown
Wilpoe144#LTR/ERVL
Wilpoe145#LTR/Unknown
Wilpoe146#LTR/Unknown
Wilpoe147#LTR/Unknown
Wilpoe148#LTR/Unknown
Wilpoe149#LINE/CR1
Wilpoe150#LTR/Unknown
Wilpoe151#LTR/Unknown
Wilpoe152#LTR/Unknown
Wilpoe153#Unknown
Wilpoe154#Unknown
Wilpoe155#LTR/Unknown
Wilpoe156#LTR/Unknown
Wilpoe157#LINE/CR1
```
Here is a naming table matching old and new names
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation
cat > Wilpoe_Te_naming_table.txt
```
```
rnd-1_family-100#Unknown	Wilpoe000#LTR/Unknown
rnd-1_family-101#Unknown	Wilpoe001#LTR/Unknown
rnd-1_family-102#Unknown	Wilpoe002#Unknown
rnd-1_family-103#LTR@ERVK	Wilpoe003#LTR/ERVK
rnd-1_family-104#Unknown	Wilpoe004#LTR/Unknown
rnd-1_family-105#Unknown	Wilpoe005#LTR/Unknown
rnd-1_family-107#LTR@ERVL	Wilpoe006#LTR/ERVL
rnd-1_family-113#LINE@CR1	Wilpoe007#LINE/CR1
rnd-1_family-114#Unknown	Wilpoe008#LTR/Unknown
rnd-1_family-117#Unknown	Wilpoe009#LTR/Unknown
rnd-1_family-119#Unknown	Wilpoe010#LTR/Unknown
rnd-1_family-122#LTR@ERVL	Wilpoe011#LTR/ERVL
rnd-1_family-126#Unknown	Wilpoe012#LTR/ERVL
rnd-1_family-128#Unknown	Wilpoe013#LTR/Unknown
rnd-1_family-129#Unknown	Wilpoe014#LTR/Unknown
rnd-1_family-130#Unknown	Wilpoe015#LTR/Unknown
rnd-1_family-134#Unknown	Wilpoe016#LINE/Unknown
rnd-1_family-138#Unknown	Wilpoe017#LTR/Unknown
rnd-1_family-143#Unknown	Wilpoe018#LTR/Unknown
rnd-1_family-145#Unknown	Wilpoe019#LTR/Unknown
rnd-1_family-150#Unknown	Wilpoe020#LTR/Unknown
rnd-1_family-152#Unknown	Wilpoe021#LTR/Unknown
rnd-1_family-16#LINE@CR1	Wilpoe022#LINE/CR1
rnd-1_family-18#LINE@CR1	Wilpoe023#LINE/CR1
rnd-1_family-21#LINE@CR1	Wilpoe024#LINE/CR1
rnd-1_family-26#LINE@CR1	Wilpoe025#LINE/CR1
rnd-1_family-27#LINE@RTE-BovB	Wilpoe026#LINE/RTE
rnd-1_family-29#Unknown	Wilpoe027#LTR/Unknown
rnd-1_family-30#Unknown	Wilpoe028#LTR/Unknown
rnd-1_family-31#LINE@CR1	Wilpoe029#LINE/CR1
rnd-1_family-43#Unknown	Wilpoe030#LTR/Unknown
rnd-1_family-44#Unknown	Wilpoe031#LTR/ERVL
rnd-1_family-45#Unknown	Wilpoe032#LTR/ERVL
rnd-1_family-47#LTR@ERVL	Wilpoe033#LTR/ERVL
rnd-1_family-51#Unknown	Wilpoe034#LTR/ERVL
rnd-1_family-52#LTR@ERVL	Wilpoe035#LTR/ERVL
rnd-1_family-53#Unknown	Wilpoe036#LTR/Unknown
rnd-1_family-54#LTR@ERVL	Wilpoe037#LTR/ERVL
rnd-1_family-57#LTR@ERVL	Wilpoe038#LTR/ERVL
rnd-1_family-59#Unknown	Wilpoe039#LTR/Unknown
rnd-1_family-63#Unknown	Wilpoe040#LTR/Unknown
rnd-1_family-64#Unknown	Wilpoe041#LTR/Unknown
rnd-1_family-68#Unknown	Wilpoe042#LTR/Unknown
rnd-1_family-79#LTR@ERVL	Wilpoe043#LTR/ERVL
rnd-1_family-81#Unknown	Wilpoe044#LINE/Unknown
rnd-1_family-84#Unknown	Wilpoe045#Unknown
rnd-1_family-88#Unknown	Wilpoe046#Unknown
rnd-1_family-97#Unknown	Wilpoe047#Unknown
rnd-1_family-98#Unknown	Wilpoe048#LTR/Unknown
rnd-1_family-99#Unknown	Wilpoe049#Unknown
rnd-3_family-53#LINE@CR1	Wilpoe050#LINE/CR1
rnd-4_family-1062#Unknown	Wilpoe051#LTR/Unknown
rnd-4_family-336#Unknown	Wilpoe052#LTR/Unknown
rnd-5_family-10335#Unknown	Wilpoe053#LTR/Unknown
rnd-5_family-10372#Satellite	Wilpoe054#Satellite
rnd-5_family-10982#Unknown	Wilpoe055#Unknown
rnd-5_family-1147#LTR@ERV1	Wilpoe056#LTR/ERV1
rnd-5_family-11568#Unknown	Wilpoe057#Unknown
rnd-5_family-12159#LINE@CR1	Wilpoe058#LINE/CR1
rnd-5_family-1373#LTR@ERVK	Wilpoe059#LTR/ERVK
rnd-5_family-14611#Unknown	Wilpoe060#LTR/ERVL
rnd-5_family-1533#Unknown	Wilpoe061#LTR/Unknown
rnd-5_family-15382#Unknown	Wilpoe062#LTR/Unknown
rnd-5_family-1599#Unknown	Wilpoe063#LTR/Unknown
rnd-5_family-16196#Unknown	Wilpoe064#LTR/Unknown
rnd-5_family-16640#Unknown	Wilpoe065#LTR/Unknown
rnd-5_family-1684#Unknown	Wilpoe066#LTR/Unknown
rnd-5_family-1704#LTR@ERVL	Wilpoe067#LTR/ERVL
rnd-5_family-1706#Unknown	Wilpoe068#LTR/ERVL
rnd-5_family-17134#Unknown	Wilpoe069#LTR/Unknown
rnd-5_family-18364#LTR@ERVK	Wilpoe070#LTR/Unknown
rnd-5_family-18768#Unknown	Wilpoe071#LTR/ERVL
rnd-5_family-2249#Unknown	Wilpoe072#LTR/Unknown
rnd-5_family-2267#LTR@ERV1	Wilpoe073#LTR/ERV1
rnd-5_family-25903#LTR@ERVK	Wilpoe074#LTR/ERVK
rnd-5_family-26406#Unknown	Wilpoe075#LTR/Unknown
rnd-5_family-3138#Unknown	Wilpoe076#Unknown
rnd-5_family-3181#Unknown	Wilpoe077#LTR/Unknown
rnd-5_family-3861#Unknown	Wilpoe078#LTR/Unknown
rnd-5_family-4174#Unknown	Wilpoe079#LINE/CR1
rnd-5_family-42824#Unknown	Wilpoe080#LTR/Unknown
rnd-5_family-439#LTR@ERVK	Wilpoe081#LTR/ERVK
rnd-5_family-4633#LINE@CR1	Wilpoe082#LINE/CR1
rnd-5_family-5294#LTR@ERVK	Wilpoe083#LTR/ERV2
rnd-5_family-6012#Unknown	Wilpoe084#Unknown
rnd-5_family-6177#Unknown	Wilpoe085#LTR/Unknown
rnd-5_family-6256#Unknown	Wilpoe086#Unknown
rnd-5_family-6290#Unknown	Wilpoe087#Unknown
rnd-5_family-6381#Unknown	Wilpoe088#LTR/ERVL
rnd-5_family-6742#Unknown	Wilpoe089#LTR/Unknown
rnd-5_family-6995#Unknown	Wilpoe090#Unknown
rnd-5_family-7082#LTR@ERVK	Wilpoe091#LTR/ERVK
rnd-5_family-7126#Unknown	Wilpoe092#Unknown
rnd-5_family-764#Unknown	Wilpoe093#LTR/Unknown
rnd-5_family-796#Unknown	Wilpoe094#LTR/Unknown
rnd-5_family-9233#tRNA	Wilpoe095#SINE/RTE
rnd-5_family-963#LTR@ERV1	Wilpoe096#LTR/ERV1
rnd-5_family-965#Unknown	Wilpoe097#LTR/Unknown
rnd-5_family-992#Unknown	Wilpoe098#LTR/Unknown
TE_00000000#MITE@unknown	Wilpoe099#Unknown
TE_00000001#MITE@unknown	Wilpoe100#LTR/Unknown
TE_00000002#MITE@unknown	Wilpoe101#SINE/SINE2
TE_00000010#MITE@unknown	Wilpoe102#SINE/AviRTE
TE_00000012#MITE@unknown	Wilpoe103#LTR/ERVL
TE_00000015#MITE@unknown	Wilpoe104#LTR/Unknown
TE_00000018#MITE@unknown	Wilpoe105#LTR/Unknown
TE_00000021#MITE@unknown	Wilpoe106#LTR/Unknown
TE_00000022#MITE@unknown	Wilpoe107#LTR/Unknown
TE_00000023#MITE@unknown	Wilpoe108#LTR/Unknown
TE_00000024#MITE@unknown	Wilpoe109#LTR/Unknown
TE_00000028#MITE@unknown	Wilpoe110#LTR/ERVL
TE_00000030#MITE@unknown	Wilpoe111#LTR/Unknown
TE_00000032#MITE@unknown	Wilpoe112#LTR/Unknown
TE_00000033#MITE@unknown	Wilpoe113#LTR/Unknown
TE_00000034#MITE@unknown	Wilpoe114#LTR/Unknown
TE_00000036#MITE@unknown	Wilpoe115#LTR/Unknown
TE_00000040#MITE@unknown	Wilpoe116#LTR/Unknown
TE_00000043#MITE@unknown	Wilpoe117#LTR/Unknown
TE_00000046#MITE@unknown	Wilpoe118#LTR/Unknown
TE_00000049#MITE@unknown	Wilpoe119#Unknown
TE_00000052#MITE@unknown	Wilpoe120#LTR/Unknown
TE_00000068#MITE@unknown	Wilpoe121#LINE/CR1
TE_00000071#MITE@unknown	Wilpoe122#LINE/CR1
TE_00000073#MITE@unknown	Wilpoe123#LTR/Unknown
TE_00001107_LTR#LTR@unknown	Wilpoe124#LTR/Unknown
TE_00001177_LTR#LTR@unknown	Wilpoe125#LTR/Unknown
TE_00001214_LTR#LTR@Gypsy	Wilpoe126#LTR/Unknown
TE_00001324_LTR#LTR@unknown	Wilpoe127#LTR/Unknown
TE_00002249_INT#LTR@Gypsy	Wilpoe128#LTR/ERVK
TE_00002464_LTR#LTR@unknown	Wilpoe129#LINE/CR1
TE_00002563_LTR#LTR@unknown	Wilpoe130#LTR/Unknown
TE_00002676_LTR#LTR@unknown	Wilpoe131#LINE/CR1
TE_00002739_LTR#LTR@Copia	Wilpoe132#LTR/Unknown
TE_00002799_LTR#LTR@unknown	Wilpoe133#LTR/Unknown
TE_00002809_LTR#LTR@Gypsy	Wilpoe134#LTR/Unknown
TE_00002832_LTR#LTR@unknown	Wilpoe135#LINE/CR1
TE_00002986_LTR#LTR@unknown	Wilpoe136#LINE/CR1
TE_00003054_INT#LTR@unknown	Wilpoe137#LTR/ERVL
TE_00003386_LTR#LTR@Gypsy	Wilpoe138#LTR/Unknown
TE_00003421_LTR#LTR@Gypsy	Wilpoe139#LTR/Unknown
TE_00003595_INT#LTR@Gypsy	Wilpoe140#LTR/ERVK
TE_00003627_LTR#LTR@unknown	Wilpoe141#LTR/Unknown
TE_00003665_LTR#LTR@Gypsy	Wilpoe142#LTR/Unknown
TE_00003759_LTR#LTR@unknown	Wilpoe143#LTR/Unknown
TE_00003992_INT#LTR@unknown	Wilpoe144#LTR/ERVL
TE_00004015_LTR#LTR@Gypsy	Wilpoe145#LTR/Unknown
TE_00004237_LTR#LTR@unknown	Wilpoe146#LTR/Unknown
TE_00004268_LTR#LTR@unknown	Wilpoe147#LTR/Unknown
TE_00004276_LTR#LTR@unknown	Wilpoe148#LTR/Unknown
TE_00004455_LTR#LTR@unknown	Wilpoe149#LINE/CR1
TE_00004571_LTR#LTR@unknown	Wilpoe150#LTR/Unknown
TE_00004612_LTR#LTR@unknown	Wilpoe151#LTR/Unknown
TE_00004824_LTR#LTR@unknown	Wilpoe152#LTR/Unknown
TE_00004863_INT#LTR@unknown	Wilpoe153#Unknown
TE_00004880_LTR#LTR@unknown	Wilpoe154#Unknown
TE_00004920_LTR#LTR@unknown	Wilpoe155#LTR/Unknown
TE_00004995_LTR#LTR@Copia	Wilpoe156#LTR/Unknown
TE_00005070_LTR#LTR@unknown	Wilpoe157#LINE/CR1
```
Now, use a python script to rename the fasta headers. This script is below in the appendix.
```bash
python /home/0_PROGRAMS/uriel/renamefasta.py
```
# Curated Library

Our final library is now in `/home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/repeat_curation/V2_Anteater_curated_TE_library_consensus.95.fasta` - lets name it something simpler

```bash
mv V2_Anteater_curated_TE_library_consensus.95.fasta Wilpoe_TE_library.fasta
```

# Our curated library is now in `Wilpoe_TE_library.fasta`. I have also placed it into the Repeat_library folder of this repository.


# Notes on TE curation by hand 

First, you want to make nicely curated consensus sequences. The steps for that could be:  
(These are used for example in Abundant recent activity of retrovirus-like retrotransposons within and among flycatcher species implies a rich source of structural variation in songbird genomes)
1) BLASTn your repeats to the genome and extract the 20 best hits for each sequence with 2000 bp flanking regions.

2) Align these best hits using MAFFT

3) If the repetitive part of the alignment ends (on the 5' end or 3' end), then you might be missing some more of the repetitive parts. Re extract the flanking regions with wider flanks (4000 bp maybe) and realign. Repeat until you are running unto unique DNA that does not seem to align between different TE copies - there are the edges to the TEs.

3) make a majority rule consensus sequence. Could do by hand, or use Advanced Consensus Maker

4) Try to classify unknowns
* Look for 5'-TG-CA-3'termini of LTRs
* count the length of TSDs for subfamilies of LTRs (RV1=4 bp; ERV2/ERVK = 6 bp; ERVs/ERVL = 5 bp)
* "characteristic eight base pair motif [42], 5?-ATTCTRTG-3?, was used to identify the 3? ends of CR1 LINEs"

5) Compare with previous curated TE libraries using CENSOR.   
This is really pretty fun. Submit your TE sequences to GIRI and see what they are!!

6) Name the repeats. follow nomenclature guidelines for your group
* Use 80-80-80 rule to put repeats into families
* Sequence similarity across entire length = name + wPoe
* Sequence similarity across part of length = name + -L_wPoe
* No sequence similarity - new family? wPoe + superfamily name
* 95% similarity defines subfamilies

7) Classify CR1 subfamilies by aligning to known CR1 repeats (get from Data S4 of "Abundant recent activity of retrovirus-like retrotransposons within and among flycatcher species implies a rich source of structural variation in songbird genomes")
* Align with MAFFT, tree with RAxML or PhyML

8) Merge your curated library with other curated libraries

9) Run RepeatMasker parameters -a -xsmall - gccalc

10) Look at TE landscapes

Have fun!

# Appendix A: renamefasta.py
This script was written by Uriel Garcilazo Cruz. Thank you Uriel!
```python
'''This script replaces names from a fasta file by using a reference table. The table has two columns. On the first column, the file names that can be found in the fasta file of interest.
On the second column, the new names. The program asks first for the table, then for the fasta file of interest. If the program couldn't find a particular name in the table, it will add
many zeros to it so that you chan check it up manually. BE CAREFUL with upper and lower cases as the script is case sensitive.

The output file contains a new fasta file with the corrected names.

Also, if the students are accessing directly through the server don't use tkinter, instead, enter the file names manually'''

import os
import tkinter.filedialog as Tkfiledialog

pathtofolder=os.getcwd()
table_of_reference_here='Wilpoe_Te_naming_table.txt'
fasta_file_here='Anteater_curated_TE_library_consensus.95.fasta'
#Old_Newnames=directory=Tkfiledialog.askopenfilename()#Select the file name with the table
#namefasta_file=directory=Tkfiledialog.askopenfilename()#Select the file name with the fasta file
Old_Newnames=os.path.join(pathtofolder,table_of_reference_here)
namefasta_file=os.path.join(pathtofolder,fasta_file_here)

lines_fastafile=[x for x in open(namefasta_file,'r').readlines()]
table_names=[d.strip().split('\t') for d in open(Old_Newnames,'r').readlines()]

finalstring=''

for line in lines_fastafile:
    if '>' in line:
        itemfound=False
        for each in table_names:
            if each[0] in line:
                finalstring+=line.replace(each[0],each[1])
                itemfound=True
        if not itemfound:
            finalstring+='00000_'+line
    else:
        finalstring+=line

exportpath=os.path.join(pathtofolder,'V2_'+fasta_file_here)
export=open(exportpath,'w')
export.write(finalstring)
export.close()
```
