# Assembling a Transcriptome

In order to run maker on a non-model organism we will need RNA evidence to provide a hint at where the genes are in the genome. This could be ESTs or RNA-seq data: nowadays, RNA-seq is more common. If you do not have RNA sequences for your species, you will have to look to public databases to see if there are any available you can use.  

# Finding RNA sequence data
We can look for RNA sequences on NCBI databases. Our best options are the EST database which has now been merged with the NCBI database.
We can [search](https://www.ncbi.nlm.nih.gov/nuccore/?term=Passeriformes+%5Borganism%5D) for Passeriformes in the nucleotide database and select "EST" as the sequence type. Hopefully there will be a species relatively close to your own species.  

There are only Taeniopygia, Lonchura, and Haemorhous in the database at the moment.

We can also search the Transcriptome Shotgun Assembly (TSA) database for Passeriformes to find assembled transcriptomes.

Here are all the passerine transcriptomes available on NCBI's TSA at the moment:  
Right now there are only oscine passerines available. The transcriptomes for Zonotrichia are most extensive, so we could use those, but we would prefer a suboscine right now, since the RNA sequences should be quite close (ideally the same species).

**Passerellidae**
[Zonotrichia albicollis](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA231930)
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/BC/GBBC01/GBBC01.1.fsa_nt.gz
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/BD/GBBD01/GBBD01.1.fsa_nt.gz
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/BD/GBBD01/GBBD01.2.fsa_nt.gz
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/BB/GBBB01/GBBB01.1.fsa_nt.gz
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/BB/GBBB01/GBBB01.2.fsa_nt.gz

**Estrildidae**
[Taeniopygia guttata](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA445653)
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GG/LD/GGLD01/GGLD01.1.fsa_nt.gz

**Fringillidae**
Carduelis chloris
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/CG/GBCG01/GBCG01.1.fsa_nt.gz

Acanthis 
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/IN/GBIN01/GBIN01.1.fsa_nt.gz
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GB/IK/GBIK01/GBIK01.1.fsa_nt.gz

**Sturnidae**
Sturnus vulgaris
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GF/DQ/GFDQ01/GFDQ01.1.fsa_nt.gz

**Muscicapidae**
Oenanthe oenanthe
wget https://sra-download.ncbi.nlm.nih.gov/traces/wgs03/wgs_aux/GF/YT/GFYT01/GFYT01.1.fsa_nt.gz

None of these are particularly close to our species! We will turn to a different database, the Sequence Read Archive (SRA) which houses raw sequence reads - unfortunately we will have to assemble those ourselves.  

We only want to look at suboscines, but there is no taxonomic filter in NCBI for suboscines so we will instead filter out the common oscines with this [search](https://www.ncbi.nlm.nih.gov/sra/?term=(passeriformes%5Borganism%5D)+NOT+Corvoidea%5Borganism%5D+NOT+Passeroidea%5Borganism%5D+NOT+Meliphagoidea%5Borganism%5D+NOT+Sylvioidea%5Borganism%5D+NOT+Passerellidae%5Borganism%5D+NOT+Muscicapidae%5Borganism%5D+NOT+Paridae%5Borganism%5D+NOT+Turdidae%5Borganism%5D+NOT+Sturnidae%5Borganism%5D), and select Source: RNA  
(passeriformes[organism]) NOT Corvoidea[organism] NOT Passeroidea[organism] NOT Meliphagoidea[organism] NOT Sylvioidea[organism] NOT Passerellidae[organism] NOT Muscicapidae[organism] NOT Paridae[organism] NOT Turdidae[organism] NOT Sturnidae[organism]  

Good news, there are results for three species of Pipridae, which is closer. Now we will have to assemble a transcritome de novo.  

There are 14 SRA results for *Manacus vitellinus*, so that is what we will assemble. We will accept data that is from 2016 or later for now.
search: `(("2016"[Publication Date] : "3000"[Publication Date])) AND (((passeriformes[organism]) NOT Corvoidea[organism] NOT Passeroidea[organism] NOT Meliphagoidea[organism] NOT Sylvioidea[organism] NOT Passerellidae[organism] NOT Muscicapidae[organism] NOT Paridae[organism] NOT Turdidae[organism] NOT Sturnidae[organism]) AND "Manacus vitellinus"[orgn:__txid328815])`  

# Assembling the Transcriptome

Luckily assembling a transcriptome is not difficult, and there is a program [Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki) that will assemble the whole transcriptome from raw reads, it will just take a several days to run on 14 threads.  

1) Download SRA files using [SRA toolkit](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc). The SRA files will appear in `~/ncbi/public`, so we will have to move them to our working directory.

```bash
mkdir ~/Genetic_resources/Manacus_RNA
cd ~/Genetic_resources/Manacus_RNA

#download the SRA files
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545929
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545930
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545931
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545932
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545933
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545934
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545935
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545936
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545937
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545938
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545939
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/prefetch SRR2545940

mv /home/else/ncbi/public/sra/*.sra ./
```
* SRR#######: These are the Accession numbers for the NCBI SRA entries that you want to download.

2) Convert SRA files to fastq files. I am running them all on separate terminals so that they run simultaneously. Note that the --defline-seq tag is required for Trinity to be able to correctly read the fastq file headers.

```bash
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545929 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545930 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545931 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545932 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545933 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545934 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545935 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545936 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545938 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545937 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545939 --defline-seq '@$sn[_$rn]/$ri'
cd ~/Genetic_resources/Manacus_RNA
~/tools/sratoolkit.2.9.6-1-ubuntu64/bin/fastq-dump -I --split-files SRR2545940 --defline-seq '@$sn[_$rn]/$ri'

rm ./*.sra
```
* --split-files: give the name of the accession number of your SRA file without the .sra extension

3) Create an input file

To run trinity most easily, we will specify a file that lists all the sequence files we want to use and which files are pairs for paired end data. It will have 4 columns, listing the condition name, biological replicate number, read1 file, and read2 file. We are not interested in the condition or replicate in this case, we will leave those columns with generic names.

As usual, to create a text file, run the cat command, copy-paste the contents of the new file to create the document, and then press ctrl-c to finish.
```bash
cat > Trinity_samples_file.txt
```
*****
>condA	condA_rep1	SRR2545929_1.fastq	SRR2545929_2.fastq	  
condB	condB_rep1	SRR2545930_1.fastq	SRR2545930_2.fastq	  
condC	condC_rep1	SRR2545931_1.fastq	SRR2545931_2.fastq	 
condD	condD_rep1	SRR2545932_1.fastq	SRR2545932_2.fastq	 
condE	condE_rep1	SRR2545933_1.fastq	SRR2545933_2.fastq	  
condF	condF_rep1	SRR2545934_1.fastq	SRR2545934_2.fastq	 
condG	condG_rep1	SRR2545935_1.fastq	SRR2545935_2.fastq	 
condH	condH_rep1	SRR2545936_1.fastq	SRR2545936_2.fastq	 
condI	condI_rep1	SRR2545937_1.fastq	SRR2545937_2.fastq	 
condJ	condJ_rep1	SRR2545938_1.fastq	SRR2545938_2.fastq	 
condK	condK_rep1	SRR2545939_1.fastq	SRR2545939_2.fastq	 
condL	condL_rep1	SRR2545940_1.fastq	SRR2545940_2.fastq	 
*****

## Run Trinity
Finally, after gathering and quality-controlling the data, you can build the transcriptome.

```bash
#setup environment. Make sure that Jellyfish, Salmon, Trimmomatic, and Bowtie2 are available in your environment.
cd ~/Genetic_resources/Manacus_RNA
TRIMMOMATIC_DIR=~/tools/Trimmomatic-0.38
export PATH=$PATH:~/tools
export PATH=$PATH:~/tools/Trimmomatic-0.38
export PATH=$PATH:~/tools/bin
export PATH=$PATH:~/tools/bowtie2-2.3.5-linux-x86_64
export PATH=$PATH:~/tools/salmon-latest_linux_x86_64/bin

~/tools/trinityrnaseq-Trinity-v2.8.5/Trinity --seqType fq --samples_file ~/Genetic_resources/Manacus_RNA/Trinity_samples_file.txt --CPU 10 --max_memory 50G --min_contig_length 150 --trimmomatic --normalize_by_read_set
```

* --seqType: specifies whether we are using fastq (fq) or fasta (fa) files
* --left: list of sequence files containing RNA sequences for read 1's of paired-end data
* --right: list of sequence files containing RNA sequences for read 2's of paired-end data
* --max_memory: max amount of memory to use (in Gb)
* --CPU: number of cores to use
* --samples_file: file listing where your sequence reads are (created above)
* --trimmomatic: perform adapter removal and quality trimming. By default, assumes standard TruSeq adapter sequences.
* (not used) --quality_trimming_params "ILLUMINACLIP:illumina.fa:2:30:10 SLIDINGWINDOW:4:5 LEADING:5 TRAILING:5 MINLEN:25" would allow you to customize the trimmomatic parameters.
* --normalize_by_read_set 

Note: **If the Trinity run fails** you can restart where you left off by simply running the exact same command again. It will detect the files that were already written and start again from there. For example, My terminal exited before the run finished when I was already 3 days into it. I reran the same command in a new terminal and it restarted. You can update --CPU if you want to change it but keep the inputs to the command the same as before.

This took about a week to run with 10 threads, but there were several interuptions as my terminal exited before finishing.  
The finished transcriptome will be in trinity_out_dir/Trinity.fasta  

## Quality check

First, make sure the transcriptome looks like a normal fasta file.
```bash
head trinity_out_dir/Trinity.fasta
```

Then, take a look at some assembly stats for your transcriptome.
```bash
~/tools/trinityrnaseq-Trinity-v2.8.5/util/TrinityStats.pl trinity_out_dir/Trinity.fasta
```
 
*Don't forget the reference:*  
Grabherr MG, Haas BJ, Yassour M, Levin JZ, Thompson DA, Amit I, Adiconis X, Fan L, Raychowdhury R, Zeng Q, Chen Z, Mauceli E, Hacohen N, Gnirke A, Rhind N, di Palma F, Birren BW, Nusbaum C, Lindblad-Toh K, Friedman N, Regev A. Full-length transcriptome assembly from RNA-seq data without a reference genome. Nat Biotechnol. 2011 May 15;29(7):644-52. doi: 10.1038/nbt.1883. PubMed PMID: 21572440.



NOTE: I would not repeat this, as there are actually Piprid transcriptomes available from Ensembl and Refseq, with more appearing.
Here is how I would have downloaded those.
```
mkdir /home/0_GENOMES1/0_Genetic_resources/Ensembl_transcriptomes
cd /home/0_GENOMES1/0_Genetic_resources/Ensembl_transcriptomes
wget ftp://ftp.ensembl.org/pub/release-98/fasta/lepidothrix_coronata/cdna/Lepidothrix_coronata.Lepidothrix_coronata-1.0.cdna.all.fa.gz
wget ftp://ftp.ensembl.org/pub/release-98/fasta/manacus_vitellinus/cds/Manacus_vitellinus.ASM171598v2.cds.all.fa.gz

mkdir /home/0_GENOMES1/0_Genetic_resources/Refseq_transcriptomes
cd /home/0_GENOMES1/0_Genetic_resources/Refseq_transcriptomes

#Empidonax trailli
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/annotation_releases/164674/100/GCF_003031625.1_ASM303162v1/GCF_003031625.1_ASM303162v1_rna.fna.gz
mv GCF_003031625.1_ASM303162v1_rna.fna.gz Emptra_rna.fa.gz

#Corapipo altera
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/annotation_releases/415028/100/GCF_003945725.1_ASM394572v1/GCF_003945725.1_ASM394572v1_rna.fna.gz 	
mv GCF_003945725.1_ASM394572v1_rna.fna.gz Coralt_rna.fa.gz

#Pipra filicauda
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/annotation_releases/649802/100/GCF_003945595.1_ASM394559v1/GCF_003945595.1_ASM394559v1_rna.fna.gz 
mv GCF_003945595.1_ASM394559v1_rna.fna.gz Pipfil_rna.fa.gz

#Neopelma chrysocephalum
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/annotation_releases/114329/100/GCF_003984885.1_ASM398488v2/GCF_003984885.1_ASM398488v2_rna.fna.gz ; mv GCF_003984885.1_ASM398488v2_rna.fna.gz Neochr_rna.fa.gz

gunzip -k ./*.fa.gz
```