Now we have some gene models. For this data to be useful, we ideally want to know more: what are these genes? What do they do?

The first thing we will do is look for conserved protein domains using InterPro Scan, which can search several different databases to each of our protein sequences and give a list of matches. These will be used to assign possible functions to the genes, including GO terms. A list of analyses within Interproscan are [here](https://github.com/ebi-pf-team/interproscan/wiki/HowToRun#included-analyses), which also includes all the instructions for running the program.

I am first just running with pfam and then tried running with all of the supported analyses. However, when I tried to run all of them it failed after days of running, so I will not try that again. I am trying the following (one-at-a-time):
* Pfam (used to save unsupported gene models from the filter) v32.0
* PANTHER v14.1
* ProSiteProfiles v2019_01
* SMART v7.1
* PRINTS v42.0
This list was based on the most frequently used ones that I noticed being used in the literature. I am not running ProDom because it has not been updated since 2006 and so is now considered [deprecated](https://github.com/ebi-pf-team/interproscan/wiki/Interproscan5_35_74_ReleaseNotes).

```bash
#set up environment
export PATH=$PATH:/home/0_PROGRAMS/interproscan/interproscan-5.39-77.0 #for all of interproscan
export PATH=$PATH:~/tools/maker/bin
#when I am working on the nodes rather than main server:
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

mkdir ~/Willisornis/functional_ann
cd ~/Willisornis/functional_ann

#search for hits in the databases with Interproscan

#Pfam
interproscan.sh -iprlookup -goterms -appl Pfam-32.0 -i /home/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4prots -f tsv
#PANTHER
time interproscan.sh -iprlookup -goterms -appl PANTHER-14.1 -i /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4protsPANTHER -f tsv
#Others
time interproscan.sh -iprlookup -goterms -appl PRINTS-42.0 -i /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4protsPrints -f tsv
time interproscan.sh -iprlookup -goterms -appl ProSiteProfiles-2019_01 -i /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4protsProsite -f tsv
time interproscan.sh -iprlookup -goterms -appl SMART-7.1 -i /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4protsSMART -f tsv
#This would run on ALL installed databases (alternative to the above) - would take days and crashed too many times for me which is why I did them separately!
#interproscan.sh -iprlookup -goterms -i /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta -b Willisornis_R4prots_allinterpro -f tsv

#concatenate all 5 interpro results if done separately
cat Willisornis_R4prot*.tsv > Willisornis_R4prot_5interpro.tsv

#add the interpro results to the maker gff
ipr_update_gff /hhome/else/Willisornis/maker/maker_R4/Willisornis_rnd4.all.maker.gff Willisornis_R4prot_5interpro.tsv > Willisornis.max.functional_5interpro.gff

#filter to retain only best genes, compare two filtering options:
perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -d Willisornis.max.functional_5interpro.gff > Willisornis.default.functional_5interpro.gff #default build: only proteins with RNA/protein evidence (AED<1)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -s Willisornis.max.functional_5interpro.gff > Willisornis.standard.functional_5interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter_PfamPANTHER.pl -s Willisornis.max.functional_5interpro.gff > Willisornis.standardplus.functional_5interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam or PANTHER)

#count number of genes
grep -cP '\tgene\t' /hhome/else/Willisornis/maker/maker_R4/Willisornis_rnd4.all.maker.gff #number of genes originally
grep -cP '\tgene\t' Willisornis.default.functional_5interpro.gff #number genes with evidence
grep -cP '\tgene\t' Willisornis.standard.functional_5interpro.gff #number genes with evidence OR Pfam domain
grep -cP '\tgene\t' Willisornis.standardplus.functional_5interpro.gff #number genes with evidence OR Pfam domain or PANTHER
less Willisornis_R4prots.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with Pfam annotations
less Willisornis_R4protsPANTHER.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with PANTHER annotations
less Willisornis_R4protsPrints.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with PRINTS annotations
less Willisornis_R4protsProsite.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with Prosite annotations
less Willisornis_R4protsSMART.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with Pfam annotations
less Willisornis_R4prot_5interpro.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with annotations
```
Interproscan
* `-iprlookup`: "Option that provides mappings from matched member database signatures to the InterPro entries that they are integrated into."
* `-appl`: the database to search. If not specified, it will search all the ones you have installed.
* `-goterms`: "Option that provides mappings to the Gene Ontology (GO). These mappings are based on the matched manually curated InterPro entries. (IMPLIES -iprlookup option)"
* `-b`: "Optionally, you can supply a path and base name (excluding a file extension) for the results file as follows:
* `-pa`: "Option that provides mappings from matches to pathway information, which is based on the matched manually curated InterPro entries. (IMPLIES -iprlookup option). The different pathways databases that I5 provides cross links to are: KEGG, MetaCyc, Reactome"
* `-i`: input fasta file of protein (not DNA!) sequences

**Timing**: with Pfam, interproscan took 10 mins. With PANTHER, it took 690m44 (6164m16 user time) on 24 threads. PRINTS, 44m38 (267m58). Prosite, 157m55 (1184m13). SMART, 190m13 (1521m59).

**Results**:

Step|# Genes|
---|---|
Raw input|43,959|
No protein/RNA evidence|29,386|
Pfam annotated|14,446|
PANTHER annotated|16,941|
PRINTS annotated|3740|
Prosite annotated|8380|
SMART annotated|7472|
Pfam OR PANTHER annotated|17,117|
Pfam OR PANTHER OR PRINTS OR Prosite|17,307
Pfam OR PANTHER OR PRINTS OR Prosite OR SMART|17,311
protein/RNA evidence|14,573|
evidence and/or Pfam|16,418|
evidence and/or PANTHER and/or Pfam|17,475|
Without evidence with Pfam|1845 6.28%|
With evidence without Pfam|1972 13.53%|

Oops I looked at the script and it is specifically looking for pfam domains, adding PANTHER does not help! In order to also keep the ones associated with PANTHER...
```bash
nano /home/0_PROGRAMS/Genome_annotation/quality_filter.pl
#change it to allow the Dbxref to equal Pfam OR PANTHER and save as quality_filter_PfamPANTHER.pl
```

So, 1845 genes were saved from the dumpster with Pfam InterproScan. This is only 6.28% of the unsupported gene dumpster! In contrast, 86.47% of the supported genes received Pfam domains. An additional 1057 were saved by PANTHER

As the next step, let's look at some [stats](https://www.biostars.org/p/402537/) with [AGAT](https://github.com/NBISweden/AGAT), and also make a FASTA file and get a BUSCO score:  
```bash
cd ~/Willisornis/functional_ann
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PERL5LIB=$PERL5LIB:/hhome/else/.conda/envs/agat/lib/site_perl/5.26.2/

#AGAT stats
conda activate agat
agat_sp_functional_statistics.pl --gff Willisornis.standardplus.functional_5interpro.gff -o Willisornis.standardplus.functional_5interpro.stats
```

### Rename genes
Let's simplify the gene names before we get too far - right now they are named very long codenames.
```bash
# create naming table (there are additional options for naming beyond defaults)
maker_map_ids --prefix WilPoe --justify 5 Willisornis.standardplus.functional_5interpro.gff > Willisornis.standardplus.name.map
# replace names in GFF files
map_gff_ids Willisornis.standardplus.name.map Willisornis.standardplus.functional_5interpro.gff
# replace names in FASTA headers
cp /hhome/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.proteins.fasta .
map_fasta_ids Willisornis.standardplus.name.map Willisornis_input.all.maker.proteins.fasta
```
* `--justify`: the number of digits to use in the name (5 = 99,999 unique numbers max)

### Filter and evaluate
Now, I will need a FASTA file of the proteins and transcripts with only the filtered genes.
```bash
#extract protein FASTA
#agat_sp_extract_sequences.pl --protein --gff Willisornis.standardplus.functional_5interpro.gff -f /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/genome/Willisornis700.fasta -o Willisornis_standardproteins.fasta #did not work, complained that "Got a sequence without letters. Could not guess alphabet"

#Use seqtk to filter a fasta based on a list
cat Willisornis.standardplus.name.map | cut -f 2 | sort | uniq > Willisornis.standardplus_genelist #extract list of genes to keep
/home/0_PROGRAMS/seqtk/seqtk subseq Willisornis_input.all.maker.proteins.fasta Willisornis.standardplus_genelist > Willisornis.standardplus.proteins.fasta

#BUSCO evaluation
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/hhome/else/Willisornis/maker/busco/Willisornis_busco_config.ini"
cp /home/0_PROGRAMS/busco/config/config.ini ./Willisornis_busco_config.ini
python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i Willisornis.standardplus.proteins.fasta  -o Willisornis.standardplus.proteins -l /home/0_PROGRAMS/busco/aves_odb9 -m proteins -c 24 -sp chicken -z --augustus_parameters='--progress=true'
python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i Willisornis_input.all.maker.proteins.fasta -o Willisornis.all.proteins -l /home/0_PROGRAMS/busco/aves_odb9 -m proteins -c 24 -sp chicken -z --augustus_parameters='--progress=true'
```
* `--protein`: extract the amino acid sequences not DNA
* `--gff`: input GFF file to extract the protein sequences with
* `-f`: fasta file corresponding to the GFF file, to take the sequences from

**Results**

BUSCO run|#genes|total BUSCOs|complete BUSCOs|complete single-copy BUSCOS|complete duplicated BUSCOS|fragmented BUSCOS|missing BUSCOS|run time (s)|
|---|---|---|---|---|---|---|---|---|
Genome|?|96.7%|C:91.6%|S:90.1%|D:1.5%|F:5.1%|M:3.3%|15862.3|
Maker round 4 (raw)|43,959|93% |C:80.1% (3937)|S:78.7%|D:1.4 (68)|F:12.9%(635)|M:7.0% (343)|207.7|
Willisornis.standardplus.proteins.fasta|17,475 |93% |C:80.1% (3937)|S:78.7%|D:1.4 (69)|F:12.9% (634)|M:7.0% (344)|202.1|

I am pleased! We cut the numbers of protein candidates down from 43,959 to 17,475 (removing 60.2% of the raw candidates) but all it costed was a single fragmented BUSCO!!

Now, we have a fasta file of protein sequences, with their protein domains annotated with 5 databases.

### Homology-based annotation
Now we will rename our proteins based on homology with other named proteins, especially the curated Swiss-Prot database.  
First, we need to get the Swissprot/Uniprot database and make a blast database for it. I already downloaded this database in the [proteomes](https://github.com/elsemikk/Willisornis_Genome/blob/master/Proteomes.md) section.  
Then, we will BLAST our proteome against this BLAST database to see what our proteins match.
```bash
#setup environment
cd /home/else/Willisornis/functional_ann
export PATH=$PATH:/home/0_PROGRAMS/ncbi-blast-2.9.0+/bin/

makeblastdb -in /home/0_GENOMES1/0_Genetic_resources/Uniprot_swissprot/uniprot_sprot.fasta -input_type fasta -dbtype prot #the blast database will be made at the location of the uniprot_sprot.fasta file, not the working directory.

blastp -db /home/0_GENOMES1/0_Genetic_resources/Uniprot_swissprot/uniprot_sprot.fasta -query Willisornis.standardplus.proteins.fasta -out maker2uni.blastp -evalue .000001 -outfmt 6 -num_alignments 1 -seg yes -soft_masking true -lcase_masking -max_hsps 1 -num_threads 23

#add the homology results to the FASTA file
maker_functional_fasta /home/0_GENOMES1/0_Genetic_resources/Uniprot_swissprot/uniprot_sprot.fasta maker2uni.blastp Willisornis.standardplus.proteins.fasta > Willisornis.standardplus.SPproteins.fasta

```
* `-outfmt 6`: place into tabular format
* `-num_alignments 1`: only one hit per query sequence. This should be the best hit, but beware if the database is very redundant this might not work as expected. There is also a bug in earlier version that affected this so make sure you are using a version more recent than Dec 2018! ([reference](https://academic-oup-com.myaccess.library.utoronto.ca/bioinformatics/article/35/9/1613/5106166))
* `-max_hsps 1`: only one hit per sequence pair. An HSP is a "Highly Similar Sequences" pair which align without gaps. Because there may be gaps in the alignment between two protein sequences, this means that you can get multiple HSPs per query/subject pair. These would not be useful to us as we simply want to find the best hit and use it to name our proteins, so we limit it to 1 HSP per query/subject combo.

Next, we can supplement Swiss-prot with Trembl. First, I will download all Trembl entries from birds.
I conducted this [search](https://www.uniprot.org/uniprot/?query=taxonomy%3A%22Aves+%289AVES%29+%5B8782%5D%22+reviewed%3Ano&sort=score): `taxonomy:"Aves [8782]" AND reviewed:no -> Download -> Download All, FASTA (Canonical), compressed -> Go`  
On Dec 19, 2019, when there were 876,730 bird sequences in Trembl.

```bash
mkdir /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl
mv ~/Downloads/uniprot*.fasta.gz ~/Downloads/Trembl_birds_Dec2019.fasta.gz
cp ~/Downloads/Trembl_birds_Dec2019.fasta.gz /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl
gunzip /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta.gz
grep -c "^>" /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta #876,730 sequences
```

```bash
#setup environment
cd /home/else/Willisornis/functional_ann
export PATH=$PATH:/home/0_PROGRAMS/ncbi-blast-2.9.0+/bin/

makeblastdb -in /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta -input_type fasta -dbtype prot #the blast database will be made at the location of the fasta file, not the working directory.

blastp -db /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta -query Willisornis.standardplus.proteins.fasta -out maker2trem.blastp -evalue .000001 -outfmt 6 -num_alignments 1 -seg yes -soft_masking true -lcase_masking -max_hsps 1 -num_threads 23

#add the homology results to the FASTA file
maker_functional_fasta /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta maker2trem.blastp Willisornis.standardplus.proteins.fasta > Willisornis.standardplus.Trproteins.fasta
```

**Results**:
With Swiss-prot, I got 16,981 hits. With Trembl, I got 17,341 hits (367 that Swissprot did not hit). 17,348 had a hit in a least one of those. I had 17,475 total genes so almost all had hits!!! 99.3%!!!!

In the end, I would like to name the proteins first with the Swiss-prot names, and then with the Trembl names for the 367 that had hits with Trembl but not Swissprot. That is because many of the Trembl proteins are named uninformative names like "Uncharacterized protein" while the Swissprot proteins are manually curated and higher quality.
```bash
cat maker2uni.blastp | cut -f 1 > swissprot_matched
grep -v -f swissprot_matched maker2trem.blastp  > onlytrembl_matched
cat maker2uni.blastp onlytrembl_matched > nr_blastp_for_naming

cat /home/0_GENOMES1/0_Genetic_resources/Uniprot_swissprot/uniprot_sprot.fasta /home/0_GENOMES1/0_Genetic_resources/Uniprot_trembl/Trembl_birds_Dec2019.fasta > /home/0_GENOMES1/0_Genetic_resources/Uniprot_cat_Tremblbirds.temp

#add the homology results to the FASTA file
maker_functional_fasta /home/0_GENOMES1/0_Genetic_resources/Uniprot_cat_Tremblbirds.temp nr_blastp_for_naming Willisornis.standardplus.proteins.fasta > Willisornis.standardplus.namedproteins.fasta

#finally, get the transcript data as well and name the same way
cp /home/else/Willisornis/maker/maker_R4/Willisornis_input.all.maker.transcripts.fasta .
map_fasta_ids Willisornis.standardplus.name.map Willisornis_input.all.maker.transcripts.fasta
/home/0_PROGRAMS/seqtk/seqtk subseq Willisornis_input.all.maker.transcripts.fasta Willisornis.standardplus_genelist > Willisornis.standardplus.transcripts.fasta
maker_functional_fasta /home/0_GENOMES1/0_Genetic_resources/Uniprot_cat_Tremblbirds.temp nr_blastp_for_naming Willisornis.standardplus.transcripts.fasta > Willisornis.standardplus.namedtranscripts.fasta

```
