This section is the workflow for annotating Hypocnemis and Rhegmatorhina.

Since they are so closely related, I will use the same repeat library and protein library and SNAP hmm. This way they will hopefully be more standardized.

# Hypocnemis ochrogyna

![bird](https://upload.wikimedia.org/wikipedia/commons/8/8e/Hypocnemis_rondoni_-_Manicore_warbling-antbird.jpg)
**This file is licensed under the Creative Commons [Attribution-Share Alike 2.0](https://creativecommons.org/licenses/by-sa/2.0/deed.en) Generic license.**

# Get genome sequence
```bash
mkdir /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker
cd /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker
cp /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch.fasta ./Hypoch_input.fasta
nano Hypoch_input.fasta #be patient, it will take a while to load. Don't press anything.
#Use the arrow keys to scroll right, and change the name of the first line from ">0" to ">0start"
#press ctrl-o to save, then ctrl-x to exit
#check that it worked ok
less Hypoch_input.fasta #all good, press q to quit.
gzip Hypoch_input.fasta
```

# Transcriptome data
I am skipping RNA alignment because we are not training a SNAP hmm this time.

# protein data
I am providing the same protein input as used for Willisornis, plus the Willisornis proteome.
```bash
cat ~/anteater/maker/maker_R1/Anteaterprot_input.fasta /home/0_GENOMES1/0_Genetic_resources/Orthofinder/Willisornis.fa > Hypochprot_input.fasta
```

# repeat library
I am providing the same repeat library input as used for Willisornis.

```bash
cp ~/anteater/RepeatMasking/Anteater_repeat_library_cleaned_withFicalbUracya.lib ./Hypochreps_input.fasta
```

# SNAP hmm file
I am providing the same SNAP hmm input as used for Willisornis round 4.
```bash
cp ~/anteater/maker/maker_R3/snap/round3/Anteater_rnd3.zff.length50_aed0.25.hmm .
```

# control files

```bash
cat > maker_exe.ctl
```
```
#-----Location of Executables Used by MAKER/EVALUATOR
makeblastdb=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/blast+/2.7.1/bin/makeblastdb #location of NCBI+ makeblastdb executable
blastn=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/blast+/2.7.1/bin/blastn #location of NCBI+ blastn executable
blastx=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/blast+/2.7.1/bin/blastx #location of NCBI+ blastx executable
tblastx=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/blast+/2.7.1/bin/tblastx #location of NCBI+ tblastx executable
formatdb= #location of NCBI formatdb executable
blastall= #location of NCBI blastall executable
xdformat= #location of WUBLAST xdformat executable
blasta= #location of WUBLAST blasta executable
RepeatMasker=/gpfs/fs0/path/to/folder/tools/RepeatMasker/RepeatMasker #location of RepeatMasker executable
exonerate=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/exonerate/2.4.0/bin/exonerate #location of exonerate executable

#-----Ab-initio Gene Prediction Algorithms
snap=/gpfs/fs0/path/to/folder/tools/snap/snap #location of snap executable
gmhmme3=/path/to/folder/tools/gm_et_linux_64/gmhmme3 #location of eukaryotic genemark executable
gmhmmp= #location of prokaryotic genemark executable
augustus=/gpfs/fs0/path/to/folder/tools/Augustus/bin/augustus #location of augustus executable
fgenesh= #location of fgenesh executable
tRNAscan-SE=/gpfs/fs0/path/to/folder/tools/tRNAscan-SE-2.0/tRNAscan-SE #location of trnascan executable
snoscan=/gpfs/fs0/path/to/folder/tools/snoscan-0.9.1/snoscan #location of snoscan executable

#-----Other Algorithms
probuild= #location of probuild executable (required for genemark)
```

```bash
cat > maker_bopt.ctl
```
```
#-----BLAST and Exonerate Statistics Thresholds
blast_type=ncbi+ #set to 'ncbi+', 'ncbi' or 'wublast'

pcov_blastn=0.8 #Blastn Percent Coverage Threhold EST-Genome Alignments
pid_blastn=0.85 #Blastn Percent Identity Threshold EST-Genome Aligments
eval_blastn=1e-10 #Blastn eval cutoff
bit_blastn=40 #Blastn bit cutoff
depth_blastn=0 #Blastn depth cutoff (0 to disable cutoff)

pcov_blastx=0.5 #Blastx Percent Coverage Threhold Protein-Genome Alignments
pid_blastx=0.4 #Blastx Percent Identity Threshold Protein-Genome Aligments
eval_blastx=1e-06 #Blastx eval cutoff
bit_blastx=30 #Blastx bit cutoff
depth_blastx=0 #Blastx depth cutoff (0 to disable cutoff)

pcov_tblastx=0.8 #tBlastx Percent Coverage Threhold alt-EST-Genome Alignments
pid_tblastx=0.85 #tBlastx Percent Identity Threshold alt-EST-Genome Aligments
eval_tblastx=1e-10 #tBlastx eval cutoff
bit_tblastx=40 #tBlastx bit cutoff
depth_tblastx=0 #tBlastx depth cutoff (0 to disable cutoff)

pcov_rm_blastx=0.5 #Blastx Percent Coverage Threhold For Transposable Element Masking
pid_rm_blastx=0.4 #Blastx Percent Identity Threshold For Transposbale Element Masking
eval_rm_blastx=1e-06 #Blastx eval cutoff for transposable element masking
bit_rm_blastx=30 #Blastx bit cutoff for transposable element masking

ep_score_limit=20 #Exonerate protein percent of maximal score threshold
en_score_limit=20 #Exonerate nucleotide percent of maximal score threshold
```

```bash
cat > maker_opts.ctl
```
changing: 
* genome=Hypoch_input.fasta
* altest= #
* protein=Hypochprot_input.fasta
* rmlib=Hypochreps_input.fasta
* est2genome=0
* augustus_species=chicken
* snaphmm=Anteater_rnd3.zff.length50_aed0.25.hmm
* keep_preds=1

```
#-----Genome (these are always required)
genome=Hypoch_input.fasta #genome sequence (fasta file or fasta embeded in GFF3 file)
organism_type=eukaryotic #eukaryotic or prokaryotic. Default is eukaryotic

#-----Re-annotation Using MAKER Derived GFF3
maker_gff= #MAKER derived GFF3 file
est_pass=0 #use ESTs in maker_gff: 1 = yes, 0 = no
altest_pass=0 #use alternate organism ESTs in maker_gff: 1 = yes, 0 = no
protein_pass=0 #use protein alignments in maker_gff: 1 = yes, 0 = no
rm_pass=0 #use repeats in maker_gff: 1 = yes, 0 = no
model_pass=0 #use gene models in maker_gff: 1 = yes, 0 = no
pred_pass=0 #use ab-initio predictions in maker_gff: 1 = yes, 0 = no
other_pass=0 #passthrough anyything else in maker_gff: 1 = yes, 0 = no

#-----EST Evidence (for best results provide a file for at least one)
est= #set of ESTs or assembled mRNA-seq in fasta format
altest= #EST/cDNA sequence file in fasta format from an alternate organism
est_gff= #aligned ESTs or mRNA-seq from an external GFF3 file
altest_gff= #aligned ESTs from a closly relate species in GFF3 format

#-----Protein Homology Evidence (for best results provide a file for at least one)
protein=Hypochprot_input.fasta  #protein sequence file in fasta format (i.e. from mutiple oransisms)
protein_gff=  #aligned protein homology evidence from an external GFF3 file

#-----Repeat Masking (leave values blank to skip repeat masking)
model_org=all #select a model organism for RepBase masking in RepeatMasker
rmlib=Hypochreps_input.fasta #provide an organism specific repeat library in fasta format for RepeatMasker
repeat_protein=/gpfs/fs0/path/to/folder/tools/maker/data/te_proteins.fasta #provide a fasta file of transposable element proteins for RepeatRunner
rm_gff= #pre-identified repeat elements from an external GFF3 file
prok_rm=0 #forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no
softmask=1 #use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)

#-----Gene Prediction
snaphmm=Anteater_rnd3.zff.length50_aed0.25.hmm #SNAP HMM file
gmhmm= #GeneMark HMM file
augustus_species=chicken #Augustus gene prediction species model
fgenesh_par_file= #FGENESH parameter file
pred_gff= #ab-initio predictions from an external GFF3 file
model_gff= #annotated gene models from an external GFF3 file (annotation pass-through)
est2genome=0 #infer gene predictions directly from ESTs, 1 = yes, 0 = no
protein2genome=1 #infer predictions from protein homology, 1 = yes, 0 = no
trna=0 #find tRNAs with tRNAscan, 1 = yes, 0 = no
snoscan_rrna= #rRNA file to have Snoscan find snoRNAs
unmask=0 #also run ab-initio prediction programs on unmasked sequence, 1 = yes, 0 = no

#-----Other Annotation Feature Types (features MAKER doesn't recognize)
other_gff= #extra features to pass-through to final MAKER generated GFF3 file

#-----External Application Behavior Options
alt_peptide=C #amino acid used to replace non-standard amino acids in BLAST databases
cpus=10 #max number of cpus to use in BLAST and RepeatMasker (not for MPI, leave 1 when using MPI)

#-----MAKER Behavior Options
max_dna_len=100000 #length for dividing up contigs into chunks (increases/decreases memory usage)
min_contig=10000 #skip genome contigs below this length (under 10kb are often useless)

pred_flank=200 #flank for extending evidence clusters sent to gene predictors
pred_stats=0 #report AED and QI statistics for all predictions as well as models
AED_threshold=1 #Maximum Annotation Edit Distance allowed (bound by 0 and 1)
min_protein=0 #require at least this many amino acids in predicted proteins
alt_splice=0 #Take extra steps to try and find alternative splicing, 1 = yes, 0 = no
always_complete=0 #extra steps to force start and stop codons, 1 = yes, 0 = no
map_forward=0 #map names and attributes forward from old GFF3 genes, 1 = yes, 0 = no
keep_preds=1 #Concordance threshold to add unsupported gene prediction (bound by 0 and 1)

split_hit=10000 #length for the splitting of hits (expected max intron size for evidence alignments)
single_exon=0 #consider single exon EST evidence when generating annotations, 1 = yes, 0 = no
single_length=250 #min length required for single exon ESTs if 'single_exon is enabled'
correct_est_fusion=0 #limits use of ESTs in annotation to avoid fusion genes

tries=2 #number of times to try a contig if there is a failure for some reason
clean_try=0 #remove all data from previous run before retrying, 1 = yes, 0 = no
clean_up=0 #removes theVoid directory with individual analysis files, 1 = yes, 0 = no
TMP= #specify a directory other than the system default temporary directory for temporary files
```

# Transfer data to Niagara and subset genome
```bash
mkdir ~/temp
cp -r ../Hypoch_maker ~/temp
#log into Niagara supercomputer and make a folder for the annotation
mkdir /path/to/folder/0_MAKER/Hypoch
cd /path/to/folder/0_MAKER/Hypoch
#get all the input files
scp else@[IP Address Goes Here]:/home/else/temp/Hypoch_maker/* . #transfers all files in this folder onto othe supercomputer
#everything should say 100% done
gunzip Hypoch_input.fasta.gz

#We will split the genome into chunks to run on each chunk separately.
module load NiaEnv
module load CCEnv
module load StdEnv
module load intel/2019.3
module load nixpkgs/16.09
module load gcccore/.8.3.0
module load gcc/7.3.0
module load perl/5.22.4
module load bioperl/1.7.1 #1.7.5
module load exonerate/2.4.0
module load trf/4.09
module load rmblast/2.9.0
module load blast+/2.7.1
module load openmpi/3.1.2
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/maker/bin
fasta_tool --chunk 100 Hypoch_input.fasta #this tool comes with maker, in the folder maker/bin.
#created 80 chunks labelled 000 to 080
```

Now to make the input files, reusing template from Anteater.
```bash
cp ../anteater/R4_Niagara_submitfile_anteatermaker_chunk*.sh .
```
example:
```bash
#!/bin/bash
#SBATCH --nodes=1 #
#SBATCH --ntasks-per-node=1#
#SBATCH --cpus-per-task=40
#SBATCH --mem=0 #set to 0 to use all available memory of each requested node. Ad
vatage, any of the ntasks of each node can use any amount of the memory provided
 to sum does not exceed thata available to the node
#SBATCH --time=01:30:00
#SBATCH --job-name Anteater_Maker_Round4
#SBATCH --output=OUTPUT_R4_01_%.txt

#export paths
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/RepeatMasker 
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/snap 
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/snap/Zoe 
export ZOE=/gpfs/fs0/path/to/folder/tools/snap/Zoe
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/exonerate-2.2.0-x86_64/bin
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/maker/bin
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/snoscan-0.9.1
export PATH=$PATH:/gpfs/fs0/path/to/folder/tools/tRNAscan-SE-2.0
export LD_PRELOAD=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx512/Compiler/gcc7.3/openmpi/3.1.2/lib/libmpi.so
export PATH=$PATH:/path/to/folder/tools/gm_et_linux_64/

#load modules
module load NiaEnv
module load CCEnv
module load StdEnv
module load intel/2019.3
module load nixpkgs/16.09
module load gcccore/.8.3.0
module load gcc/7.3.0
module load perl/5.22.4
module load bioperl/1.7.1 #1.7.5
module load exonerate/2.4.0
module load trf/4.09
module load rmblast/2.9.0
module load blast+/2.7.1
module load openmpi/3.1.2
module load augustus/3.3
export PATH=/gpfs/fs0/path/to/folder/tools/Augustus/bin:/gpfs/fs0/path/to/folder/tools/Augustus/scripts:$PATH
export AUGUSTUS_CONFIG_PATH=/gpfs/fs0/path/to/folder/tools/Augustus/config/

export OMP_NUM_THREADS=40
export PERL5LIB=$PERL5LIB:/gpfs/fs0/path/to/folder/tools/tRNAscan-SE-2.0/lib
export PERL5LIB=$PERL5LIB:/path/to/folder/tools/perl/lib/perl5
cp /gpfs/fs0/path/to/folder/0_MAKER/anteater/gm_key ~/.gm_key

#User sets these options
WD=/path/to/folder/0_MAKER/anteater

#automated code to run (don't change)
   cd $WD
   #job1
   (maker -cpus 10 -base Anteater_input -g Anteater_input_000.fasta) &
   (maker -cpus 10 -base Anteater_input -g Anteater_input_001.fasta) &
   (maker -cpus 10 -base Anteater_input -g Anteater_input_002.fasta) &
   (maker -cpus 10 -base Anteater_input -g Anteater_input_003.fasta) &
wait
```

Modify Willisornis template for Hypocnemis
```bash
sed -i 's/Anteater_input/Hypoch_input/g' ./R4_Niagara_submitfile_anteatermaker_chunk*.sh
sed -i 's/anteater/Hypoch/g' ./R4_Niagara_submitfile_anteatermaker_chunk*.sh
sed -i 's/Anteater_Maker_Round4/Hypoch_Maker_Round1.4/g' ./R4_Niagara_submitfile_anteatermaker_chunk*.sh
sed -i 's/OUTPUT_R4/OUTPUT_R1/g' ./R4_Niagara_submitfile_anteatermaker_chunk*.sh
sed -i 's/time=01:30:00/time=11:59:00/g' ./R4_Niagara_submitfile_anteatermaker_chunk*.sh
find R4_Niagara_submitfile_anteatermaker_chunk*.sh | sed 's/R4_Niagara_submitfile_anteatermaker_chunk//g' | while read line ; do mv R4_Niagara_submitfile_anteatermaker_chunk"$line" R1_Niagara_submitfile_Hypochmaker_chunk"$line" ; done
sed -i 's/time=00:59:00/time=11:59:00/g' ./R1_Niagara_submitfile_Hypochmaker_chunk*.sh
sed -i 's/time=00:45:00/time=11:59:00/g' ./R1_Niagara_submitfile_Hypochmaker_chunk*.sh
```

The chunks of anteater only went up to 63 but now we have 80. We need 5 more
```bash
cat > nums.txt
```
```
17
18
19
20
```
```bash
#make copies of the input file
cat nums.txt | while read line ; do cat R1_Niagara_submitfile_Hypochmaker_chunk01.sh | sed "s/OUTPUT_R1_01/OUTPUT_R1_$line/g" | sed "s/job1/job$line/g" > R1_Niagara_submitfile_Hypochmaker_chunk"$line".sh ; done 
#Now I went in by hand editing the input fasta files to go up from 000 to 063 with 4 per file.
nano R1_Niagara_submitfile_Hypochmaker_chunk17.sh #64, 65, 66, 67
nano R1_Niagara_submitfile_Hypochmaker_chunk18.sh #68 69 70 71
nano R1_Niagara_submitfile_Hypochmaker_chunk19.sh #72 73 74 75
nano R1_Niagara_submitfile_Hypochmaker_chunk20.sh #76 77 78 79 80 #cpus 8
chmod +x ./*.sh #make executable!
```
Submit!
```bash
sbatch R1_Niagara_submitfile_Hypochmaker_chunk01.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk02.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk03.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk04.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk05.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk06.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk07.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk08.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk09.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk10.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk11.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk12.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk13.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk14.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk15.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk16.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk17.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk18.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk19.sh
sbatch R1_Niagara_submitfile_Hypochmaker_chunk20.sh
```
```bash
squeue -u $USER
squeue --start -j 2394061
```

After maker is done, check results
```bash
cd /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker
scp -r elsemikk@niagara.scinet.utoronto.ca:/path/to/folder/0_MAKER/Hypoch .
```

```bash
cd Hypoch

#check if there are any that failed and count how many were skipped vs finished
grep -c "SKIPPED_SMALL" ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log 
grep -c "STARTED" ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log 
grep -c "FINISHED" ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log 
grep -c "DIED_SKIPPED_PERMANENT" ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log #none! That is good
wc -l ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log #just counting lines to make sure the preceeding numbers add up

#make sure that all of the contigs were run
grep -c "^>" Hypoch_input.fasta #Count contigs of original fasta file
cat Hypoch_input_0*.fasta | grep -c "^>" #Count input contigs of the many chunks - should add to the same number as the original above
cut -f 1 ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log | sort | uniq | wc -l #count number of unique contigs in the log (contigs that actually ran in the maker pipeline)

grep "^>" Hypoch_input.fasta | sed 's/>//g' | sed 's/ .*$//g' | sort | uniq | wc -l
cut -f 1 ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log | sort | uniq
#If all looks good:

#Create files merging all data into a fasta and GFF3 file
time gff3_merge -s -d ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log > Hypoch.all.maker.gff
time fasta_merge -d ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log
#Create a gff3 file without sequences
time gff3_merge -n -s -d ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log > Hypoch.all.maker.noseq.gff
#done
```

Type|number|
---|---|
Total scaffolds|27,493|
run scaffolds|27,256|
Skipped_small|23,732|
started|3545|
finished|3523|
died|0|
lines|30,809|

Why did 237 contigs not run?
```bash
#which ones did not work?
grep "^>" Hypoch_input.fasta | sed 's/>//g' | sed 's/ .*$//g' | sort | uniq > temp
cut -f 1 ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log | sort | uniq > temp2
grep -vxf temp2 temp
```
Turns out that they are just small contigs, its ok, maker skips small contigs. I checked the output logs and these contigs WERE run, but gave the message "Skipping the contig because it is too short!!". So, it is nothing to be alarmed about and I can move on - they were run but for some reason that was not recorded in the log.

```bash
export PERLBREW_ROOT=/opt/perl5 #perlbrew will be installed in opt
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

/home/0_PROGRAMS/GAAS/annotation/Tools/Maker/maker_check_progress.sh ./Hypoch_input.maker.output/Hypoch_input_master_datastore_index.log
```
42 contigs without STARTED signal and 1 contig without FINISHED signal. strange... It says that there was an issue with contig 53 but it was "well terminated" and does not indicated that it failed permanently (retried).

### Assess
How many genes? On average how long are they?
```bash
cat Hypoch.all.maker.gff | awk '{ if ($3 == "gene") print $0 }' | awk '{ sum += ($5 - $4) } END { print NR, sum / NR }'
```
attribute|number|
---|---|
number genes|46,578|
average length|28,448.6|

Busco completeness
```bash
cd /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker
mkdir busco
cd busco

#get config file. Edit if needed (make sure paths to prerequisite programs are correct. Mt default also has tblastn not forced to single threaded, and augustus species is chicken).
cp /home/0_PROGRAMS/busco/config/config.ini ./Hypoch_busco_config.ini

#set up environment
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/hhome/else/anteater/maker/busco/anteater_busco_config.ini"

#Run BUSCO!
#assess transcripts
time python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i ../Hypoch/Hypoch_input.all.maker.transcripts.fasta  -o tranannotation_eval -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m transcriptome -c 24 -sp chicken -z --augustus_parameters='--progress=true'

#assess proteins
python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i ../Hypoch/Hypoch_input.all.maker.proteins.fasta  -o protannotation_eval -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m proteins -c 24 -sp chicken -z --augustus_parameters='--progress=true'
```
Slightly better than Anteater!

BUSCO run|#genes|total BUSCOs|complete BUSCOs|complete single-copy BUSCOS|complete duplicated BUSCOS|fragmented BUSCOS|missing BUSCOS|run time (s)|
|---|---|---|---|---|---|---|---|---|
transcripts|46,578|94.5|82.4|80.5|1.9|11.9|5.7|50m14|
proteins|46,578|94.5|82.4|80.4|2.0|11.9|5.7|553|
filtered proteins||94.5|82.4|80.4|2.0|11.9|5.7|228|
Willisornis|43,959|93% |C:80.1%|S:78.7%|D:1.4|F:12.9%|M:7.0%|207.7|

AED scores
```bash
perl /home/0_PROGRAMS/Genome_annotation/AED_cdf_generator.pl -b 0.025 Hypoch/Hypoch.all.maker.gff  > AED_cdf_distribution.txt
```
Gene stats
```bash
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PERL5LIB=$PERL5LIB:/hhome/else/.conda/envs/agat/lib/site_perl/5.26.2/

conda activate agat
agat_sp_functional_statistics.pl --gff ./Hypoch/Hypoch.all.maker.noseq.gff -o ./Hypoch/Hypoch_genestats
```

### Hypocnemis Functional annotation
I am only doing this to filter out false positives, I will not actually use the annotations.
```bash
#set up environment
export PATH=$PATH:/home/0_PROGRAMS/interproscan/interproscan-5.39-77.0 #for all of interproscan
export PATH=$PATH:~/tools/maker/bin
#when I am working on the nodes rather than main server:
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

mkdir /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/functional_ann
cd /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/functional_ann

#search for hits in the databases with Interproscan

#Pfam
time interproscan.sh -iprlookup -goterms -appl Pfam-32.0 -i /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch_input.all.maker.proteins.fasta -b Hypoch_prots -f tsv

#PANTHER (took 982m (5683m user time)
time interproscan.sh -iprlookup -goterms -appl PANTHER-14.1 -i /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch_input.all.maker.proteins.fasta -b Hypoch_protsPANTHER -f tsv

#concatenate results
cat Hypoch_prots*.tsv > Hypoch_prots_2interpro.tsv

#add the interpro results to the maker gff
ipr_update_gff /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch.all.maker.gff Hypoch_prots_2interpro.tsv > Hypoch.max.functional_2interpro.gff

#filter to retain only best genes, compare two filtering options:
perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -d Hypoch.max.functional_2interpro.gff > Hypoch.default.functional_2interpro.gff #default build: only proteins with RNA/protein evidence (AED<1)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -s Hypoch.max.functional_2interpro.gff > Hypoch.standard.functional_2interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter_PfamPANTHER.pl -s Hypoch.max.functional_2interpro.gff > Hypoch.standardplus.functional_2interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam or PANTHER)

#count number of genes
grep -cP '\tgene\t' /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch.all.maker.gff #number of genes originally
grep -cP '\tgene\t' Hypoch.max.functional_2interpro.gff #should be same as above
grep -cP '\tgene\t' Hypoch.default.functional_2interpro.gff #number genes with evidence
grep -cP '\tgene\t' Hypoch.standard.functional_2interpro.gff #number genes with evidence OR Pfam domain
grep -cP '\tgene\t' Hypoch.standardplus.functional_2interpro.gff #number genes with evidence OR Pfam domain or PANTHER
less Hypoch_prots.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with Pfam annotations
less Hypoch_protsPANTHER.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with PANTHER annotations
less Hypoch_prots_2interpro.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with annotations

```
**Results**:

Step|# Genes|
---|---|
Raw input|46,578|
No protein evidence|30,122|
Pfam annotated|14,733|
PANTHER annotated|17,166|
Pfam OR PANTHER annotated|17,341|
protein evidence|16,456|
evidence and/or Pfam|17,314|
evidence and/or PANTHER and/or Pfam|18,020|
Without evidence with Pfam|858 2.85%|
With evidence without Pfam|2581 15.68%|

We have 18,020 Hypoch proteins.

Quickly simplify the names
```bash
# create naming table (there are additional options for naming beyond defaults)
maker_map_ids --prefix Hypoch --justify 5 Hypoch.standardplus.functional_2interpro.gff > Hypoch.standardplus.name.map
# replace names in GFF files
map_gff_ids Hypoch.standardplus.name.map Hypoch.standardplus.functional_5interpro.gff
# replace names in FASTA headers
cp /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch_input.all.maker.proteins.fasta .
map_fasta_ids Hypoch.standardplus.name.map Hypoch_input.all.maker.proteins.fasta

#extract the fasta
#Use seqtk to filter a fasta based on a list
cat Hypoch.standardplus.name.map | cut -f 2 | sort | uniq > Hypoch.standardplus_genelist #extract list of genes to keep
/home/0_PROGRAMS/seqtk/seqtk subseq Hypoch_input.all.maker.proteins.fasta Hypoch.standardplus_genelist > Hypoch.standardplus.proteins.fasta

#transcripts too
cp /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/Hypoch/Hypoch_input.all.maker.transcripts.fasta .
map_fasta_ids Hypoch.standardplus.name.map Hypoch_input.all.maker.transcripts.fasta
/home/0_PROGRAMS/seqtk/seqtk subseq Hypoch_input.all.maker.transcripts.fasta Hypoch.standardplus_genelist > Hypoch.standardplus.transcripts.fasta

#check BUSCO
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/hhome/else/anteater/maker/busco/anteater_busco_config.ini"

python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i Hypoch.standardplus.proteins.fasta  -o Hypoch.standardplus.proteins -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m proteins -c 22 -sp chicken -z --augustus_parameters='--progress=true'
```

Time to add it to OrthoFinder!

```bash
#load environment
export PATH=$PATH:/home/0_PROGRAMS/fastme-2.1.5/binaries
export PATH=$PATH:/home/0_PROGRAMS/ #for Diamond
export PATH=$PATH:/home/0_PROGRAMS/OrthoFinder

cd /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder
mkdir temp_proteomes_to_add
cp /home/0_GENOMES1/0_Genetic_resources/temp_genomes/Hypoch_maker/functional_ann/Hypoch.standardplus.proteins.fasta ./temp_proteomes_to_add

time orthofinder -b /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Jan20/WorkingDirectory -f /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/temp_proteomes_to_add -t 24 > Orthofinder_PasserinePlusplusHypoch.log

```
**timing**: orthofinder took 63m49 on 24 threads (365m user time).

# Rhegmatorhina
Now for Rhegmatorhina! I repeated the exact same process as above, but for the genome published for Rhegmatorhina.
It waited for a few weeks in the queue on Niagara, but finally ran!

```bash
mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina
scp -r elsemikk@niagara.scinet.utoronto.ca:/path/to/folder/0_MAKER/Rhemel .
mv Rhemel maker
rm -r Maker/garbage
cd maker
```

```bash
#check if there are any that failed and count how many were skipped vs finished
grep -c "SKIPPED_SMALL" ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log 
grep -c "STARTED" ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log 
grep -c "FINISHED" ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log 
grep -c "DIED_SKIPPED_PERMANENT" ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log #none! That is good
wc -l ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log #just counting lines to make sure the preceeding numbers add up

#make sure that all of the contigs were run
grep -c "^>" Rhemel_input.fasta #Count contigs of original fasta file
cat Rhemel_input_0*.fasta | grep -c "^>" #Count input contigs of the many chunks - should add to the same number as the original above
cut -f 1 ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log | sort | uniq | wc -l #count number of unique contigs in the log (contigs that actually ran in the maker pipeline)

grep "^>" Rhemel_input.fasta | sed 's/>//g' | sed 's/ .*$//g' | sort | uniq | wc -l
cut -f 1 ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log | sort | uniq
#If all looks good:

#Create files merging all data into a fasta and GFF3 file
time gff3_merge -s -d ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log> Rhemel.all.maker.gff
time fasta_merge -d ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log
#Create a gff3 file without sequences
time gff3_merge -n -s -d ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log > Rhemel.all.maker.noseq.gff
#done
```
Type|number (incomplete round)|number (finished)|
---|---|
Total scaffolds|619|619|
run scaffolds|618|619|
Skipped_small|0|0|
started|609|610|
finished|618|619|
died|0|0|
lines|1229|1231|

In the first round, you can see there is a missing contig - only 618 run but the genome assembly has 619 contigs! That was contig 0 and it was because the code to run them started the count at 1. oops, that costed a few weeks of waiting in the queue again for the final contig to run, but it eventually did rerun.

```bash
export PERLBREW_ROOT=/opt/perl5 #perlbrew will be installed in opt
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

/home/0_PROGRAMS/GAAS/annotation/Tools/Maker/maker_check_progress.sh ./Rhemel_input.maker.output/Rhemel_input_master_datastore_index.log
```
9 contigs without STARTED signal. Not a problem.

### Assess
How many genes? On average how long are they?
```bash
cat Rhemel.all.maker.gff | awk '{ if ($3 == "gene") print $0 }' | awk '{ sum += ($5 - $4) } END { print NR, sum / NR }'
```

* # genes: 41005 
* average length: 31441.5

Busco completeness
```bash
cd ..
mkdir busco
cd busco

#get config file. Edit if needed (make sure paths to prerequisite programs are correct. Mt default also has tblastn not forced to single threaded, and augustus species is chicken).
# mine I am reusing from anteater

#set up environment
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/hhome/else/anteater/maker/busco/anteater_busco_config.ini"

#Run BUSCO!
#assess transcripts
time python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i ../Rhemel/Rhemel_input.all.maker.transcripts.fasta  -o tranannotation_eval -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m transcriptome -c 24 -sp chicken -z --augustus_parameters='--progress=true'

#assess proteins
python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i ../Rhemel/Rhemel_input.all.maker.proteins.fasta  -o protannotation_eval -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m proteins -c 24 -sp chicken -z --augustus_parameters='--progress=true'
```
Transcripts: C:81.1%[S:79.7%,D:1.4%],F:10.6%,M:8.3%,n:4915
Proteins: C:81.0%[S:79.5%,D:1.5%],F:10.8%,M:8.2%,n:4915

AED scores
```bash
perl /home/0_PROGRAMS/Genome_annotation/AED_cdf_generator.pl -b 0.025 maker/Rhemel_input.all.maker.gff  > AED_cdf_distribution.txt
```
Gene stats
```bash
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PERL5LIB=$PERL5LIB:/hhome/else/.conda/envs/agat/lib/site_perl/5.26.2/

conda activate agat
agat_sp_functional_statistics.pl --gff ./maker/Rhemel_input.all.maker.noseq.gff -o ./maker/Rhemel_genestats
```

### Functional annotation
I am only doing this to filter out false positives
```bash
#set up environment
export PATH=$PATH:/home/0_PROGRAMS/interproscan/interproscan-5.39-77.0 #for all of interproscan
export PATH=$PATH:~/tools/maker/bin
#when I am working on the nodes rather than main server:
export PERLBREW_ROOT=/opt/perl5
/opt/perl5/bin/perlbrew switch perl-5.30.0 #A sub-shell is launched with perl-5.30.0 as the activated perl. Run 'exit' to finish it.
export PATH=/opt/tools/maker/bin:$PATH

mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/functional_ann
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/functional_ann

#search for hits in the databases with Interproscan

#Pfam
interproscan.sh -iprlookup -goterms -appl Pfam-32.0 -i /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/Rhemel/Rhemel_input.all.maker.proteins.fasta -b Rhemel_prots -f tsv
#PANTHER
time interproscan.sh -iprlookup -goterms -appl PANTHER-14.1 -i /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/Rhemel/Rhemel_input.all.maker.proteins.fasta -b Rhemel_protsPANTHER -f tsv

#concatenate results
cat Rhemel_prots*.tsv > Rhemel_prots_2interpro.tsv

#add the interpro results to the maker gff
ipr_update_gff /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/Rhemel/Rhemel.all.maker.gff Rhemel_prots_2interpro.tsv > Rhemel.max.functional_2interpro.gff

#filter to retain only best genes, compare two filtering options:
perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -d Rhemel.max.functional_2interpro.gff > Rhemel.default.functional_2interpro.gff #default build: only proteins with RNA/protein evidence (AED<1)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter.pl -s Rhemel.max.functional_2interpro.gff > Rhemel.standard.functional_2interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam)

perl /home/0_PROGRAMS/Genome_annotation/quality_filter_PfamPANTHER.pl -s Rhemel.max.functional_2interpro.gff > Rhemel.standardplus.functional_2interpro.gff #standard build: all proteins with evidence (AED<1) and/or a functional annotation (Pfam or PANTHER)

#count number of genes
grep -cP '\tgene\t' ../Rhemel/Rhemel.all.maker.gff #number of genes originally
grep -cP '\tgene\t' Rhemel.default.functional_2interpro.gff #number genes with evidence
grep -cP '\tgene\t' Rhemel.standard.functional_2interpro.gff #number genes with evidence OR Pfam domain
grep -cP '\tgene\t' Rhemel.standardplus.functional_2interpro.gff #number genes with evidence OR Pfam domain or PANTHER
less Rhemel_prots.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with Pfam annotations
less Rhemel_protsPANTHER.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with PANTHER annotations
less Rhemel_prots_2interpro.tsv | cut -f 1 | sort | uniq | wc -l #number of genes with annotations

```
**Timing**: Interproscan took 551m35 to do Panther.

Step|# Genes|
---|---|
Raw input|41,005|
No protein evidence|26,782|
Pfam annotated|12,607|
PANTHER annotated|14,459|
Pfam OR PANTHER annotated|14,596|
protein evidence|14,223|
evidence and/or Pfam|14,726|
evidence and/or PANTHER and/or Pfam|15,117|
Without evidence with Pfam|503 1.87%|
With evidence without Pfam|2119 14.90%|

We have 15,117 Rhemel proteins. That is fewer than anteater.

Quickly simplify the names
```bash
# create naming table (there are additional options for naming beyond defaults)
maker_map_ids --prefix Rhemel --justify 5 Rhemel.standardplus.functional_2interpro.gff > Rhemel.standardplus.name.map
# replace names in GFF files
map_gff_ids Rhemel.standardplus.name.map Rhemel.standardplus.functional_2interpro.gff
# replace names in FASTA headers
cp ../Rhemel/Rhemel_input.all.maker.proteins.fasta .
map_fasta_ids Rhemel.standardplus.name.map Rhemel_input.all.maker.proteins.fasta #overwrites input, careful!

#extract the fasta
#Use seqtk to filter a fasta based on a list
cat Rhemel.standardplus.name.map | cut -f 2 | sort | uniq > Rhemel.standardplus_genelist #extract list of genes to keep
/home/0_PROGRAMS/seqtk/seqtk subseq Rhemel_input.all.maker.proteins.fasta Rhemel.standardplus_genelist > Rhemel.standardplus.proteins.fasta

#transcripts too
cp ../Rhemel/Rhemel_input.all.maker.transcripts.fasta .
map_fasta_ids Rhemel.standardplus.name.map Rhemel_input.all.maker.transcripts.fasta
/home/0_PROGRAMS/seqtk/seqtk subseq Rhemel_input.all.maker.transcripts.fasta Rhemel.standardplus_genelist > Rhemel.standardplus.transcripts.fasta

#check BUSCO
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/hhome/else/anteater/maker/busco/anteater_busco_config.ini"

python /home/0_PROGRAMS/busco_v3/scripts/run_BUSCO.py -i Rhemel.standardplus.proteins.fasta  -o Rhemel.standardplus.proteins -l /home/0_PROGRAMS/busco_v3/aves_odb9 -m proteins -c 22 -sp chicken -z --augustus_parameters='--progress=true'
```
8.2% missing, exactly the same as before filtering!

Time to add it to OrthoFinder!

```bash
#load environment
export PATH=$PATH:/home/0_PROGRAMS/fastme-2.1.5/binaries
export PATH=$PATH:/home/0_PROGRAMS/ #for Diamond
export PATH=$PATH:/home/0_PROGRAMS/OrthoFinder

cd /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/temp_proteomes_to_add
cp /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/anteater/Rhegmatorhina/functional_ann/Rhemel.standardplus.proteins.fasta .

ulimit -n 6144

time orthofinder -b /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/ -f /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/temp_proteomes_to_add -t 24 -a 6 > Orthofinder_PasserinePlusplusNPRhemel.log
```
**timing**: orthofinder took 249m on 24 threads (1867m user time), and then crashed due to IOerror (too many open files) even though I had set ulimit to 6144! It is fine, it crashed at the resolving gene trees part, but all that I wanted was the orthogroup assignment.