# Gene family expansions and contractions with Café

There is a very good tutorial for café [here](https://iu.app.box.com/v/cafetutorial-pdf) which I have loosely followed with modifications (for example I am using OrthoFinder instead of mcl).

### Step 1: Obtain Proteome Data
I have obtained proteomes in two places: Ensembl, and Refseq. Ensembl is partly based on Refseq as well as taking evidence from additional sources so I will give priority to Ensembl proteomes and then take Refseq proteomes from any additional species not covered.  
In this dataset I have used Ensembl release 99.

The details of how I downloaded and filtered the datasets can be found in the positive selection page of this repository. Briefly, I filtered the proteomes to retain only the longest isoform of each gene.

In addition to Ensembl 99, I added other species available through NCBI's RefSeq (species list available [here](https://www.ncbi.nlm.nih.gov/genome/annotation_euk/all/), comparison of Refseq and Ensembl [here](https://m.ensembl.org/Help/Faq?id=294). This is a little more work to get the useable data because while the Ensembl data has information on the gene and the sequence length of each protein, the Refseq data does not necessarily have the gene IDs in the FASTA headers, only the transcript IDs which are different for each protein isoform. We do not want redundant isoforms for the same protein in our dataset as this will severely affect the downstream analysis! We need to also download a gff file which has the information on what gene each transcript belongs to, and then we will use that to keep only a single (longest) isoform per transcript. I found a script to do this [here](https://www.biostars.org/p/388936/) by user SMK (I just had to change a piece of the code from `.+` to `*+` to deal with the fact that the format is slightly different from the format the code was written for (fewer fields in the FASTA header).
The filtering of the Refseq data again is detailed in the dN/dS positive selection page.

 
Quality check: how complete are they? "Completeness" can be measured using BUSCO.
```bash
#Ensembl proteomes
cd /home/0_GENOMES1/0_Genetic_resources/Ensembl_99
grep -c "^>" longest.simpleheader*.AA.fa #beware this is not the number of genes, as it likely includes multiple isoforms

#get config file. Edit if needed (make sure paths to prerequisite programs are correct. Mt default also has tblastn not forced to single threaded, and augustus species is chicken).
cp /home/0_PROGRAMS/busco/config/config.ini ./busco_config.ini

#set up environment
conda activate Augustus #has augustus installed
export BUSCO_CONFIG_FILE="/home/0_GENOMES1/0_Genetic_resources/Ensembl_99/busco_config.ini"

cat Ensembl99_cieslist.txt | parallel python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i longest.simpleheader{1}.AA.fa  -o {1}_proteome_eval -l /home/0_PROGRAMS/busco/aves_odb9 -m proteins -c 1 -sp chicken -z --augustus_parameters='--progress=true'

#Refseq proteomes
cd /home/0_GENOMES1/0_Genetic_resources/Refseq
cp /home/0_PROGRAMS/busco/config/config.ini ./busco_config.ini
export BUSCO_CONFIG_FILE="/home/0_GENOMES1/0_Genetic_resources/Refseq/busco_config.ini"

cat RefnotEnsPasserines_list.txt | parallel python /home/0_PROGRAMS/busco/scripts/run_BUSCO.py -i longest.simpleheader{1}.AA.fa  -o {1}_proteome_eval -l /home/0_PROGRAMS/busco/aves_odb9 -m proteins -c 1 -sp chicken -z --augustus_parameters='--progress=true'
```
They look fine. Accnis has maybe too many proteins - but I am not including it anyways as it is not in my group of interest.


# Step 2: Assign orthologs

I have previously identified ortholog/paralog groups between proteomes from all the Passerines/Falconiformes/Psittaciformes on Ensembl 99 and Refseq (Jan 2020) using OrthoFinder, so I will be reusing that dataset. The details of how this was done are in the page describing dN/dS calculation to test for positive selection.

I just need to remove some taxa which are currently under embargo, so I will not include them in this analysis.
```bash
#remove unwanted taxa from taxon list of orthofinder run
nano /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/OrthoFinder/Results_Feb15/WorkingDirectory/SpeciesIDs.txt
#hashtag out 25 (Strhab) and 30-85 (nonpasserines)

#load environment
export PATH=$PATH:/home/0_PROGRAMS/fastme-2.1.5/binaries
export PATH=$PATH:/home/0_PROGRAMS/ #for Diamond
export PATH=$PATH:/home/0_PROGRAMS/OrthoFinder

#rerun orthofinder after specifying taxa to remove
cd /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/OrthoFinder/Results_Feb15/WorkingDirectory/
cp ../../../*.fa .
cp ../../../Blast* .
time orthofinder -b . -t 24 > Orthofinder_PasserinePlus.log
```

# Step 3: filter and parse for Café
Next, we will parse the output to use as input to Café. We need a tab-delimited file where column 1 is a gene description (or blank), column 2 is a gene ID (or blank), and the following columns are the number gene in that gene family in that species, and the header is the species name. We have a file like this, `Orthogroups/Orthogroups.GeneCount.tsv` except the first column is missing and the last column (total) should not be present.

(If we used mcl to cluster instead of OrthoFinder, we could use a script like this: `python /home/0_PROGRAMS/CAFE/cafe_tutorial/python_scripts/cafetutorial_mcl2rawcafe.py -i dump.Ensembl_blast_output.mci.I30 -o unfiltered_cafe_input.txt -sp "ENSG00 ENSPTR ENSPPY ENSPAN ENSNLE ENSMMU ENSCJA ENSRNO ENSMUS ENSFCA ENSECA ENSBTA"` )

Luckily the OrthoFinder output includes an OrthoMCL output format.
Then, we will filter a bit. The filter will retain only families found in at least 2 taxa (to be phylogenetically informative as to ancestral states), and will split the file to separate out proteins that have more than 100 copies in an individual, since this may skew the results. 

Note I would not necessarily recommend the weird way I put together the large vs small family files but at least it worked.
```bash
#set up environment
mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe
 
#get genecounts file from orthofinder
cp /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/OrthoFinder/Results_Feb15/WorkingDirectory/OrthoFinder/Results_Mar26_2/Orthogroups/Orthogroups.GeneCount.tsv .

#First, remove the last column | then add a null first column, output is tab-delimited 
awk 'BEGIN{OFS="\t"}NF{NF-=1};1' Orthogroups.GeneCount.tsv | awk 'BEGIN{OFS="\t"}{print "(null)\t"$0}' > Cafe_input.tsv

#simplify and standardize my species codes
sed -i 's/longest\.simpleheader\.//g' Cafe_input.tsv
sed -i 's/\.AA//g' Cafe_input.tsv
sed -i 's/Willisornis/Wilpoe/g' Cafe_input.tsv
sed -i 's/.standardplus.proteins//g' Cafe_input.tsv
sed -i 's/\tOrthogroup/ID\tOrthogroup/' Cafe_input.tsv

#remove redundant species, and a few that are under embargo (not allowed to analyze)
#cut -f 28,40,61,62,73 Cafe_input.tsv | head #verify which columns will be deleted
#cut -f 1-27,29-39,41-60,63-72,74-91 Cafe_input.tsv > temp && mv temp Cafe_input.tsv #removed duplicate Galgal and Campar, and embargoed Calana, Strhab, and Aquchr
#I wish I could use all of the bird phylogeny, but the program just keeps crashing, so i reduced it to only this passerines + parrots dataset
cut -f 1-27,29-32,89 Cafe_input.tsv > temp && mv temp Cafe_input.tsv 

#temporarily remove the headers for the following code to work
tail -n +2 Cafe_input.tsv > temp_filterinput
#filter to separate gene families with more than 100 copies in any species
conda activate python2
python /home/0_PROGRAMS/CAFE-4.2.1/cafe_tutorial/python_scripts/cafetutorial_clade_and_size_filter.py -i temp_filterinput -o filtered_Cafe_input.tsv -s 

#put the headers back and deal with the fact it thought the first orthogroup was the header
grep OG0000000 filtered_Cafe_input.tsv > temp_OG0
grep -v OG0000000 filtered_Cafe_input.tsv > temp_smallfams
grep -v OG0000000 large_filtered_Cafe_input.tsv > temp_largefams
head -n 1 Cafe_input.tsv > temp_header
cat temp_header temp_smallfams > filtered_Cafe_input.tsv
cat temp_header temp_OG0 temp_largefams > large_filtered_Cafe_input.tsv
rm temp_OG0 temp_smallfams temp_largefams temp_header temp_filterinput
```
* `-sp`: this is a list of clades of interest. Each clade is separated by a space
* `-s`: size cutoff selection, will separate genes found in less than or more than 100 copies.

I included all of the genomes annotations that were available on Ensembl/Refseq (at the time), but some of these are under embargo. I removed these before I analyzed the dataset.
* Aquila chrysaetos chrysaetos (European golden eagle) genome assembly bAquChr1 (column 41)
* Calypte anna (Anna's hummingbird) genome sequencing and assembly, primary haplotype, v1 (column 62)
* Strigops habroptila (Kakapo) Genome sequencing and assembly, primary haplotype, v1 (column 29)

### Step 4: Get a species tree
There has already beeen a robust, time-calibrated tree created for passeriformes, likely much more accurate than a tree I could make quickly. My goal for this project is not to reinvent the Passeriform phylogeny, so I will just use this previously-published tree to use an input to Café. A currently recent passerine phylogeny was made by Oliveros et al. 2019.  
However, it does not include data for all of the taxa I am using so I need to extend towards other sources as well. In order to be consistent, I will just use the dates reported by Jetz et al. 

**Note**: ideally I would try to use a phylogeny based on the whole genome sequences, however I cannot do this since the manakin genomes are currently quite new and under data embargo for phylogenetic analysis.  

**specifications**: It must be ultrametric, and have branch lengths, and be in Newick format.

```bash
cat > Psittacopasserae.tree
```
Here is with nonpasserines:

I will just use the Jetz et al. ages for consistency. I wrote it out in Newick format with the relevant taxa.
```
(((Tinamus guttatus:73.8,Nothoprocta perdicaria:73.8):12.1,(Struthio camelus:61,(Dromaius novaehollandiae:42.5,((Apteryx mantelli mantelli:5.35,Apteryx rowi:5.35):4.32,(Apteryx haastii:3.42,Apteryx owenii:3.42):6.25):32.83):18.5):24.9):27.3,(((Anas platyrhynchos:24.1,(Anser cygnoides:0.9669,Anser brachyrhynchus:0.9669):23.1331):48.3,(Numida meleagris:27.3,(Coturnix japonica:18.8,((Gallus gallus:17,Pavo cristatus:17):1.1,(Meleagris gallopavo:14.6,(Phasianus colchicus:7.35,Chrysolophus pictus:7.35):7.25):3.5):0.7):8.5):45.1):25.2,((Eurypyga helias:77.1,(Antrostomus carolinensis:73,(Calypte anna:58.1,Chaetura pelagica:58.1):14.9):4.1):2.5,((Phaethon lepturus:74.8,(((Columba livia:64.8,Mesitornis unicolor:64.8):3.5,(Pterocles gutturalis:64.3,Opisthocomus hoazin:64.3):4):6.2,((Balearica regulorum:69.9,(Chlamydotis macqueenii:67.6,Cuculus canorus:67.6):2.3):1.6,(Tauraco erythrolophus:66.5,(Gavia stellata:65.8,(((Pelecanus crispus:59.2,(Nipponia nippon:51.6,Egretta garzetta:51.6):7.6):0.6,Phalacrocorax carbo:59.8):4.0,(Fulmarus glacialis:63.2,(Aptenodytes forsteri:25.4,Pygoscelis adeliae:25.4):37.8):0.6):2.0):0.7):5):3):0.3):4.8,((Charadrius vociferus:67.9,(Calidris pugnax:21.5,Calidris pygmaea:21.5):46.4):11.6,((((Leptosomus discolor:75.8,(Apaloderma vittatum:72.1,(Buceros rhinoceros:70.7,(Merops nubicus:69.6,Picoides pubescens:69.6):1.1):1.4):3.7):0.1,(Tyto alba:69.5,Athene cunicularia:69.5):6.4):1.8,(Accipiter nisus:26.9,Aquila chrysaetos:26.9):50.8):0.7,(Colius striatus:76.4,(Cariama cristata:76.2,((Falco peregrinus:3.85,Falco cherug:3.85):71.75,((Nestor notabilis:55.6,(Amazona collaria:30.4,Melopsittacus undulatus:30.4):25.2):19.2,(Acanthisitta chloris:66.8,((((Willisornis poecilinotus:8.41,Rhegmatorhina melanosticta:8.41):2.69,Hypocnemis ochrogyna:11.1):35.9,(((((Lepidothrix coronata:6.1,Manacus vitellinus:6.1):0.7,Pipra filicauda:6.8):3.3,Corapipo altera:10.1):3.0,Neopelma chrysocephalum:13.1):13.2,Empidonax trailli:26.3):20.7):17.6,((Corvus cornix:4.17,Corvus brachyrhynchos:4.17):45.93,((Pseudopodoces humilis:23.27,(Parus major:19,Cyanistes caeruleus:19):4.27):22.33,((Ficedula albicollis:28.1,Sturnus vulgaris:28.1):17,(((Taeniopygia guttata:11.9,Lonchura striata:11.9):0.9,Erythrura gouldiae:12.8):20.7,(Serinus canaria:26,((Camarhynchus parvulus:4.47,Geospiza fortis:4.47):17.03,(Zonotrichia albicollis:5.28,Junco hyemalis:5.28):16.22):4.5):7.5):11.6):0.5):4.5):14.5):2.2):8):0.8):0.6):0.2):2.0):1.1):0.1):0):18):15.6);
```

Here I have removed the embargoed ones:
```
(((Tinamus guttatus:73.8,Nothoprocta perdicaria:73.8):12.1,(Struthio camelus:61,(Dromaius novaehollandiae:42.5,((Apteryx mantelli mantelli:5.35,Apteryx rowi:5.35):4.32,(Apteryx haastii:3.42,Apteryx owenii:3.42):6.25):32.83):18.5):24.9):27.3,(((Anas platyrhynchos:24.1,(Anser cygnoides:0.9669,Anser brachyrhynchus:0.9669):23.1331):48.3,(Numida meleagris:27.3,(Coturnix japonica:18.8,((Gallus gallus:17,Pavo cristatus:17):1.1,(Meleagris gallopavo:14.6,(Phasianus colchicus:7.35,Chrysolophus pictus:7.35):7.25):3.5):0.7):8.5):45.1):25.2,((Eurypyga helias:77.1,(Antrostomus carolinensis:73,Chaetura pelagica:73):4.1):2.5,((Phaethon lepturus:74.8,(((Columba livia:64.8,Mesitornis unicolor:64.8):3.5,(Pterocles gutturalis:64.3,Opisthocomus hoazin:64.3):4):6.2,((Balearica regulorum:69.9,(Chlamydotis macqueenii:67.6,Cuculus canorus:67.6):2.3):1.6,(Tauraco erythrolophus:66.5,(Gavia stellata:65.8,(((Pelecanus crispus:59.2,(Nipponia nippon:51.6,Egretta garzetta:51.6):7.6):0.6,Phalacrocorax carbo:59.8):4.0,(Fulmarus glacialis:63.2,(Aptenodytes forsteri:25.4,Pygoscelis adeliae:25.4):37.8):0.6):2.0):0.7):5):3):0.3):4.8,((Charadrius vociferus:67.9,(Calidris pugnax:21.5,Calidris pygmaea:21.5):46.4):11.6,((((Leptosomus discolor:75.8,(Apaloderma vittatum:72.1,(Buceros rhinoceros:70.7,(Merops nubicus:69.6,Picoides pubescens:69.6):1.1):1.4):3.7):0.1,(Tyto alba:69.5,Athene cunicularia:69.5):6.4):1.8,Accipiter nisus:77.7):0.7,(Colius striatus:76.4,(Cariama cristata:76.2,((Falco peregrinus:3.85,Falco cherug:3.85):71.75,((Nestor notabilis:55.6,(Amazona collaria:30.4,Melopsittacus undulatus:30.4):25.2):19.2,(Acanthisitta chloris:66.8,((((Willisornis poecilinotus:8.41,Rhegmatorhina melanosticta:8.41):2.69,Hypocnemis ochrogyna:11.1):35.9,(((((Lepidothrix coronata:6.1,Manacus vitellinus:6.1):0.7,Pipra filicauda:6.8):3.3,Corapipo altera:10.1):3.0,Neopelma chrysocephalum:13.1):13.2,Empidonax trailli:26.3):20.7):17.6,((Corvus cornix:4.17,Corvus brachyrhynchos:4.17):45.93,((Pseudopodoces humilis:23.27,(Parus major:19,Cyanistes caeruleus:19):4.27):22.33,((Ficedula albicollis:28.1,Sturnus vulgaris:28.1):17,(((Taeniopygia guttata:11.9,Lonchura striata:11.9):0.9,Erythrura gouldiae:12.8):20.7,(Serinus canaria:26,((Camarhynchus parvulus:4.47,Geospiza fortis:4.47):17.03,(Zonotrichia albicollis:5.28,Junco hyemalis:5.28):16.22):4.5):7.5):11.6):0.5):4.5):14.5):2.2):8):0.8):0.6):0.2):2.0):1.1):0.1):0):18):15.6);
```
Here is just the Passerines/parrots
```
((Nesnot:55.6,(Amacol:30.4,Melund:30.4):25.2):19.2,(Acachl:66.8,((((Wilpoe:8.41,Rhemel:8.41):2.69,Hypoch:11.1):35.9,(((((Lepcor:6.1,Manvit:6.1):0.7,Pipfil:6.8):3.3,Coralt:10.1):3.0,Neochr:13.1):13.2,Emptra:26.3):20.7):17.6,((Corcor:4.17,Corbra:4.17):45.93,((Psehum:23.27,(Parmaj:19,Cyacae:19):4.27):22.33,((Ficalb:28.1,Stuvul:28.1):17,(((Taegut:11.9,Lonstr:11.9):0.9,Erygou:12.8):20.7,(Sercan:26,((Campar:4.47,Geofor:4.47):17.03,(Zonalb:5.28,Junhie:5.28):16.22):4.5):7.5):11.6):0.5):4.5):14.5):2.2):8);
```

Translate to codes:
```bash
sed -i 's/Tinamus guttatus/Tingut/g' Psittacopasserae.tree
sed -i 's/Nothoprocta perdicaria/Notper/g' Psittacopasserae.tree
sed -i 's/Anser cygnoides/Anscyg/g' Psittacopasserae.tree
sed -i 's/Coturnix japonica/Cotjap/g' Psittacopasserae.tree
sed -i 's/Meleagris gallopavo/Melgal/g' Psittacopasserae.tree
sed -i 's/Charadrius vociferus/Chavoc/g' Psittacopasserae.tree
sed -i 's/Leptosomus discolor/Lepdis/g' Psittacopasserae.tree
sed -i 's/Merops nubicus/Mernub/g' Psittacopasserae.tree
sed -i 's/Athene cunicularia/Athcun/g' Psittacopasserae.tree
sed -i 's/Colius striatus/Colstr/g' Psittacopasserae.tree
sed -i 's/Falco cherug/Falche/g' Psittacopasserae.tree
sed -i 's/Amazona collaria/Amacol/g' Psittacopasserae.tree
sed -i 's/Willisornis poecilinotus/Wilpoe/g' Psittacopasserae.tree
sed -i 's/Manacus vitellinus/Manvit/g' Psittacopasserae.tree
sed -i 's/Neopelma chrysocephalum/Neochr/g' Psittacopasserae.tree
sed -i 's/Corvus brachyrhynchos/Corbra/g' Psittacopasserae.tree
sed -i 's/Cyanistes caeruleus/Cyacae/g' Psittacopasserae.tree
sed -i 's/Taeniopygia guttata/Taegut/g' Psittacopasserae.tree
sed -i 's/Serinus canaria/Sercan/g' Psittacopasserae.tree
sed -i 's/Zonotrichia albicollis/Zonalb/g' Psittacopasserae.tree
sed -i 's/Struthio camelus/Strcam/g' Psittacopasserae.tree
sed -i 's/Dromaius novaehollandiae/Dronov/g' Psittacopasserae.tree
sed -i 's/Apteryx haastii/Apthaa/g' Psittacopasserae.tree
sed -i 's/Anser brachyrhynchus/Ansbra/g' Psittacopasserae.tree
sed -i 's/Gallus gallus/Galgal/g' Psittacopasserae.tree
sed -i 's/Phasianus colchicus/Phacol/g' Psittacopasserae.tree
sed -i 's/Calidris pugnax/Calpug/g' Psittacopasserae.tree
sed -i 's/Apaloderma vittatum/Apavit/g' Psittacopasserae.tree
sed -i 's/Picoides pubescens/Picpub/g' Psittacopasserae.tree
sed -i 's/Accipiter nisus/Accnis/g' Psittacopasserae.tree
sed -i 's/Cariama cristata/Carcri/g' Psittacopasserae.tree
sed -i 's/Strigops habroptila/Strhab/g' Psittacopasserae.tree
sed -i 's/Melopsittacus undulatus/Melund/g' Psittacopasserae.tree
sed -i 's/Hypocnemis ochrogyna/Hypoch/g' Psittacopasserae.tree
sed -i 's/Pipra filicauda/Pipfil/g' Psittacopasserae.tree
sed -i 's/Empidonax trailli/Emptra/g' Psittacopasserae.tree
sed -i 's/Pseudopodoces humilis/Psehum/g' Psittacopasserae.tree
sed -i 's/Ficedula albicollis/Ficalb/g' Psittacopasserae.tree
sed -i 's/Lonchura striata/Lonstr/g' Psittacopasserae.tree
sed -i 's/Camarhynchus parvulus/Campar/g' Psittacopasserae.tree
sed -i 's/Junco hyemalis/Junhie/g' Psittacopasserae.tree
sed -i 's/Apteryx mantelli mantelli/Aptaus/g' Psittacopasserae.tree
sed -i 's/Apteryx owenii/Aptowe/g' Psittacopasserae.tree
sed -i 's/Numida meleagris/Nummel/g' Psittacopasserae.tree
sed -i 's/Pavo cristatus/Pavcri/g' Psittacopasserae.tree
sed -i 's/Chrysolophus pictus/Chrpic/g' Psittacopasserae.tree
sed -i 's/Calidris pygmaea/Calpyg/g' Psittacopasserae.tree
sed -i 's/Buceros rhinoceros/Bucrhi/g' Psittacopasserae.tree
sed -i 's/Tyto alba/Tytalb/g' Psittacopasserae.tree
sed -i 's/Aquila chrysaetos/Aquchr/g' Psittacopasserae.tree
sed -i 's/Falco peregrinus/Falper/g' Psittacopasserae.tree
sed -i 's/Nestor notabilis/Nesnot/g' Psittacopasserae.tree
sed -i 's/Acanthisitta chloris/Acachl/g' Psittacopasserae.tree
sed -i 's/Lepidothrix coronata/Lepcor/g' Psittacopasserae.tree
sed -i 's/Corapipo altera/Coralt/g' Psittacopasserae.tree
sed -i 's/Corvus cornix/Corcor/g' Psittacopasserae.tree
sed -i 's/Parus major/Parmaj/g' Psittacopasserae.tree
sed -i 's/Sturnus vulgaris/Stuvul/g' Psittacopasserae.tree
sed -i 's/Erythrura gouldiae/Erygou/g' Psittacopasserae.tree
sed -i 's/Geospiza fortis/Geofor/g' Psittacopasserae.tree
sed -i 's/Apteryx rowi/Aptrow/g' Psittacopasserae.tree
sed -i 's/Anas platyrhynchos/Anapla/g' Psittacopasserae.tree
sed -i 's/Tauraco erythrolophus/Tauery/g' Psittacopasserae.tree
sed -i 's/Chlamydotis macqueenii/Chlmac/g' Psittacopasserae.tree
sed -i 's/Cuculus canorus/Cuccan/g' Psittacopasserae.tree
sed -i 's/Gavia stellata/Gavste/g' Psittacopasserae.tree
sed -i 's/Pelecanus crispus/Pelcri/g' Psittacopasserae.tree
sed -i 's/Nipponia nippon/Nipnip/g' Psittacopasserae.tree
sed -i 's/Columba livia/Colliv/g' Psittacopasserae.tree
sed -i 's/Calypte anna/Calana/g' Psittacopasserae.tree
sed -i 's/Antrostomus carolinensis/Antcar/g' Psittacopasserae.tree
sed -i 's/Eurypyga helias/Eurhel/g' Psittacopasserae.tree
sed -i 's/Egretta garzetta/Egrgar/g' Psittacopasserae.tree
sed -i 's/Phalacrocorax carbo/Phacar/g' Psittacopasserae.tree
sed -i 's/Fulmarus glacialis/Fulgla/g' Psittacopasserae.tree
sed -i 's/Aptenodytes forsteri/Aptfor/g' Psittacopasserae.tree
sed -i 's/Pygoscelis adeliae/Pygade/g' Psittacopasserae.tree
sed -i 's/Rhegmatorhina melanosticta/Rhemel/g' Psittacopasserae.tree
sed -i 's/Balearica regulorum/Balreg/g' Psittacopasserae.tree
sed -i 's/Opisthocomus hoazin/Opihoa/g' Psittacopasserae.tree
sed -i 's/Pterocles gutturalis/Ptegut/g' Psittacopasserae.tree
sed -i 's/Mesitornis unicolor/Mesuni/g' Psittacopasserae.tree
sed -i 's/Phaethon lepturus/Phalep/g' Psittacopasserae.tree
sed -i 's/Chaetura pelagica/Chapel/g' Psittacopasserae.tree
```
Result:
```
((Nesnot:55.6,(Amacol:30.4,Melund:30.4):25.2):19.2,(Acachl:66.8,((((Wilpoe:8.41,Rhemel:8.41):2.69,Hypoch:11.1):35.9,(((((Lepcor:6.1,Manvit:6.1):0.7,Pipfil:6.8):3.3,Coralt:10.1):3.0,Neochr:13.1):13.2,Emptra:26.3):20.7):17.6,((Corcor:4.17,Corbra:4.17):45.93,((Psehum:23.27,(Parmaj:19,Cyacae:19):4.27):22.33,(Ficalb:45.1,(((Taegut:11.9,Lonstr:11.9):0.9,Erygou:12.8):20.7,(Sercan:26,((Campar:4.47,Geofor:4.47):17.03,(Zonalb:5.28,Junhie:5.28):16.22):4.5):7.5):11.6):0.5):4.5):14.5):2.2):8);
```
### Step 5: Coffee Time
First, we estimate a single birth-death lambda for the tree, which we must provide to cafe. Then, we will summarize the results, and then plot them on a phylogenetic tree to visualize it.

Firstly, some scripts in the café pipeline were written for python 2 so I am making myself a python 2.7 environment
```bash
conda create --name python2717 python=2.7.17
conda activate python2717
conda install matplotlib
pip install matplotlib
conda install matplotlib=2.0.2
```
Here is the lamdas_tree file:
```bash
cat > lamdas_tree
```
```
((Falper:1,Falche:1):1,(((Strhab:1,Nesnot:1):1,(Amacol:1,Melund:1):1):1,(Acachl:1,((Wilpoe:2,(((((Lepcor:2,Manvit:2):2,Pipfil:2):2,Coralt:2):2,Neochr:2):2,Emptra:2):2):2,((Corcor:3,Corbra:3):3,((Psehum:3,(Parmaj:3,Cyacae:3):3):3,((Ficalb:3,Stuvul:3):3,(((Taegut:3,Lonstr:3):3,Erygou:3):3,(Sercan:3,((Campar:3,Geofor:3):3,(Zonalb:3,Junhie:3):3):3):3):3):3):3):3):3):3):3);
```
Before we estimate anything, we need to deal with the fact that our data is imperfect. Errors in genome assembly, annotation, and ortholog assignment mean that the data probably contains hidden errors in the number of genes in each gene family in each species. Fragmented annotations or uncollapsed haplotypes in the assembly can lead to duplicate counting of the same gene, and assembly gaps or unannotated proteins can lead to undercounting a gene family. We can deal with this by estimating assembly error that will be taken into account by cafe.
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe
time /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -e -o error_model > error_model.log

less error_model/Base_error_model.txt #view results
```
* `-i`: input file of gene family counts
* `-t`: input phylogeny of taxa
* `-e`: estimate the assembly error. Once it is estimated then you can provide the path to that output but for some reason do not put a space between `-e` and the path to the output file

Now I am running cafe, including the error model.
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe

#No among family rate variation, try with an without error model
time /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -o single_lambda -eerror_model/Base_error_model.txt > single_lambda.log

#no error model
time /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -o single_lambda_noe > single_lambda_noe.log

#Yes among family rate variation, estimate lambda and alpha and three discrete gamma rate categories (skipped)
time /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -k 3 -o three_lambda > three_lambda.log
#with error model (skipped)
time /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -k 3 -o three_lambda_error -eerror_model/Base_error_model.txt > three_lambda_error.log

#Now analyze the large families using the previous values of lambda
time /home/0_PROGRAMS/CAFE/bin/cafexp -i large_filtered_Cafe_input.tsv -t Psittacopasserae.tree -l 0.001319516 -o large_results > large_results.log
```
* `-k`: "Number of gamma categories to use. If specified, the Gamma model will be used to run calculations; otherwise the Base model will be used"
* `-o`: output folder in which to place the results
* `-i`: input file of gene family counts
* `-t`: input phylogeny of taxa
* `-e`: estimate the assembly error. Once it is estimated then you can provide the path to that output but for some reason do not put a space between `-e` and the path to the output file
* `-l`: provide a value for lambda instead of calculating it


**Timing**: Error model: 7m3 (14m53 user); k=3 with error model failed after 14m48 (386m usr).

**Notes**: if you get infinity log likelihoods or it cannot come up with any meaningful starting values, make sure that you do not have any high-variance families in the dataset. They can ruin everything. I realized I still had one gene family with over 100 genes in one taxon, and it was making all the runs fail.

### Interpreting results

The estimates of lambda/epsilon are found in `"$run"_results.txt`

```bash
less Base_asr.tre #gives reconstructed numbers of each gene family at each node
less Base_branch_probabilities.tab #gives calculated likelihood of the gene family size at each node for each gene family
less Base_clade_results.txt #how many families are expanded or contracted at each node

grep "Lambda" ./*/*_results.txt
grep "Epsilon" ./*/*_results.txt
```

Without Rhemel/Hypoch:  

Method|lambda|error
---|---|---|
(bad) unfiltered data|0.011226995547482|0.0077777
error calculation|0.0013047669952907|0.0242847
No error|0.0024234584706124|NA
single lambda|0.0013195807370738|(0.0242847)
3 lambda no error|0.0056132433007506|NA|
3 lambda|0.0024348375726112|(0.0242847)|
large|(0.0013195807370738)|(0.0242847)

With Rhemel/Hypoch:  

Method|lambda|error
---|---|---|
error calculation|0.0015129619926465|0.0276292
No error|0.0023935352133143|NA

As expected, adding an error term makes the estimate of lambda smaller. In this case, almost half as small! Also, filtering out high-variance families had a *huge* effect on lambda, an order of magnitude.
Here is some published data to compare to (from here: https://www.biorxiv.org/content/biorxiv/early/2018/03/22/287086.full.pdf):  
Runs replicating the same settings were very similar.

species |λ (No Error Model) |ε (Estimated error) |λ (Error Model = ε)|
---|---|---|---|
8 bird species in this study |0.00221 |0.01025 |0.00205
12 Drosophila species*| 0.00121| 0.04102 |0.00059
10 mammal species* |0.00238 |0.07324 |0.00186
16 fungi species*| 0.0008 |0.02771 |0.00061

Now that we have run it once, I would like to do it better. I am going to run 10 times and take the average in order to better estimate lambda, and check convergence.
```bash
printf '%s\n' {1..10} | parallel /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -e -o error_model_{1} >> error_model_100runs.log 2>> error_model_100runs.errorlog
#MORE!
printf '%s\n' {11..20} | parallel /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -e -o error_model_{1} >> error_model_100runs.log 2>> error_model_100runs.errorlog
#MORE
printf '%s\n' {21..50} | parallel /home/0_PROGRAMS/CAFE/bin/cafexp -i filtered_Cafe_input.tsv -t Psittacopasserae.tree -e -o error_model_{1} >> error_model_100runs.log 2>> error_model_100runs.errorlog

#Now analyze the large families using the previous values of lambda
printf '%s\n' {1..50} | parallel /home/0_PROGRAMS/CAFE/bin/cafexp -i large_filtered_Cafe_input.tsv -t Psittacopasserae.tree -l 0.001488216 -o large_results_{1} >> large_results_100runs.log
```

Summarise the results of all runs
```bash
grep "Lambda" ./error_model_*/Base_results.txt | sed 's/^.* //g' > lambdas.txt
grep "Epsilon" ./error_model_*/Base_results.txt | sed 's/^.* //g' > epsilons.txt
```
R code:
```R
R
epsilons <- read.table("epsilons.txt")
lambdas <- read.table("lambdas.txt")

mean(epsilons[,1])
sd(epsilons[,1])

mean(lambdas[,1])
sd(lambdas[,1])
q()
n
```
**Reportable results**: 
Without Hypoch/Rhemel:
**Mean epsilon: 0.02427359  
Standard deviation: 2.056175e-05**

**Mean lambda: 0.001319516  
Standard deviation: 6.09575e-7**

With Hypoch/Rhemel:
**Mean epsilon: 0.02786335  
Standard deviation: 0.0001334974**

**Mean lambda: 0.001488216  
Standard deviation: 3.841434e-05**

As you see, there is a low standard deviation between estimates!

Now we can investigate which orthogroups are expanded/contracted
```bash
mkdir analysis
cd analysis

#Check what clade number the taxa of interest are
head -n 3 ../error_model_1/Base_asr.tre #Wilpoe is 29, Rhemel 28, Hypoch 36, (Wilpoe,Rhemel) 37, and Thamnophilidae is 41.

#get a list of the significantly expanded/contracted/rapid genes in Willisornis
printf '%s\n' {1..50} | while read number ; do grep "Wilpoe<29>\*" ../error_model_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > ant_sig_changes_error_model_"$number" ; done 
#get a list of the significantly expanded/contracted/rapid genes in Willisornis/Hypoch/Rhemel (Thamnophilidae)
printf '%s\n' {1..50} | while read number ; do grep "<41>\*" ../error_model_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Thamno_sig_changes_error_model_"$number" ; done 
#get a list of the significantly expanded/contracted/rapid genes in Willisornis/Hypoch/Rhemel (Thamnophilidae)
printf '%s\n' {1..50} | while read number ; do grep "Rhemel<28>\*" ../error_model_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Rhemel_sig_changes_error_model_"$number" ; done 
#get a list of the significantly expanded/contracted/rapid genes in Willisornis/Hypoch/Rhemel (Thamnophilidae)
printf '%s\n' {1..50} | while read number ; do grep "Hypoch<36>\*" ../error_model_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Hypoch_sig_changes_error_model_"$number" ; done 

#count how many were found at each run
wc -l ant_sig_changes_error_model_* #when I ran without the other 2 antbirds, most found 50, except number 6 found 51. With the other two antbirds, it found 73-92 per round.
wc -l Thamno_sig_changes_error_model_* #33-34 in most, and one has 25
wc -l Rhemel_sig_changes_error_model_* #95-120
wc -l Hypoch_sig_changes_error_model_* #72-102

#see what is uniquely found in run 6 (this was the time I had no Hypoch/Rhemel)
printf '%s\n' {1..10} | while read number ; do grep -v -f sig_changes_error_model_"$number" sig_changes_error_model_6 ; done
#OG0013931 is not found in the other runs
#make sure all the others are identical
printf '%s\n' {1..10} | while read number ; do grep -v -f sig_changes_error_model_"$number" sig_changes_error_model_1 ; done
#yep, all 50 were found in all other runs (no output returned by the above command)

#Get list of families that have expanded in Willisornis, whether or not significant
#for the cut -f 1,41: 41 is the column number. It is the taxon number (the number in <>) plus 2 (because the numbers start at zero and there is a label column.)
printf '%s\n' {1..50} | while read number ; do cut -f 1,31 ../error_model_$number/Base_change.tab | grep -v "+0" | grep -v "-" > antexpandedfams_error_model_"$number" ; done #848 in all ten replicates
printf '%s\n' {1..50} | while read number ; do cut -f 1,43 ../error_model_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Thamnoexpandedfams_error_model_"$number" ; done #540 in most (539-543)
printf '%s\n' {1..50} | while read number ; do cut -f 1,30 ../error_model_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Rhemelexpandedfams_error_model_"$number" ; done #233 in all 50 replicates
printf '%s\n' {1..50} | while read number ; do cut -f 1,38 ../error_model_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Hypochexpandedfams_error_model_"$number" ; done #550 in all 50 replicates

wc -l *expandedfams_error_model_* #remember the header is one line!

```
Now, we should summarize these results across all 50 runs to get a single list. Uriel Garcilazo wrote this script for getting a concatenated list of all the orthogroups. Change "nameroot" as needed. I am looping through several. Note I am not looping through the large families runs because they were all identical between replicates.
```python
import os
from os.path import join as jn
import numpy as np

folderpath= os.getcwd()
#nameroot='Thamnoexpandedfams_error_model_'
#nameroot='antexpandedfams_error_model_'
#nameroot='Rhemelexpandedfams_error_model_'
#nameroot='Hypochexpandedfams_error_model_'
#nameroot='ant_sig_changes_error_model_'
#nameroot='Thamno_sig_changes_error_model_'
#nameroot='Rhemel_sig_changes_error_model_'
#nameroot='Hypoch_sig_changes_error_model_'
#nameroot='follower_sig_changes_error_model_'
nameroot='followerexpandedfams_error_model_'

files = [x for x in os.listdir(folderpath) if nameroot in x]
orthologs=''
superstring=''
for file in files:
    with open(jn(folderpath,file),'r') as fl:
        fl=fl.readlines()
        fl=fl[1:]
        for f in fl:
            superstring+=f.strip().split('\t')[0]+'\n'

print(superstring)
export= open(jn(folderpath,'output'),'w')
export.write(superstring)
export.close()
```
Loop through to create antexpandedfams (847)83, Thamnoexpandedfams (539)29, Rhemelexpandedfams (232)108, Hypochexpandedfams (549)85
```bash
cat output | sort | uniq > OGtotaltemp
cat OGtotaltemp | while read line ; do grep -c "$line" output >> tempcounts ; done
paste OGtotaltemp tempcounts >> OGcountstemp
#cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Thamnoexpandedfams ; fi ; done
#cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> antexpandedfams ; fi ; done
#cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Rhemelexpandedfams ; fi ; done
#cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Hypochexpandedfams ; fi ; done
##cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Thamno_sig_changes ; fi ; done
cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> ant_sig_changes ; fi ; done
##cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Rhemel_sig_changes ; fi ; done
##cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> Hypoch_sig_changes ; fi ; done
rm -f OGcountstemp OGtotaltemp tempcounts output

#count how many in each
wc -l *fams
wc -l *changes
```
Now we have a list of the expanded orthogroups, and the significantly changing orthogroups. Compare the two to get significantly expanding orthogroups
```bash
#compare to significant list to keep significantly expanded fams
for taxon in ant Rhemel Hypoch Thamno ; do grep -f "$taxon"_sig_changes "$taxon"expandedfams > "$taxon"expandedfams.sig ; done
wc -l *expandedfams*.sig #ant:26, thamno:17, Rhemel: 12, Hypoch: 45

#repeat for contracted fams
printf '%s\n' {1..10} | while read number ; do cut -f 1,41 ../error_model_$number/Base_change.tab | grep -v "+" > antcontractedfams_error_model_"$number" ; done #1733 in all ten replicates
printf '%s\n' {1..10} | while read number ; do grep -f sig_changes_error_model_"$number" antcontractedfams_error_model_"$number" > antcontractedfams_error_model_"$number".sig ; done 
wc -l antcontractedfams_error_model_*.sig #29 for all replicates
```

Don't forget the large orthgroups that were analyzed separately!
```bash
#repeat for large families
printf '%s\n' {1..50} | while read number ; do grep "Wilpoe<29>\*" ../large_results_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > ant_sig_changes_large_"$number" ; done 
printf '%s\n' {1..50} | while read number ; do grep "<41>\*" ../large_results_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Thamno_sig_changes_large_"$number" ; done 
printf '%s\n' {1..50} | while read number ; do grep "Rhemel<28>\*" ../large_results_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Rhemel_sig_changes_large_"$number" ; done 
printf '%s\n' {1..50} | while read number ; do grep "Hypoch<36>\*" ../large_results_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > Hypoch_sig_changes_large_"$number" ; done 
wc -l *_sig_changes_large_* #Thamno: 0, Rhemel: 3, Hypoch: 2, Willisornis: 3
for taxon in ant Rhemel Hypoch Thamno ; do printf '%s\n' {1..10} | while read number ; do grep -v -f "$taxon"_sig_changes_large_"$number" "$taxon"_sig_changes_large_1 ; done ; done #nothing returned, all runs identical

printf '%s\n' {1..50} | while read number ; do cut -f 1,31 ../large_results_$number/Base_change.tab | grep -v "+0" | grep -v "-" > antexpandedfams_large_"$number" ; done #848 in all ten replicates
printf '%s\n' {1..50} | while read number ; do cut -f 1,43 ../large_results_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Thamnoexpandedfams_large_"$number" ; done #540 in most (539-543)
printf '%s\n' {1..50} | while read number ; do cut -f 1,30 ../large_results_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Rhemelexpandedfams_large_"$number" ; done #233 in all 50 replicates
printf '%s\n' {1..50} | while read number ; do cut -f 1,38 ../large_results_$number/Base_change.tab | grep -v "+0" | grep -v "-" > Hypochexpandedfams_large_"$number" ; done #550 in all 50 replicates
wc -l *expandedfams_large_* #remember the header is one line!

for taxon in ant Rhemel Hypoch Thamno ; do grep -f "$taxon"_sig_changes_large_1 "$taxon"expandedfams_large_1 | cut -f 1 > "$taxon"expandedfams_large.sig ; done
#0 on thamno, Rhemel: 2, Hypoch: 2, ant: 2. Same in all three: OG#1 and 4
```

Combine the results of the large families with the small families
```bash
for taxon in ant Rhemel Hypoch Thamno ; do cat "$taxon"expandedfams_large.sig "$taxon"expandedfams.sig > "$taxon"expandedfams_all.sig ; done

wc -l *expandedfams_all.sig
```
Ok. So now we have a list of orthogroup families that are significantly expanded.  
What do these do?

# GO Analysis

First things first. What genes are in these orthogroups?
```bash
mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/expanded
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/expanded

#get a fasta of the expanded families from the OrthoFinder output
for taxon in ant Rhemel Hypoch Thamno ; do cat ../analysis/"$taxon"expandedfams_all.sig | cut -f 1 | while read Orthogroup ; do cp /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/OrthoFinder/Results_Feb15/WorkingDirectory/OrthoFinder/Results_Mar26_2/Orthogroup_Sequences/"$Orthogroup".fa . ; done ; done

#Get a list of the Willisornis genes in these orthogroups
for taxon in ant Rhemel Hypoch Thamno ; do cat ../analysis/"$taxon"expandedfams.sig | cut -f 1 | while read Orthogroup ; do grep "WilPoe" "$Orthogroup".fa >> "$taxon"_expanded_geneIDs.txt ; done ; done
for taxon in ant Rhemel Hypoch Thamno ; do sed -i 's/>//g' "$taxon"_expanded_geneIDs.txt ; done
for taxon in ant Rhemel Hypoch Thamno ; do sed -i 's/-RA//g' "$taxon"_expanded_geneIDs.txt ; done

#human readable
for taxon in ant Rhemel Hypoch Thamno ; do cat ../analysis/"$taxon"expandedfams.sig | cut -f 1 | while read Orthogroup ; do echo "$Orthogroup" >> human_"$taxon"_expanded_geneIDs.txt ; grep "WilPoe" "$Orthogroup".fa >> human_"$taxon"_expanded_geneIDs.txt ; done ; done
for taxon in ant Rhemel Hypoch Thamno ; do sed -i 's/>//g' human_"$taxon"_expanded_geneIDs.txt ; done
for taxon in ant Rhemel Hypoch Thamno ; do sed -i 's/-RA//g' human_"$taxon"_expanded_geneIDs.txt ; done
```

"we use statistical techniques to tell us how surprised we should be"
There are a lot of available programs to do GO enrichment analysis, but I will be using ErmineJ, because it is free and has a very user-friendly documentation with tutorials and can be run with command line (but there is also a GUI)  .
When you have a binary list of genes (yes in the list vs not in the list) then the analysis to do is ORA, overrepresentation analysis, to see if certain GO/KEGG/other annotations are overrepresenetd in your list compared to the starting set. This means you need two things: a list of all the starting genes and their annotations, and a list of your chosen genes.

There is a nice tutorial [here](https://erminej.msl.ubc.ca/help/tutorials/running-an-analysis-ora/)/

1) [Annotation file](https://erminej.msl.ubc.ca/help/input-files/gene-annotations/)
This is a tab-delimited file  .txt file. The first line is a header. The first column is gene names. The second column is a duplicate of the first column in our case. The third column is for a description. I will make it another duplicate for now. The fourth column is a list of GO terms (eg GO:12345,GO2345,GO5643). 

```bash
cd ~/Willisornis/functional_ann

#extract GO terms
grep -P '\tgene\t' Willisornis.standardplus.functional_5interpro.gff | sed 's/maker.*ID=//g' | sed 's/;Name=.*Dbxref=/\t/g' | sed 's/;Ontology_term=/\t/g' | sed 's/;Name=.*$//g' > Gene_mappings.txt

mkdir /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/enrichment_analysis
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/enrichment_analysis

awk 'BEGIN{FS="\t"}{print $2,$2,$2,$4}' ~/Willisornis/functional_ann/Gene_mappings.txt > GO_mappings.ermineJ.txt

cat > temp_header.txt
gene gene gene GO
#ctrl-d
 
cat temp_header.txt GO_mappings.ermineJ.txt > temp && mv temp GO_mappings.ermineJ.txt
rm temp_header.txt
sed -i 's/ /\t/g' GO_mappings.ermineJ.txt
```

2) [Gene Score File](https://erminej.msl.ubc.ca/help/input-files/gene-scores/)  
The gene score file will tell ErmineJ which genes are in our chosen set and which are not. There must be a 1 line header (the first line will be skipped). There will be two columns: the first has our gene names, and the second has the gene score. In this case it will be a binary score of "NA" (not on our list) or "1" (on our list). As a refinement, we could put the p-value instead of a 1 vs 0.

```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/enrichment_analysis

#this is a bad  way
for taxon in ant Rhemel Hypoch Thamno ; do awk 'BEGIN{FS="\t"}{print $2,"0"}' ~/Willisornis/functional_ann/Gene_mappings.txt > "$taxon"_expanded_genesets ; done
for taxon in ant Rhemel Hypoch Thamno ; do cat /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/expanded/"$taxon"_expanded_geneIDs.txt | sed 's/ .*$//g' | while read gene ; do sed -i "s/$gene 0/$gene 1/g" "$taxon"_expanded_genesets ; done ; done

#make tab separated instead of space separated
for taxon in ant Rhemel Hypoch Thamno ; do sed -i 's/ /\t/g' "$taxon"_expanded_genesets ; done

#repeat for contracted families
#awk 'BEGIN{FS="\t"}{print $2,"0"}' ~/Willisornis/functional_ann/Gene_mappings.txt > contracted_genesets
#cat /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/contracted/contracted_geneIDs.txt | while read gene ; do sed -i "s/$gene 0/$gene 1/g" contracted_genesets ; done
#sed -i 's/ /\t/g' contracted_genesets
```

Ignore the data profile files, we do not have this type of data.

Run ermine!
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/enrichment_analysis

ERMINEJ_HOME=/home/0_PROGRAMS/ermineJ-3.1.2/
for taxon in ant Rhemel Hypoch Thamno ; do /home/0_PROGRAMS/ermineJ-3.1.2/bin/ermineJ.sh -a GO_mappings.ermineJ.txt -s "$taxon"_expanded_genesets -c /home/0_PROGRAMS/ermineJ-3.1.2/go_daily-termdb.rdf-xml.gz -o "$taxon"_expanded_genesets.ermine.results -y 5 -b ; done

#contracted families
/home/0_PROGRAMS/ermineJ-3.1.2/bin/ermineJ.sh -a GO_mappings.ermineJ.txt -s contracted_genesets -c /home/0_PROGRAMS/ermineJ-3.1.2/go_daily-termdb.rdf-xml.gz -o contracted_genesets.ermine.results -y 5 -b

```
* `-y`: min size of gene sets to consider
* `-b`: scores go from small to big

Alternately, rather than focus on just the significant ones, we can do GO analysis on all of the genes, but this is not a very good idea so I would not report it.

Now a ranked list regardless of significance
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/analysis
cut -f 1,41 ../error_model_1/Base_change.tab > orthogroup_expansiveness
cut -f 1,2 /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Jan20/Orthogroups/Orthogroups.tsv > orthogroup2gene.dict
```
Need to replace Orthogroup IDs with gene IDs:  
*This script was a gift from Uriel Garcilazo-Cruz*  
```bash
cat > orthogroup2gene_scores.py
```
```python
import os
file1="orthogroup_expansiveness" #This file contains the list of orthologs and their scores
file2='orthogroup2gene.dict' #This file contains the list of gene names that belong to each ortholog name
mainpath=os.getcwd()
fileI=[x.strip().split('\t') for x in open(os.path.join(mainpath,file1),'r').readlines()]
fileII=[x.strip().split('\t') for x in open(os.path.join(mainpath,file2),'r').readlines()]
count=0
#for i in range(len(fileI)):
#    if count<100:
#        if i!=0:#The first row in the table are the headers, which we don't need
#            print(fileI[i])
#    else:
#        break
#    count+=1
first=True
finalstringfileII=[]
for h in fileII:
    if not first:
        try:
            for name in h[1].split(', '):
                finalstringfileII.append([h[0],name]) #Put the names of each gene into a new row with its ortholog
        except:
            pass
    else:
        pass
    first=False

finalstring=''
for r in finalstringfileII:
    for e in fileI:
        if r[0] == e[0]:
            finalstring+=r[1]+'\t'+e[1]+'\n'

export=open(os.path.join(mainpath,"output.txt"),"w")
export.write(finalstring)
export.close()
```
```bash
python orthogroup2gene_scores.py
mv output.txt genescores_expansiveness.txt
sed -i 's/-RA//g' genescores_expansiveness.txt
sed -i 's/+//g' genescores_expansiveness.txt

```

```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/enrichment_analysis
ERMINEJ_HOME=/home/0_PROGRAMS/ermineJ-3.1.2/

/home/0_PROGRAMS/ermineJ-3.1.2/bin/ermineJ.sh -a GO_mappings.ermineJ.txt -s ../analysis/genescores_expansiveness.txt -c /home/0_PROGRAMS/ermineJ-3.1.2/go_daily-termdb.rdf-xml.gz -o expansiveness_genesets.ermine.results -y 5 -b
#Gene set resampling
/home/0_PROGRAMS/ermineJ-3.1.2/bin/ermineJ.sh -a GO_mappings.ermineJ.txt -s ../analysis/genescores_expansiveness.txt -c /home/0_PROGRAMS/ermineJ-3.1.2/go_daily-termdb.rdf-xml.gz -o expansiveness_genesets.ermine.GSR.results -y 5 -b -n GSR

```

# Visualization
I am going to use R to visualize the results on a phylogeny. I would like to show the number of genes within orthogroups at each tip of the phylogeny. To do this, I want a:
* csv file
* header = orthogroup name
* column 1 = species names
* data = counts of each orthogroup.

Starting with Cafe_input.tsv

```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe
mkdir phylogenomics
cd phylogenomics

cut -f 2-100 ../Cafe_input.tsv > gene_counts.csv #remove the null first column

#Also, we will need the corresponding phylogeny
cp ../Psittacopasserae.tree .
```

Switch the names back
```bash
sed -i 's/Tingut/Tinamus guttatus/g' gene_counts.csv
sed -i 's/Notper/Nothoprocta perdicaria/g' gene_counts.csv
sed -i 's/Anscyg/Anser cygnoides/g' gene_counts.csv
sed -i 's/Cotjap/Coturnix japonica/g' gene_counts.csv
sed -i 's/Melgal/Meleagris gallopavo/g' gene_counts.csv
sed -i 's/Chavoc/Charadrius vociferus/g' gene_counts.csv
sed -i 's/Lepdis/Leptosomus discolor/g' gene_counts.csv
sed -i 's/Mernub/Merops nubicus/g' gene_counts.csv
sed -i 's/Athcun/Athene cunicularia/g' gene_counts.csv
sed -i 's/Colstr/Colius striatus/g' gene_counts.csv
sed -i 's/Falche/Falco cherug/g' gene_counts.csv
sed -i 's/Amacol/Amazona collaria/g' gene_counts.csv
sed -i 's/Wilpoe/Willisornis poecilinotus/g' gene_counts.csv
sed -i 's/Manvit/Manacus vitellinus/g' gene_counts.csv
sed -i 's/Neochr/Neopelma chrysocephalum/g' gene_counts.csv
sed -i 's/Corbra/Corvus brachyrhynchos/g' gene_counts.csv
sed -i 's/Cyacae/Cyanistes caeruleus/g' gene_counts.csv
sed -i 's/Taegut/Taeniopygia guttata/g' gene_counts.csv
sed -i 's/Sercan/Serinus canaria/g' gene_counts.csv
sed -i 's/Zonalb/Zonotrichia albicollis/g' gene_counts.csv
sed -i 's/Strcam/Struthio camelus/g' gene_counts.csv
sed -i 's/Dronov/Dromaius novaehollandiae/g' gene_counts.csv
sed -i 's/Apthaa/Apteryx haastii/g' gene_counts.csv
sed -i 's/Ansbra/Anser brachyrhynchus/g' gene_counts.csv
sed -i 's/Galgal/Gallus gallus/g' gene_counts.csv
sed -i 's/Phacol/Phasianus colchicus/g' gene_counts.csv
sed -i 's/Calpug/Calidris pugnax/g' gene_counts.csv
sed -i 's/Apavit/Apaloderma vittatum/g' gene_counts.csv
sed -i 's/Picpub/Picoides pubescens/g' gene_counts.csv
sed -i 's/Accnis/Accipiter nisus/g' gene_counts.csv
sed -i 's/Carcri/Cariama cristata/g' gene_counts.csv
sed -i 's/Strhab/Strigops habroptila/g' gene_counts.csv
sed -i 's/Melund/Melopsittacus undulatus/g' gene_counts.csv
sed -i 's/Hypoch/Hypocnemis ochrogyna/g' gene_counts.csv
sed -i 's/Pipfil/Pipra filicauda/g' gene_counts.csv
sed -i 's/Emptra/Empidonax trailli/g' gene_counts.csv
sed -i 's/Psehum/Pseudopodoces humilis/g' gene_counts.csv
sed -i 's/Ficalb/Ficedula albicollis/g' gene_counts.csv
sed -i 's/Lonstr/Lonchura striata/g' gene_counts.csv
sed -i 's/Campar/Camarhynchus parvulus/g' gene_counts.csv
sed -i 's/Junhie/Junco hyemalis/g' gene_counts.csv
sed -i 's/Aptaus/Apteryx mantelli mantelli/g' gene_counts.csv
sed -i 's/Aptowe/Apteryx owenii/g' gene_counts.csv
sed -i 's/Nummel/Numida meleagris/g' gene_counts.csv
sed -i 's/Pavcri/Pavo cristatus/g' gene_counts.csv
sed -i 's/Chrpic/Chrysolophus pictus/g' gene_counts.csv
sed -i 's/Calpyg/Calidris pygmaea/g' gene_counts.csv
sed -i 's/Bucrhi/Buceros rhinoceros/g' gene_counts.csv
sed -i 's/Tytalb/Tyto alba/g' gene_counts.csv
sed -i 's/Aquchr/Aquila chrysaetos/g' gene_counts.csv
sed -i 's/Falper/Falco peregrinus/g' gene_counts.csv
sed -i 's/Nesnot/Nestor notabilis/g' gene_counts.csv
sed -i 's/Acachl/Acanthisitta chloris/g' gene_counts.csv
sed -i 's/Lepcor/Lepidothrix coronata/g' gene_counts.csv
sed -i 's/Coralt/Corapipo altera/g' gene_counts.csv
sed -i 's/Corcor/Corvus cornix/g' gene_counts.csv
sed -i 's/Parmaj/Parus major/g' gene_counts.csv
sed -i 's/Stuvul/Sturnus vulgaris/g' gene_counts.csv
sed -i 's/Erygou/Erythrura gouldiae/g' gene_counts.csv
sed -i 's/Geofor/Geospiza fortis/g' gene_counts.csv
sed -i 's/Aptrow/Apteryx rowi/g' gene_counts.csv
sed -i 's/Anapla/Anas platyrhynchos/g' gene_counts.csv
sed -i 's/Tauery/Tauraco erythrolophus/g' gene_counts.csv
sed -i 's/Chlmac/Chlamydotis macqueenii/g' gene_counts.csv
sed -i 's/Cuccan/Cuculus canorus/g' gene_counts.csv
sed -i 's/Gavste/Gavia stellata/g' gene_counts.csv
sed -i 's/Pelcri/Pelecanus crispus/g' gene_counts.csv
sed -i 's/Nipnip/Nipponia nippon/g' gene_counts.csv
sed -i 's/Colliv/Columba livia/g' gene_counts.csv
sed -i 's/Calana/Calypte anna/g' gene_counts.csv
sed -i 's/Antcar/Antrostomus carolinensis/g' gene_counts.csv
sed -i 's/Eurhel/Eurypyga helias/g' gene_counts.csv
sed -i 's/Egrgar/Egretta garzetta/g' gene_counts.csv
sed -i 's/Phacar/Phalacrocorax carbo/g' gene_counts.csv
sed -i 's/Fulgla/Fulmarus glacialis/g' gene_counts.csv
sed -i 's/Aptfor/Aptenodytes forsteri/g' gene_counts.csv
sed -i 's/Pygade/Pygoscelis adeliae/g' gene_counts.csv
sed -i 's/Rhemel/Rhegmatorhina melanosticta/g' gene_counts.csv
sed -i 's/Balreg/Balearica regulorum/g' gene_counts.csv
sed -i 's/Opihoa/Opisthocomus hoazin/g' gene_counts.csv
sed -i 's/Ptegut/Pterocles gutturalis/g' gene_counts.csv
sed -i 's/Mesuni/Mesitornis unicolor/g' gene_counts.csv
sed -i 's/Phalep/Phaethon lepturus/g' gene_counts.csv
sed -i 's/Chapel/Chaetura pelagica/g' gene_counts.csv
```
I have had some issues installing R pakages on the main server, so I am doing this on my personal computer in Rstudio for the graphical interface for viewing plots.

This R script is what I used to plot the orthogroup counts on the phylogenetic tree to make figure 4.
```R
# This script was used to visualize the distribution of orthogroups across the phylogeny of birds analyzed with Cafe.

setwd("/Users/elsemikkelsen/Desktop/Willisornis/orthogroups") #change to the directory containing the tree and orthogroup data
#First, install some libraries
# you only need to do this one time on your system
#install.packages("dplyr") #if not already installed
#install.packages("phytools")
#install.packages("readr") 
#install.packages("ape")

#load the libraries required by this script
library(ape)
library(readr)
library(dplyr)
library(phytools) #for ancestral state reconstruction
library(tidyverse)

#First, we need a phylogeny
tree<-read.tree("Psittacopasserae.tree") #ape package - interpret the text string as a tree

#you can plot this to visualize the tree if you like
plot(tree,no.margin=TRUE,edge.width=2) #a straight edge tree
roundPhylogram(tree) #a rounded tree (phytools package)
plot(unroot(tree),type="unrooted",no.margin=TRUE,lab4ut="axial", edge.width=2) #an unrooted tree

#view structure of the tree
str(tree)
is.binary.tree(tree) #check that the tree in binary (should be True)

#read in the dataframe of orthogroup counts
genecounts<-read.delim("gene_counts.csv",header=TRUE,row.names=1)
#remove extraneous species not analyzed in Cafe (Falco, which was only used for positive selection analyses)
genecounts <- genecounts %>%
  select(-Falco.cherug, -Falco.peregrinus)

#transpose the dataframe so that taxa are the rows and orthogroups are the columns
genecounts<-t(genecounts) 

#This function will create a plot of the specified orthogroup given a genecounts and tree file
plot_orthogroups <- function(Orthogroup, genecounts, tree){
  #isolate this orthogroup as a matrix, giving count values for each taxon
  OG_data <- as.matrix(genecounts)[,Orthogroup]
  #estimate ancestral states at each node, with 95% CI
  OG_anc <- fastAnc(tree, OG_data, vars=T, CI=T) #uses a phylogeny, and a matrix of character states (counts of gene numbers)
  
  #Now plot the data onto a tree to visualize it
  OG_tree <- contMap(tree, OG_data, plot=F) #map continuous traits onto the tree
  OG_tree<-setMap(OG_tree,invert=TRUE) #invert colours so that red is higher and blue is lower
  
  #plot(OG0_tree,type="fan",legend=0.7*max(nodeHeights(birdtree)),fsize=c(0.7,0.9)) #fan tree
  plot(OG_tree,fsize=c(0.4,1),outline=F,lwd=c(3,7),leg.txt=Orthogroup) #plot the tree. Do not use outlines (looks crowded)
}

### First option: visualize a particular orthogroup
#choose what orthogroup to analyze right now
Orthogroup <- "OG0000107" #I am going to show this orthogroup which appeared the most striking for Willisornis
#Orthogroup <- "OG0000979"
for(i in 1:length(Orthogroup)){
  quartz()
  plot_orthogroups(Orthogroup[i], genecounts, tree)
}

### Second option: visualize all orthogroups expanded in particular clades

#Here are some examples of orthogroup sets that I visualized. You can make a list of whichever groups you would like to visualize.
#Orthogroups expanded in Thamnophilidae
thamnophilidae <- c("OG0000680", "OG0000974", "OG0001023", "OG0001029", "OG0001050", "OG0001239", "OG0001873", "OG0001915", "OG0012573", "OG0012968")

#Orthogroups expanded in Thamnophilidae & Willisornis
thamnophilidae.Willisornis <- c("OG0000010", "OG0001105", "OG0001178", "OG0011455")

#Orthogroups expanded in Thamnophilidae & Rhegmatorhina
thamnophilidae.Rhegmatorhina <- c("OG0000264", "OG0011571", "OG0011529")

#species specific expansions
Willisornis <- c("OG0000011", "OG0000012", "OG0000015", "OG0000019", "OG0000029", "OG0000037", "OG0000041", "OG0000055", "OG0000057", "OG0000065", "OG0000107", "OG0000210", "OG0000229", "OG0000283", "OG0000979", "OG0001030", "OG0001053", "OG0001054", "OG0001133", "OG0001157", "OG0001347", "OG0014009")
Rhegmatorhina <- c("OG0000014", "OG0000035", "OG0000077", "OG0000087", "OG0000223", "OG0000570", "OG0000680", "OG0001029", "OG0012957")

#total, including overlaps
Thamnophilidae.all <- c("OG0000010", "OG0000264", "OG0000680", "OG0000974", "OG0001023", "OG0001029", "OG0001050", "OG0001105", "OG0001178", "OG0001239", "OG0001873", "OG0001915", "OG0011455", "OG0011529", "OG0011571", "OG0012573", "OG0012968")

#Once you have picked a list, you can visualize it (replace Thamnophilidae.all with the name of the list containing the groups of interest)
for(i in 1:length(Thamnophilidae.all)){
  plot_orthogroups(Thamnophilidae.all[i], genecounts, tree)
}

### Tabulate orthogroups for antbirds
#I would like to get a table of the expanded orthogroups for the antbirds

#here are those orthogroups that are expanded in any antbird:
tabulate <- c("OG0000001", "OG0000004", "OG0000006", "OG0000010", "OG0000011", "OG0000012", "OG0000013", "OG0000014", "OG0000015", "OG0000016", "OG0000017", "OG0000019", "OG0000020", "OG0000021", "OG0000022", "OG0000024", "OG0000029", "OG0000035", "OG0000037", "OG0000041", "OG0000048", "OG0000051", "OG0000053", "OG0000055", "OG0000057", "OG0000065", "OG0000069", "OG0000077", "OG0000082", "OG0000087", "OG0000090", "OG0000094", "OG0000106", "OG0000107", "OG0000158", "OG0000210", "OG0000223", "OG0000229", "OG0000264", "OG0000283", "OG0000322", "OG0000418", "OG0000570", "OG0000680", "OG0000734", "OG0000756", "OG0000819", "OG0000873", "OG0000974", "OG0000979", "OG0001023", "OG0001029", "OG0001030", "OG0001050", "OG0001053", "OG0001054", "OG0001100", "OG0001105", "OG0001133", "OG0001152", "OG0001157", "OG0001178", "OG0001239", "OG0001347", "OG0001366", "OG0001580", "OG0001626", "OG0001723", "OG0001873", "OG0001915", "OG0002132", "OG0004927", "OG0011455", "OG0011529", "OG0011571", "OG0012573", "OG0012957", "OG0012968", "OG0013277", "OG0013520", "OG0013734", "OG0014009")
read.delim("gene_counts.csv",header=TRUE,row.names=1) %>%
  select(Willisornis.poecilinotus, Rhegmatorhina.melanosticta, Hypocnemis.ochrogyna) %>% #keep only the antbird data
  rownames_to_column %>%
  filter(rowname %in% tabulate) #filter to keep the data only for orthogroups (rownames) in the list

#Now you can plot all of these groups if you like
for(i in 1:length(tabulate)){
  quartz()
  plot_orthogroups(tabulate[i], genecounts, tree)
}
```

# Appendix: Expansions shared by Willisornis and Rhegmatorhina
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/analysis

#get a list of the significantly expanded/contracted/rapid genes in Willisornis/Rhegmatorhina ancestor
printf '%s\n' {1..50} | while read number ; do grep "<37>\*" ../error_model_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > follower_sig_changes_error_model_"$number" ; done 

#count how many were found at each run
wc -l follower_sig_changes_error_model_* #15-24 per round

#Get list of families that have expanded in them, whether or not significant
#for the cut -f 1,41: 41 is the column number. It is the taxon number (the number in <>) plus 2 (because the numbers start at zero and there is a label column.)
printf '%s\n' {1..50} | while read number ; do cut -f 1,39 ../error_model_$number/Base_change.tab | grep -v "+0" | grep -v "-" > followerexpandedfams_error_model_"$number" ; done #74-75 

wc -l followerexpandedfams_error_model_* #remember the header is one line!

#run the python code from above

cat output | sort | uniq > OGtotaltemp
cat OGtotaltemp | while read line ; do grep -c "$line" output >> tempcounts ; done
paste OGtotaltemp tempcounts >> OGcountstemp
cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> followerexpandedfams ; fi ; done

#cat OGcountstemp | while read orthogroup counts ; do if [ "$counts" -gt 47 ] ; then echo "$orthogroup" >> follower_sig_changes ; fi ; done

rm -f OGcountstemp OGtotaltemp tempcounts output
wc -l followerexpandedfams #73
wc -l follower_sig_changes #19

#compare to significant list to keep significantly expanded fams
for taxon in follower ; do grep -f follower_sig_changes followerexpandedfams > followerexpandedfams.sig ; done
wc -l followerexpandedfams.sig #3
```

Don't forget the large orthgroups that were analyzed separately!
```bash
#repeat for large families
printf '%s\n' {1..50} | while read number ; do cut -f 1,39 ../large_results_$number/Base_change.tab | grep -v "+0" | grep -v "-" > followerexpandedfams_large_"$number" ; done

printf '%s\n' {1..50} | while read number ; do grep "<37>\*" ../large_results_$number/Base_asr.tre | sed 's/^.*OG/OG/g' | sed 's/ = .*$//g' > follower_sig_changes_large_"$number" ; done 
wc -l follower_sig_changes_large_* #0
#None are significant, no need to look further.
```

First things first. What genes are in these orthogroups?
```bash
cd /home/0_GENOMES1/0_WEIRLAB_GENOMES_CHROMIUMX/Willisornis/Cafe/expanded

#get a fasta of the expanded families from the OrthoFinder output
for taxon in follower ; do cat ../analysis/"$taxon"expandedfams.sig | cut -f 1 | while read Orthogroup ; do cp /home/0_GENOMES1/0_Genetic_resources/orthologues/PasserinesPlus/OrthoFinder/Results_Feb02_EnsNP/WorkingDirectory/OrthoFinder/Results_Feb02_2/WorkingDirectory/OrthoFinder/Results_Feb15/WorkingDirectory/OrthoFinder/Results_Mar26_2/Orthogroup_Sequences/"$Orthogroup".fa . ; done ; done

#Get a list of the Willisornis genes in these orthogroups
for taxon in follower ; do cat ../analysis/"$taxon"expandedfams.sig | cut -f 1 | while read Orthogroup ; do grep "WilPoe" "$Orthogroup".fa >> "$taxon"_expanded_geneIDs.txt ; done ; done
for taxon in follower ; do sed -i 's/>//g' "$taxon"_expanded_geneIDs.txt ; done
for taxon in follower ; do sed -i 's/-RA//g' "$taxon"_expanded_geneIDs.txt ; done

#human readable
for taxon in follower ; do cat ../analysis/"$taxon"expandedfams.sig | cut -f 1 | while read Orthogroup ; do echo "$Orthogroup" >> human_"$taxon"_expanded_geneIDs.txt ; grep "WilPoe" "$Orthogroup".fa >> human_"$taxon"_expanded_geneIDs.txt ; done ; done
for taxon in follower ; do sed -i 's/>//g' human_"$taxon"_expanded_geneIDs.txt ; done
for taxon in follower ; do sed -i 's/-RA//g' human_"$taxon"_expanded_geneIDs.txt ; done
```